{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "24658622",
   "metadata": {},
   "source": [
    "A notebook for practicing MAFFT (Multiple Alignment using Fast Fourier Transform).  The manual page is here: https://mafft.cbrc.jp/alignment/software/manual/manual.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "cd613310",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v7.490 (2021/Oct/30)\n",
      "Description:\tUbuntu 20.04.2 LTS\n"
     ]
    }
   ],
   "source": [
    "! mafft --version\n",
    "! lsb_release -d "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a42b309e",
   "metadata": {},
   "source": [
    "# Looking at the files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b281078",
   "metadata": {},
   "source": [
    "The fasta files I am practicing with are the 27 that passed hi_qc, from `result.illumina.20210513` (see the second commit of filter_fa_qc.sh for details on their extraction).  The files are:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "ebfd35cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "practice_fastas/COG111-E-NC_S83.primertrimmed.consensus.fa\r\n",
      "practice_fastas/COG111-NC_S89.primertrimmed.consensus.fa\r\n",
      "practice_fastas/COG111-PC_S86.primertrimmed.consensus.fa\r\n",
      "practice_fastas/COG112-PC_S114.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-220299_S80.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-22A50F_S96.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-22AE40_S97.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-22D87C_S81.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-22DDC8_S95.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-22DFFF_S94.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-230B31_S87.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-23264D_S112.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-23267A_S106.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-232689_S137.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-2326A7_S131.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-2326C5_S143.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-2326F2_S128.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-232829_S109.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-232847_S115.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-232883_S142.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-232892_S118.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-2328DE_S133.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-232935_S135.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-232971_S125.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-232980_S124.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-23299F_S140.primertrimmed.consensus.fa\r\n",
      "practice_fastas/NORW-232BD5_S138.primertrimmed.consensus.fa\r\n"
     ]
    }
   ],
   "source": [
    "! ls practice_fastas/*.consensus.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f96eb709",
   "metadata": {},
   "source": [
    "Let's take a look at a random file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "fe4a8baa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Consensus_NORW-220299_S80.primertrimmed.consensus_threshold_0.75_quality_20\r\n",
      "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNCTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACNGGACACGAGTAACNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNGCTCGTACGTGGCTTTGGAGACTCCGTGGAGGAGGTCTTATCAGAGGCACGTCAACATCTTAAAGATGGCACTTGTGGCTTAGTAGAAGTTGAAAAAGGCGTTTTGCCTCAACTTGAACAGCCCTATGTGTTCATCAAACGTTCGGATGCTCGAACTGCACCTCATGGTCATGTTATGGTTGAGCTGGTAGCAGAACTCGAAGGCATTCAGTACGGTCGTAGTGGTGAGACACTTGGTGTCCTTGTCCCTCATGTGGGCGAAATACCAGTGGCTTACCGCAAGGTTCTTCTTCGTAAGAACGGTAATAAAGGAGCTGGTGGCCATAGTTACGGCGCCGATCTAAAGTCATTTGACTTAGGCGACGAGCTTGGCACTGATCCTTATGAAGATTTTCAAGAAAACTGGAACACTAAACATAGCAGTGGTGTTACCCGTGAACTCATGCGTGAGCTTAACGGAGGGGCATACACTCGCTATGTCGATAACAACTTCTGTGGCCCTGATGGCTACCCTCTTGAGTGCATTAAAGACCTTCTAGCACGTGCTGGTAAAGCTTCATGCACTTTGTCTGAACAACTGGACTTTATTGACACTAAGAGGGGTGTATACTGCTGCCGTGAACATGAGCATGAAATTGCTTGGTACACGGAACGTTCTGAAAAGAGCTATGAATTGCAGACACCTTTTGAAATTAAATTGGCAAAGAAATTTGACACCTTCAATGGGGAATGTCCAAATTTTGTATTTCCCTTAAATTCCATAATCAAGACTATTCAACCAAGGGTTGAAAAGAAAAAGCTTGATGGCTTTATGGGTAGAATTCGATCTGTCTATCCAGTTGCGTCACCAAATGAATGCAACCAAATGTGCCTTTCAACTCTCATGAAGTGTGATCATTGTGGTGAAACTTCATGGCAGACGGGCGATTTTGTTAAAGCCACTTGCGAATTTTGTGGCACTGAGAATTTGACTAAAGAAGGTGCCACTACTTGTGGTTACTTACCCCAAAATGCTGTTGTTAAAATTTATTGTCCAGCATGTCACAATTCAGAAGTAGGACCTGAGCATAGTCTTGCCGAATACCATAATGAATCTGGCTTGAAAACCATTCTTCGTAAGGGTGGTCGCACTATTGCCTTTGGAGGCTGTGTGTTCTCTTATGTTGGTTGCCATAACAAGTGTGCCTATTGGGTTCCACGTGCTAGCGCTAACATAGGTTGTAACCATACAGGTGTTGTTGGAGAAGGTTCCGAAGGTCTTAATGACAACCTTCTTGAAATACTCCAAAAAGAGAAAGTCAACATCAATATTGTTGGTGACTTTAAACTTAATGAAGAGATCGCCATTATTTTGGCATCTTTTTCTGCTTCCACAAGTGCTTTTGTGGAAACTGTGAAAGGTTTGGATTATAAAGCATTCAAACAAATTGTTGAATCCTGTGGTAATTTTAAAGTTACAAAAGGAAAAGCTAAAAAAGGTGCCTGGAATATTGGTGAACAGAAATCAATACTGAGTCCTCTTTATGCATTTGCATCAGAGGCTGCTCGTGTTGTACGATCAATTTTCTCCCGCACTCTTGAAACTGCTCAAAATTCTGTGCGTGTTTTACAGAAGGCCGCTATAACAATACTAGATGGAATTTCACAGTATTCACTGAGACTCATTGATGCTATGATGTTCACATCTGATTTGGCTACTAACAATCTAGTTGTAATGGCCTACATTACAGGTGGTGTTGTTCAGTTGACTTCGCAGTGGCTAACTAACATCTTTGGCACTGTTTATGAAAAACTCAAACCCGTCCTTGATTGGCTTGAAGAGAAGTTTAAGGAAGGTGTAGAGTTTCTTAGAGACGGTTGGGAAATTGTTAAATTTATCTCAACCTGTGCTTGTGAAATTGTCGGTGGACAAATTGTCACCTGTGCAAAGGAAATTAAGGAGAGTGTTCAGACATTCTTTAAGCTTGTAAATAAATTTTTGGCTTTGTGTGCTGACTCTATCATTATTGGTGGAGCTAAACTTAAAGCCTTGAATTTAGGTGAAACATTTGTCACGCACTCAAAGGGATTGTACAGAAAGTGTGTTAAATCCAGAGAAGAAACTGGCCTACTCATGCCTCTAAAAGCCCCAAAAGAAATTATCTTCTTAGAGGGAGAAACACTTCCCACAGAAGTGTTAACAGAGGAAGTTGTCTTGAAAACTGGTGATTTACAACCATTAGAACAACCTACTAGTGAAGCTGTTGAAGCTCCATTGGTTGGTACACCAGTTTGTATTAACGGGCTTATGTTGCTCGAAATCAAAGACACAGAAAAGTACTGTGCCCTTGCACCTAATATGATGGTAACAAACAATACCTTCACACTCAAAGGCGGTGCACCAACAAAGGTTACTTTTGGTGATGACACTGTGATAGAAGTGCAAGGTTACAAGAGTGTGAATATCACTTTTGAACTTGATGAAAGGATTGATAAAGTACTTAATGAGAAGTGCTCTGCCTATACAGTTGAACTCGGTACAGAAGTAAATGAGTTCGCCTGTGTTGTGGCAGATGCTGTCATAAAAACTTTGCAACCAGTATCTGAATTACTTACACCACTGGGCATTGATTTAGATGAGTGGAGTATGGCTACATACTACTTATTTGATGAGTCTGGTGAGTTTAAATTGGCTTCACATATGTATTGTTCTTTTTACCCTCCAGATGAGGATGAAGAAGAAGGTGATTGTGAAGAAGAAGAGTTTGAGCCATCAACTCAATATGAGTATGGTACTGAAGATGATTACCAAGGTAAACCTTTGGAATTTGGTGCCACTTCTGCTGCTCTTCAACTTGAAGAAGAGCAAGAAGAAGATTGGTTAGATGATGATAGTCAACAAACTGTTGGTCAACAAGACGGCAGTGAGGACAATCAGACAACTATTATTCAAACAATTGTTGAGGTTCAACCTCAATTAGAGATGGAACTTACACCAGTTGTTCAGACTATTGAAGTGAATAGTTTTAGTGGTTATTTAAAACTTACTGACAATGTATACATTAAAAATGCAGACATTGTGGAAGAAGCTAAAAAGGTAAAACCAACAGTGGTTGTTAATGCAGCCAATGTTTACCTTAAACATGGAGGAGGTGTTGCAGGAGCCTTAAATAAGGCTACTAACAATGCCATGCAAGTTGAATCTGATGATTACATAGCTACTAATGGACCACTTAAAGTGGGTGGTAGTTGTGTTTTAAGCGGACACAATCTTGCTAAACACTGTCTTCATGTTGTCGGCCCAAATGTTAACAAAGGTGAAGACATTCAACTTCTTAAGAGTGCTTATGAAAATTTTAATCAGCACGAAGTTCTACTTGCACCATTATTATCAGCTGGTATTTTTGGTGCTGACCCTATACATTCTTTAAGAGTTTGTGTAGATACTGTTCGCACAAATGTCTACTTAGCTGTCTTTGATAAAAATCTCTATGACAAACTTGTTTCAAGCTTTTTGGAAATGAAGAGTGAAAAGCAAGTTGAACAAAAGATCGCTGAGATTCCTAAAGAGGAAGTTAAGCCATTTATAACTGAAAGTAAACCTTCAGTTGAACAGAGAAAACAAGATGATAAGAAAATCAAAGCTTGTGTTGAAGAAGTTACAACAACTCTGGAAGAAACTAAGTTCCTCACAGAAAACTTGTTACTTTATATTGACATTAATGGCAATCTTCATCCAGATTCTGCCACTCTTGTTAGTGACATTGACATCACTTTCTTAAAGAAAGATGCTCCATATATAGTGGGTGATGTTGTTCAAGAGGGTGTTTTAACTGCTGTGGTTATACCTACTAAAAAGGCTGGTGGCACTACTGAAATGCTAGCGAAAGCTTTGAGAAAAGTGCCAACAGACAATTATATAACCACTTACCCGGGTCAGGGTTTAAATGGTTACACTGTAGAGGAGGCAAAGACAGTGCTTAAAAAGTGTAAAAGTGCCTTTTACATTCTACCATCTATTATCTCTAATGAGAAGCAAGAAATTCTTGGAACTGTTTCTTGGAATTTGCGAGAAATGCTTGCACATGCAGAAGAAACACGCAAATTAATGCCTGTCTGTGTGGAAACTAAAGCCATAGTTTCAACTATACAGCGTAAATATAAGGGTATTAAAATACAAGAGGGTGTGGTTGATTATGGTGCTAGATTTTACTTTTACACCAGTAAAACAACTGTAGCGTCACTTATCAACACACTTAACGATCTAAATGAAACTCTTGTTACAATGCCACTTGGCTATGTAACACATGGCTTAAATTTGGAAGAAGCTGCTCGGTATATGAGATCTCTCAAAGTGCCAGCTACAGTTTCTGTTTCTTCACCTGATGCTGTTACAGCGTATAATGGTTATCTTACTTCTTCTTCTAAAACACCTGAAGAACATTTTATTGAAACCATCTCACTTGCTGGTTCCTATAAAGATTGGTCCTATTCTGGACAATCTACACAACTAGGTATAGAATTTCTTAAGAGAGGTGATAAAAGTGTATATTACACTAGTAATCCTACCACATTCCACCTAGATGGTGAAGTTATCACCTTTGACAATCTTAAGACACTTCTTTCTTTGAGAGAAGTGAGGACTATTAAGGTGTTTACAACAGTAGACAACATTAACCTCCACACGCAAGTTGTGGACATGTCAATGACATATGGACAACAGTTTGGTCCAACTTATTTGGATGGAGCTGATGTTACTAAAATAAAACCTCATAATTCACATGAAGGTAAAACATTTTATGTTTTACCTAATGATGACACTCTACGTGTTGAGGCTTTTGAGTACTACCACACAACTGATCCTAGTTTTCTGGGTAGGTACATGTCAGCATTAAATCACACTAAAAAGTGGAAATACCCACAAGTTAATGGTTTAACTTCTATTAAATGGGCAGATAACAACTGTTATCTTGCCACTGCATTGTTAACACTCCAACAAATAGAGTTGAAGTTTAATCCACCTGCTCTACAAGATGCTTATTACAGAGCAAGGGCTGGTGAAGCTGATAACTTTTGTGCACTTATCTTAGCCTACTGTAATAAGACAGTAGGTGAGTTAGGTGATGTTAGAGAAACAATGAGTTACTTGTTTCAACATGCCAATTTAGATTCTTGCAAAAGAGTCTTGAACGTGGTGTGTAAAACTTGTGGACAACAGCAGACAACCCTTAAGGGTGTAGAAGCTGTTATGTACATGGGCACACTTTCTTATGAACAATTTAAGAAAGGTGTTCAGATACCTTGTACGTGTGGTAAACAAGCTACAAAATATCTAGTACAACAGGAGTCACCTTTTGTTATGATGTCAGCACCACCTGCTCAGTATGAACTTAAGCATGGTACATTTACTTGTGCTAGTGAGTACACTGGTAATTACCAGTGTGGTCACTATAAACATATAACTTCTAAAGAAACTTTGTATTGCATAGACGGTGCTTTACTTACAAAGTCCTCAGAATACAAAGGTCCTATTACGGATGTTTTCTACAAAGAAAACAGTTACACAACAACCATAAAACCAGTTACTTATAAATTGGATGGTGTTGTTTGTACAGAAATTGACCCTAAGTTGGACAATTATTATAAGAAAGACAATTCTTATTTTACAGAGCAACCAATTGATCTTGTACCAAACCAACCATATCCAAACGCAAGCTTCGATAATTTTAAGTTTGTATGTGATAATATCAAATTTGCTGATGATTTAAACCAGTTAACTGGTTATAAGAAACCTGCTTCAAGAGAGCTTAAAGTTACATTTTTCCCTGACTTAAATGGTGATGTGGTGGCTATTGATTATAAACACTACACACCCTCTTTTAAGAAAGGAGCTAAATTGTTACATAAACCTATTGTTTGGCATGTTAACAATGCAACTAATAAAGCCACGTATAAACCAAATACCTGGTGTATACGTTGTCTTTGGAGCACAAAACCAGTTGAAACATCAAATTCGTTTGATGTACTGAAGTCAGAGGACGCGCAGGGAATGGATAATCTTGCCTGCGAAGATCTAAAACCAGTCTCTGAAGAAGTAGTGGAAAATCCTACCATACAGAAAGACGTTCTTGAGTGTAATGTGAAAACTACCGAAGTTGTAGGAGACATTATACTTAAACCAGCAAATAATAGTTTAAAAATTACAGAAGAGGTTGGCCACACAGATCTAATGGCTGCTTATGTAGACAATTCTAGTCTTACTATTAAGAAACCTAATGAATTATCTAGAGTATTAGGTTTGAAAACCCTTGCTACTCATGGTTTAGCTGCTGTTAATAGTGTCCCTTGGGATACTATAGCTAATTATGCTAAGCCTTTTCTTAACAAAGTTGTTAGTACAACTACTAACATAGTTACACGGTGTTTAAACCGTGTTTGTACTAATTATATGCCTTATTTCTTTACTTTATTGCTACAATTGTGTACTTTTACTAGAAGTACAAATTCTAGAATTAAAGCATCTATGCCGACTACTATAGCAAAGAATACTGTTAAGAGTGTCGGTAAATTTTGTCTAGAGGCTTCATTTAATTATTTGAAGTCACCTAATTTTTCTAAACTGATAAATATTACAATTTGGTTTTTACTATTAAGTGTTTGCCTAGGTTCTTTAATCTACTCAACCGCTGCTTTAGGTGTTTTAATGTCTAATTTAGGCATGCCTTCTTACTGTACTGGTTACAGAGAAGGCTATTTGAACTCTACTAATGTCACTATTGCAACCTACTGTACTGGTTCTATACCTTGTAGTGTTTGTCTTAGTGGTTTAGATTCTTTAGACACCTATCCTTCTTTAGAAACTATACAAATTACCATTTCATCTTTTAAATGGGATTTAACTGCTTTTGGCTTAGTTGCAGAGTGGTTTTTGGCATATATTCTTTTCACTAGGTTTTTCTATGTACTTGGATTGGCTGCAATCATGCAATTGTTTTTCAGCTATTTTGCAGTACATTTTATTAGTAATTCTTGGCTTATGTGGTTAATAATTAATCTTGTACAAATGGCCCCGATTTCAGCTATGGTTAGAATGTACATCTTCTTTGCATCATTTTATTATGTATGGAAAAGTTATGTGCATGTTGTAGACGGTTGTAATTCATCAACTTGTATGATGTGTTACAAACGTAATAGAGCAACAAGAGTCGAATGTACAACTATTGTTAATGGTGTTAGAAGGTCCTTTTATGTCTATGCTAATGGAGGTAAAGGCTTTTGCAAACTACACAATTGGAATTGTGTTAATTGTGATACATTCTGTGCTGGTAGTACATTTATTAGTGATGAAGTTGCGAGAGACTTGTCACTACAGTTTAAAAGACCAATAAATCCTACTGACCAGTCTTCTTACATCGTTGATAGTGTTACAGTGAAGAATGGTTCCATCCATCTTTACTTTGATAAAGCTGGTCAAAAGACTTATGAAAGACATTCTCTCTCTCATTTTGTTAACTTAGACAACCTGAGAGCTAATAACACTAAAGGTTCATTGCCTATTAATGTTATAGTTTTTGATGGTAAATCAAAATGTGAAGAATCATCTGCAAAATCAGCGTCTGTTTACTACAGTCAGCTTATGTGTCAACCTATACTGTTACTAGATCAGGCATTAGTGTCTGATGTTGGTGATAGTGCGGAAGTTGCAGTTAAAATGTTTGATGCTTACGTTAATACGTTTTCATCAACTTTTAACGTACCAATGGAAAAACTCAAAACACTAGTTGCAACTGCAGAAGCTGAACTTGCAAAGAATGTGTCCTTAGACAATGTCTTATCTACTTTTATTTCAGCAGCTCGGCAAGGGTTTGTTGATTCAGATGTAGAAACTAAAGATGTTGTTGAATGTCTTAAATTGTCACATCAATCTGACATAGAAGTTACTGGCGATAGTTGTAATAACTATATGCTCACCTATAACAAAGTTGAAAACATGACACCCCGTGACCTTGGTGCTTGTATTGACTGTAGTGCGCGTCATATTAATGCGCAGGTAGCAAAAAGTCACAACATTGCTTTGATATGGAACGTTAAAGATTTCATGTCATTGTCTGAACAACTACGAAAACAAATACGTAGTGCTGCTAAAAAGAATAACTTACCTTTTAAGTTGACATGTGCAACTACTAGACAAGTTGTTAATGTTGTAACAACAAAGATAGCACTTAAGGGTGGTAAAATTGTTAATAATTGGTTGAAGCAGTTAATTAAAGTTACACTTGTGTTCCTTTTTGTTGCTGCTATTTTCTATTTAATAACACCTGTTCATGTCATGTCTAAACATACTGACTTTTCAAGTGAAATCATAGGATACAAGGCTATTGATGGTGGTGTCACTCGTGACATAGCATCTACAGATACTTGTTTTGCTAACAAACATGCTGATTTTGACACATGGTTTAGCCAGCGTGGTGGTAGTTATACTAATGACAAAGCTTGCCCATTGATTGCTGCAGTCATAACAAGAGAAGTGGGTTTTGTCGTGCCTGGTTTGCCTGGCACGATATTACGCACAACTAATGGTGACTTTTTGCATTTCTTACCTAGAGTTTTTAGTGCAGTTGGTAACATCTGTTACACACCATCAAAACTTATAGAGTACACTGACTTTGCAACATCAGCTTGTGTTTTGGCTGCTGAATGTACAATTTTTAAAGATGCTTCTGGTAAGCCAGTACCATATTGTTATGATACCAATGTACTAGAAGGTTCTGTTGCTTATGAAAGTTTACGCCCTGACACACGTTATGTGCTCATGGATGGCTCTATTATTCAATTTCCTAACACCTACCTTGAAGGTTCTGTTAGAGTGGTAACAACTTTTGATTCTGAGTACTGTAGGCACGGCACTTGTGAAAGATCAGAAGCTGGTGTTTGTGTATCTACTAGTGGTAGATGGGTACTTAACAATGATTATTACAGATCTTTACCAGGAGTTTTCTGTGGTGTAGATGCTGTAAATTTACTTACTAATATGTTTACACCACTAATTCAACCTATTGGTGCTTTGGACATATCAGCATCTATAGTAGCTGGTGGTATTGTAGCTATCGTAGTAACATGCCTTGCCTACTATTTTATGAGGTTTAGAAGAGCTTTTGGTGAATACAGTCATGTAGTTGCCTTTAATACTTTACTATTCCTTATGTCATTCACTGTACTCTGTTTAACACCAGTTTACTCATTCTTACCTGGTGTTTATTCTGTTATTTACTTGTACTTGACATTTTATCTTACTAATGATGTTTCTTTTTTAGCACATATTCAGTGGATGGTTATGTTCACACCTTTAGTACCTTTCTGGATAACAATTGCTTATATCATTTGTATTTCCACAAAGCATTTCTATTGGTTCTTTAGTAATTACCTAAAGAGACGTGTAGTCTTTAATGGTGTTTCCTTTAGTACTTTTGAAGAAGCTGCGCTGTGCACCTTTTTGTTAAATAAAGAAATGTATCTAAAGTTGCGTAGTGATGTGCTATTACCTCTTACGCAATATAATAGATACTTAGCTCTTTATAATAAGTACAAGTATTTTAGTGGAGCAATGGATACAACTAGCTACAGAGAAGCTGCTTGTTGTCATCTCGCAAAGGCTCTCAATGACTTCAGTAACTCAGGTTCTGATGTTCTTTACCAACCACCACAAACCTCTATCACCTCAGCTGTTTTGCAGAGTGGTTTTAGAAAAATGGCATTCCCATCTGGTAAAGTTGAGGGTTGTATGGTACAAGTAACTTGTGGTACAACTACACTTAACGGTCTTTGGCTTGATGACGTAGTTTACTGTCCAAGACATGTGATCTGCACCTCTGAAGACATGCTTAACCCTAATTATGAAGATTTACTCATTCGTAAGTCTAATCATAATTTCTTGGTACAGGCTGGTAATGTTCAACTCAGGGTTATTGGACATTCTATGCAAAATTGTGTACTTAAGCTTAAGGTTGATACAGCCAATCCTAAGACACCTAAGTATAAGTTTGTTCGCATTCAACCAGGACAGACTTTTTCAGTGTTAGCTTGTTACAATGGTTCACCATCTGGTGTTTACCAATGTGCTATGAGGCCCAATTTCACTATTAAGGGTTCATTCCTTAATGGTTCATGTGGTAGTGTTGGTTTTAACATAGATTATGACTGTGTCTCTTTTTGTTACATGCACCATATGGAATTACCAACTGGAGTTCATGCTGGCACAGACTTAGAAGGTAACTTTTATGGACCTTTTGTTGACAGGCAAACAGCACAAGCAGCTGGTACGGACACAACTATTACAGTTAATGTTTTAGCTTGGTTGTACGCTGCTGTTATAAATGGAGACAGGTGGTTTCTCAATCGATTTACCACAACTCTTAATGACTTTAACCTTGTGGCTATGAAGTACAATTATGAACCTCTAACACAAGACCATGTTGACATACTAGGACCTCTTTCTGCTCAAACTGGAATTGCCGTTTTAGATATGTGTGCTTCATTAAAAGAATTACTGCAAAATGGTATGAATGGACGTACCATATTGGGTAGTGCTTTATTAGAAGATGAATTTACACCTTTTGATGTTGTTAGACAATGCTCAGGTGTTACTTTCCAAAGTGCAGTGAAAAGAACAATCAAGGGTACACACCACTGGTTGTTACTCACAATTTTGACTTCACTTTTAGTTTTAGTCCAGAGTACTCAATGGTCTTTGTTCTTTTTTTTGTATGAAAATGCCTTTTTACCTTTTGCTATGGGTATTATTGCTATGTCTGCTTTTGCAATGATGTTTGTCAAACATAAGCATGCATTTCTCTGTTTGTTTTTGTTACCTTCTCTTGCCACTGTAGCTTATTTTAATATGGTCTATATGCCTGCTAGTTGGGTGATGCGTATTATGACATGGTTGGATATGGTTGATACTAGTTTGAAGCTAAAAGACTGTGTTATGTATGCATCAGCTGTAGTGTTACTAATCCTTATGACAGCAAGAACTGTGTATGATGATGGTGCTAGGAGAGTGTGGACACTTATGAATGTCTTGACACTCGTTTATAAAGTTTATTATGGTAATGCTTTAGATCAAGCCATTTCCATGTGGGCTCTTATAATCTCTGTTACTTCTAACTACTCAGGTGTAGTTACAACTGTCATGTTTTTGGCCAGAGGTATTGTTTTTATGTGTGTTGAGTATTGCCCTATTTTCTTCATAACTGGTAATACACTTCAGTGTATAATGCTAGTTTATTGTTTCTTAGGCTATTTTTGTACTTGTTACTTTGGCCTCTTTTGTTTACTCAACCGCTACTTTAGACTGACTCTTGGTGTTTATGATTACTTAGTTTCTACACAGGAGTTTAGATATATGAATTCACAGGGACTACTCCCACCCAAGAATAGCATAGATGCCTTCAAACTCAACATTAAATTGTTGGGTGTTGGTGGCAAACCTTGTATCAAAGTAGCCACTGTACAGTCTAAAATGTCAGATGTAAAGTGCACATCAGTAGTCTTACTCTCAGTTTTGCAACAACTCAGAGTAGAATCATCATCTAAATTGTGGGCTCAATGTGTCCAGTTACACAATGACATTCTCTTAGCTAAAGATACTACTGAAGCCTTTGAAAAAATGGTTTCACTACTTTCTGTTTTGCTTTCCATGCAGGGTGCTGTAGACATAAACAAGCTTTGTGAAGAAATGCTGGACAACAGGGCAACCTTACAAGCTATAGCCTCAGAGTTTAGTTCCCTTCCATCATATGCAGCTTTTGCTACTGCTCAAGAAGCTTATGAGCAGGCTGTTGCTAATGGTGATTCTGAAGTTGTTCTTAAAAAGTTGAAGAAGTCTTTGAATGTGGCTAAATCTGAATTTGACCGTGATGCAGCCATGCAACGTAAGTTGGAAAAGATGGCTGATCAAGCTATGACCCAAATGTATAAACAGGCTAGATCTGAGGACAAGAGGGCAAAAGTTACTAGTGCTATGCAGACAATGCTTTTCACTATGCTTAGAAAGTTGGATAATGATGCACTCAACAACATTATCAACAATGCAAGAGATGGTTGTGTTCCCTTGAACATAATACCTCTTACAACAGCAGCCAAACTAATGGTTGTCATACCAGACTATAACACATATAAAAATACGTGTGATGGTACAACATTTACTTATGCATCAGCATTGTGGGAAATCCAACAGGTTGTAGATGCAGATAGTAAAATTGTTCAACTTAGTGAAATTAGTATGGACAATTCACCTAATTTAGCATGGCCTCTTATTGTAACAGCTTTAAGGGCCAATTCTGCTGTCAAATTACAGAATAATGAGCTTAGTCCTGTTGCACTACGACAGATGTCTTGTGCTGCCGGTACTACACAAACTGCTTGCACTGATGACAATGCGTTAGCTTACTACAACACAACAAAGGGAGGTAGGTTTGTACTTGCACTGTTATCCGATTTACAGGATTTGAAATGGGCTAGATTCCCTAAGAGTGATGGAACTGGTACTATCTATACAGAACTGGAACCACCTTGTAGGTTTGTTACAGACACACCTAAAGGTCCTAAAGTGAAGTATTTATACTTTATTAAAGGATTAAACAACCTAAATAGAGGTATGGTACTTGGTAGTTTAGCTGCCACAGTACGTCTACAAGCTGGTAATGCAACAGAAGTGCCTGCCAATTCAACTGTATTATCTTTCTGTGCTTTTGCTGTAGATGCTGCTAAAGCTTACAAAGATTATCTAGCTAGTGGGGGACAACCAATCACTAATTGTGTTAAGATGTTGTGTACACACACTGGTACTGGTCAGGCAATAACAGTTACACCGGAAGCCAATATGGATCAAGAATCCTTTGGTGGTGCATCGTGTTGTCTGTACTGCCGTTGCCACATAGATCATCCAAATCCTAAAGGATTTTGTGACTTAAAAGGTAAGTATGTACAAATACCTACAACTTGTGCTAATGACCCTGTGGGTTTTACACTTAAAAACACAGTCTGTACCGTCTGCGGTATGTGGAAAGGTTATGGCTGTAGTTGTGATCAACTCCGCGAACCCATGCTTCAGTCAGCTGATGCACAATCGTTTTTAAACGGGTTTGCGGTGTAAGTGCAGCCCGTCTTACACCGTGCGGCACAGGCACTAGTACTGATGTCGTATACAGGGCTTTTGACATCTACAATGATAAAGTAGCTGGTTTTGCTAAATTCCTAAAAACTAATTGTTGTCGCTTCCAAGAAAAGGACGAAGATGACAATTTAATTGATTCTTACTTTGTAGTTAAGAGACACACTTTCTCTAACTACCAACATGAAGAAACAATTTATAATTTACTTAAGGATTGTCCAGCTGTTGCTAAACATGACTTCTTTAAGTTTAGAATAGACGGTGACATGGTACCACATATATCACGTCAACGTCTTACTAAATACACAATGGCAGACCTCGTCTATGCTTTAAGGCATTTTGATGAAGGTAATTGTGACACATTAAAAGAAATACTTGTCACATACAATTGTTGTGATGATGATTATTTCAATAAAAAGGACTGGTATGATTTTGTAGAAAACCCAGATATATTACGCGTATACGCCAACTTAGGTGAACGTGTACGCCAAGCTTTGTTAAAAACAGTACAATTCTGTGATGCCATGCGAAATGCTGGTATTGTTGGTGTACTGACATTAGATAATCAAGATCTCAATGGTAACTGGTATGATTTCGGTGATTTCATACAAACCACGCCAGGTAGTGGAGTTCCTGTTGTAGATTCTTATTATTCATTGTTAATGCCTATATTAACCTTGACCAGGGCTTTAACTGCAGAGTCACATGTTGACACTGACTTAACAAAGCCTTACATTAAGTGGGATTTGTTAAAATATGACTTCACGGAAGAGAGGTTAAAACTCTTTGACCGTTATTTTAAATATTGGGATCAGACATACCACCCAAATTGTGTTAACTGTTTGGATGACAGATGCATTCTGCATTGTGCAAACTTTAATGTTTTATTCTCTACAGTGTTCCCACTTACAAGTTTTGGACCACTAGTGAGAAAAATATTTGTTGATGGTGTTCCATTTGTAGTTTCAACTGGATACCACTTCAGAGAGCTAGGTGTTGTACATAATCAGGATGTAAACTTACATAGCTCTAGACTTAGTTTTAAGGAATTACTTGTGTATGCTGCTGACCCTGCTATGCACGCTGCTTCTGGTAATCTATTACTAGATAAACGCACTACGTGCTTTTCAGTAGCTGCACTTACTAACAATGTTGCTTTTCAAACTGTCAAACCTGGTAATTTTAACAAAGACTTCTATGACTTTGCTGTGTCTAAGGGTTTCTTTAAGGAAGGAAGTTCTGTTGAATTAAAACACTTCTTCTTTGCTCAGGATGGTAATGCTGCTATCAGCGATTATGACTACTATCGTTATAATCTACCAACAATGTGTGATATCAGACAACTACTATTTGTAGTTGAAGTTGTTGATAAGTACTTTGATTGTTACGATGGTGGCTGTATTAATGCTAACCAAGTCATCGTCAACAACCTAGACAAATCAGCTGGTTTTCCATTTAATAAATGGGGTAAGGCTAGACTTTATTATGATTCAATGAGTTATGAGGATCAAGATGCACTTTTCGCATATACAAAACGTAATGTCATCCCTACTATAACTCAAATGAATCTTAAGTATGCCATTAGTGCAAAGAATAGAGCTCGCACCGTAGCTGGTGTCTCTATCTGTAGTACTATGACCAATAGACAGTTTCATCAAAAATTATTGAAATCAATAGCCGCCACTAGAGGAGCTACTGTAGTAATTGGAACAAGCAAATTCTATGGTGGTTGGCACAACATGTTAAAAACTGTTTATAGTGATGTAGAAAACCCTCATCTTATGGGTTGGGATTATCCTAAATGTGATAGAGCCATGCCTAACATGCTTAGAATTATGGCCTCACTTGTTCTTGCTCGCAAACATACAACGTGTTGTAGCTTGTCACACCGTTTCTATAGATTAGCTAATGAGTGTGCTCAAGTATTGAGTGAAATGGTCATGTGTGGCGGTTCACTATATGTTAAACCAGGTGGAACCTCATCAGGAGATGCCACAACTGCTTATGCTAATAGTGTTTTTAACATTTGTCAAGCTGTCACGGCCAATGTTAATGCACTTTTATCTACTGATGGTAACAAAATTGCCGATAAGTATGTCCGCAATTTACAACACAGACTTTATGAGTGTCTCTATAGAAATAGAGATGTTGACACAGACTTTGTGAATGAGTTTTACGCATATTTGCGTAAACATTTCTCAATGATGATACTCTCTGACGATGCTGTTGTGTGTTTCAATAGCACTTATGCATCTCAAGGTCTAGTGGCTAGCATAAAGAACTTTAAGTCAGTTCTTTATTATCAAAACAATGTTTTTATGTCTGAAGCAAAATGTTGGACTGAGACTGACCTTACTAAAGGACCTCATGAATTTTGCTCTCAACATACAATGCTAGTTAAACAGGGTGATGATTATGTGTACCTTCCTTACCCAGATCCATCAAGAATCCTAGGGGCCGGCTGTTTTGTAGATGATATCGTAAAAACAGATGGTACACTTATGATTGAACGGTTCGTGTCTTTAGCTATAGATGCTTACCCACTTACTAAACATCCTAATCAGGAGTATGCTGATGTCTTTCATTTGTACTTACAATACATAAGAAAGCTACATGATGAGTTAACAGGACACATGTTAGACATGTATTCTGTTATGCTTACTAATGATAACACCTCAAGGTATTGGGAACCTGAGTTTTATGAGGCTATGTACACACCGCATACAGTCTTACAGGCTGTTGGGGCTTGTGTTCTTTGCAATTCACAGACTTCATTAAGATGTGGTGCTTGCATACGTAGACCATTCTTATGTTGTAAATGCTGTTACGACCATGTCATATCAACATCACATAAATTAGTCTTGTCTGTTAATCCGTATGTTTGCAATGCTCCAGGTTGTGATGTCACAGATGTGACTCAACTTTACTTAGGAGGTATGAGCTATTATTGTAAATCACATAAACCATCCATTAGTTTTCCATTGTGTGCTAATGGACAAGTTTTTGGTTTATATAAAAATACATGTGTTGGTAGCGATAATGTTACTGACTTTAATGCAATTGCAACATGTGACTGGACAAATGCTGGTGATTACATTTTAGCTAACACCTGTACTGAAAGACTCAAGCTTTTTGCAGCAGAAACGCTCAAAGCTACTGAGGAGACATTTAAACTGTCTTATGGTATTGCTACTGTACGTGAAGTGCTGTCTGACAGAGAATTACATCTTTCATGGGAAGTTGGTAAACCTAGACCACCACTTAACCGAAATTATGTCTTTACTGGTTATCGTGTAACTAAAAACAGTAAAGTACAAATAGGAGAGTACACCTTTGAAAAAGGTGACTATGGTGATGCTGTTGTTTACCGAGGTACAACAACTTACAAATTAAATGTTGGTGATTATTTTGTGCTGACATCACATACAGTAATGCCATTAAGTGCACCTACACTAGTGCCACAAGAGCACTATGTTAGAATTACTGGCTTATACCCAACACTCAATATCTCAGATGAGTTTTCTAGCAATGTTGCAAATTATCAAAAGGTTGGTATGCAAAAGTATTCTACACTCCAGGGACCACCTGGTACTGGTAAGAGTCATTTTGCTATTGGCCTAGCTCTCTACTACCCTTCTGCTCGCATAGTGTATACAGCTTGCTCTCATGCCGCTGTTGATGCACTATGTGAGAAGGCATTAAAATATTTGCCTATAGATAAATGTAGTAGAATTATACCTGCACGTGCTCGTGTAGAGTGTTTTGATAAATTCAAAGTGAATTCAACATTAGAACAGTATGTCTTTTGTACTGTAAATGCATTGCCTGAGACGACAGCAGATATAGTTGTCTTTGATGAAATTTCAATGGCCACAAATTATGATTTGAGTGTTGTCAATGCCAGATTACGTGCTAAGCACTATGTGTACATTGGCGACCCTGCTCAATTACCTGCACCACGCACATTGCTAACTAAGGGCACACTAGAACCAGAATATTTCAATTCAGTGTGTAGACTTATGAAAACTATAGGTCCAGACATGTTCCTCGGAACTTGTCGGCGTTGTCCTGCTGAAATTGTTGACACTGTGAGTGCTTTGGTTTATGATAATAAGCTTAAAGCACATAAAGACAAATCAGCTCAATGCTTTAAAATGTTTTATAAGGGTGTTATCACGCATGATGTTTCATCTGCAATTAACAGGCCACAAATAGGCGTGGTAAGAGAATTCCTTACACGTAACCCTGCTTGGAGAAAAGCTGTCTTTATTTCACCTTATAATTCACAGAATGCTGTAGCCTCAAAGATTTTGGGACTACCAACTCAAACTGTTGATTCATCACAGGGCTCAGAATATGACTATGTCATATTCACTCAAACCACTGAAACAGCTCACTCTTGTAATGTAAACAGATTTAATGTTGCTATTACCAGAGCAAAAGTAGGCATACTTTGCATAATGTCTGATAGAGACCTTTATGACAAGTTGCAATTTACAAGTCTTGAAATTCCACGTAGGAATGTGGCAACTTTACAAGCTGAAAATGTAACAGGACTCTTTAAAGATTGTAGTAAGGTAATCACTGGGTTACATCCTACACAGGCACCTACACACCTCAGTGTTGACACTAAATTCAAAACTGAAGGTTTATGTGTTGACATACCTGGCATACCTAAGGACATGACCTATAGAAGACTCATCTCTATGATGGGTTTTAAAATGAATTATCAAGTTAATGGTTACCCTAACATGTTTATCACCCGCGAAGAAGCTATAAGACATGTACGTGCATGGATTGGCTTCGATGTCGAGGGGTGTCATGCTACTAGAGAAGCTGTTGGTACCAATTTACCTTTACAGCTAGGTTTTTCTACAGGTGTTAACCTAGTTGCTGTACCTACAGGTTATGTTGATACACCTAATAATACAGATTTTTCCAGAGTTAGTGCTAAACCACCGCCTGGAGATCAATTTAAACACCTCATACCACTTATGTACAAAGGACTTCCTTGGAATGTAGTGCGTATAAAGATTGTACAAATGTTAAGTGACACACTTAAAAATCTCTCTGACAGAGTCGTATTTGTCTTATGGGCACATGGCTTTGAGTTGACATCTATGAAGTATTTTGTGAAAATAGGACCTGAGCGCACCTGTTGTCTATGTGATAGACGTGCCACATGCTTTTCCACTGCTTCAGACACTTATGCCTGTTGGCATCATTCTATTGGATTTGATTACGTCTATAATCCGTTTATGATTGATGTTCAACAATGGGGTTTTACAGGTAACCTACAAAGCAACCATGATCTGTATTGTCAAGTCCATGGTAATGCACATGTAGCTAGTTGTGATGCAATCATGACTAGGTGTCTAGCTGTCCACGAGTGCTTTGTTAAGCGTGTTGACTGGACTATTGAATATCCTATAATTGGTGATGAACTGAAGATTAATGCGGCTTGTAGAAAGGTTCAACACATGGTTGTTAAAGCTGCATTATTAGCAGACAAATTCCCAGTTCTTCACGACATTGGTAACCCTAAAGCTATTAAGTGTGTACCTCAAGCTGATGTAGAATGGAAGTTCTATGATGCACAGCCTTGTAGTGACAAAGCTTATAAAATAGAAGAATTATTCTATTCTTATGCCACACATTCTGACAAATTCACAGATGGTGTATGCCTATTTTGGAATTGCAATGTCGATAGATATCCTGCTAATTCCATTGTTTGTAGATTTGACACTAGAGTGCTATCTAACCTTAACTTGCCTGGTTGTGATGGTGGCAGTTTGTATGTAAATAAACATGCATTCCACACACCAGCTTTTGATAAAAGTGCTTTTGTTAATTTAAAACAATTACCATTTTTCTATTACTCTGACAGTCCATGTGAGTCTCATGGAAAACAAGTAGTGTCAGATATAGATTATGTACCACTAAAGTCTGCTACGTGTATAACACGTTGCAATTTAGGTGGTGCTGTCTGTAGACATCATGCTAATGAGTACAGATTGTATCTCGATGCTTATAACATGATGATCTCAGCTGGCTTTAGCTNNNNNNNNNACAAACAATTTGATACTTATAACCTCTGGAACACTTTTACAAGACTTCAGAGTTTAGAAAATGTGGCTTTTAATGTTGTAAATAAGGGACACTTTGATGGACAACAGGGTGAAGTACCAGTTTCTATCATTAATAACACTGTTTACACAAAAGTTGATGGTGTTGATGTAGAATTGTTTGAAAATAAAACAACATTACCTGTTAATGTAGCATTTGAGCTTTGGGCTAAGCGCAACATTAAACCAGTACCAGAGGTGAAAATACTCAATAATTTGGGTGTGGACATTGCTGCTAATACTGTGATCTGGGACTACAAAAGAGATGCTCCAGCACATATATCTACTATTGGTGTTTGTTCTATGACTGACATAGCCAAGAAACCAACTGAAACGATTTGTGCACCACTCACTGTCTTTTTTGATGGTAGAGTTGATGGTCAAGTAGACTTATTTAGAAATGCCCGTAATGGTGTTCTTATTACAGAAGGTAGTGTTAAAGGTTTACAACCATCTGTAGGTCCCAAACAAGCTAGTCTTAATGGAGTCACATTAATTGGAGAAGCCGTAAAAACACAGTTCAATTATTATAAGAAAGTTGATGGTGTTGTCCAACAATTACCTGAAACTTACTTTACTCAGAGTAGAAATTTACAAGAATTTAAACCCAGGAGTCAAATGGAAATTGATTTCTTAGAATTAGCTATGGATGAATTCATTGAACGGTATAAATTAGAAGGCTATGCCTTCGAACATATCGTTTATGGAGATTTTAGTCATAGTCAGTTAGGTGGTTTACATCTACTGATTGGACTAGCTAAACGTTTTAAGGAATCACCTTTTGAATTAGAAGATTTTATTCCTATGGACAGTACAGTTAAAAACTATTTCATAACAGATGCGCAAACAGGTTCATCTAAGTGTGTGTGTTCTGTTATTGATTTATTAATTGATGATTTTGTTGAAATAATAAAATCCCAAGATTTATCTGTAGTTTCTAAGGTTGTCAAAGTGACTATTGACTATACAGAAATTTCATTTATGCTTTGGTGTAAAGATGGCCATGTAGAAACATTTTACCCAAAATTACAATCTAGTCAAGCGTGGCAACCGGGTGTTGCTATGCCTAATCTTTACAAAATGCAAAGAATGCTATTAGAAAAGTGTGACCTTCAAAATTATGGTGATAGTGCAACATTACCTAAAGGCATAATGATGAATGTCGCAAAATATACTCAACTGTGTCAATATTTAAACACATTAACATTAGCTGTACCCTATAATATGAGAGTTATACATTTTGGTGCTGGTTCTGATAAAGGAGTTGCACCAGGTACAGCTGTTTTAAGACAGTGGTTGCCTACGGGTACGCTGCTTGTCGATTCAGATCTTAATGACTTTGTCTCTGATGCAGATTCAACTTTGATTGGTGATTGTGCAACTGTACATACAGCTAATAAATGGGATCTCATTATTAGTGATATGTACGACCCTAAGACTAAAAATGTTACAAAAGAAAATGACTCTAAAGAGGGTTTTTTCACTTACATTTGTGGGTTTATACAACAAAAGCTAGCTCTTGGAGGTTCCGTGGCTATAAAGATAACAGAACATTCTTGGAATGCTGATCTTTATAAGCTCATGGGACACTTCGCATGGTGGACAGCCTTTGTTACTAATGTGAATGCGTCATCATCTGAAGCATTTTTAATTGGATGTAATTATCTTGGCAAACCACGCGAACAAATAGATGGTTATGTCATGCATGCAAATTACATATTTTGGAGGAATACAAATCCAATTCAGTTGTCTTCCTATTCTTTATTTGACATGAGTAAATTTCCCCTTAAATTAAGGGGTACTGCTGTTATGTCTTTAAAAGAAGGTCAAATCAATGATATGATTTTATCTCTTCTTAGTAAAGGTAGACTTATAATTAGAGAAAACAACAGAGTTGTTATTTCTAGTGATGTTCTTGTTAACAACTAAACGAACAATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTGTTAATCTTACAACCAGAACTCAATTACCCCCTGCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAAGTTTTCAGATCCTCAGTTTTACATTCAACTCAGGACTTGTTCTTACCTTTCTTTTCCAATGTTACTTGGTTCCATGCTATCTCTGGGACCAATGGTACTAAGAGGTTTGATAACCCTGTCCTACCATTTAATGATGGTGTTTATTTTGCTTCCACTGAGAAGTCTAACATAATAAGAGGCTGGATTTTTGGTACTACTTTAGATTCGAAGACCCAGTCCCTACTTATTGTTAATAACGCTACTAATGTTGTTATTAAAGTCTGTGAATTTCAATTTTGTAATGATCCATTTTTGGGTGTTTACCACAAAAACAACAAAAGTTGGATGGAAAGTGAGTTCAGAGTTTATTCTAGTGCGAATAATTGCACTTTTGAATATGTCTCTCAGCCTTTTCTTATGGACCTTGAAGGAAAACAGGGTAATTTCAAAAATCTTAGGGAATTTGTGTTTAAGAATATTGATGGTTATTTTAAAATATATTCTAAGCACACGCCTATTAATTTAGTGCGTGATCTCCCTCAGGGTTTTTCGGCTTTAGAACCATTGGTAGATTTGCCAATAGGTATTAACATCACTAGGTTTCAAACTTTACTTGCTTTACATAGAAGTTATTTGACTCCTGGTGATTCTTCTTCAGGTTGGACAGCTGGTGCTGCAGCTTATTATGTGGGTTATCTTCAACCTAGGACTTTTCTATTAAAATATAATGAAAATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCTCAGAAACAAAGTGTACGTTGAAATCCTTCACTGTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACTTGTGCCCTTTTGGTGAAGTTTTTAACGCCACCAGATTTGCATCTGTTTATGCTTGGAACAGGAAGAGAATCAGCAACTGTGTTGCTGATTATTCTGTCCTATATAATTCCGCATCATTTTCCACTTTTAAGTGTTATGGAGTGTCTCCTACTAAATTAAATGATCTCTGCTTTACTAATGTCTATGCAGATTCATTTGTAATTAGAGGTGATGAAGTCAGACAAATCGCTCCAGGGCAAACTGGAAAGATTGCTGATTATAATTATAAATTACCAGATGATTTTACAGGCTGCGTTATAGCTTGGAATTCTAACAATCTTGATTCTAAGGTTGGTGGTAATTATAATTACCTGTATAGATTGTTTAGGAAGTCTAATCTCAAACCTTTTGAGAGAGATATTTCAACTGAAATCTATCAGGCCGGTAGCACACCTTGTAATGGTGTTGAAGGTTTTAATTGTTACTTTCCTTTACAATCATATGGTTTCCAACCCACTTATGGTGTTGGTTACCAACCATACAGAGTAGTAGTACTTTCTTTTGAACTTCTACATGCACCAGCAACTGTTTGTGGACCTAAAAAGTCTACTAATTTGGTTAAAAACAAATGTGTCAATTTCAACTTCAATGGTTTAACAGGCACAGGTGTTCTTACTGAGTCTAACAAAAAGTTTCTGCCTTTCCAACAATTTGGCAGAGACATTGATGACACTACTGATGCTGTCCGTGATCCACAGACACTTGAGATTCTTGACATTACACCATGTTCTTTTGGTGGTGTCAGTGTTATAACACCAGGAACAAATACTTCTAACCAGGTTGCTGTTCTTTATCAGGGTGTTAACTGCACAGAAGTCCCTGTTGCTATTCATGCAGATCAACTTACTCCTACTTGGCGTGTTTATTCTACAGGTTTTAATGTTTTTCAAACACGTGCAGGCTGTTTAATAGGGGCTGAACATGTCAACAACTCATATGAGTGTGACATACCCATTGGTGCAGGTATATGCGCTAGTTATCAGACTCAGACTAATTCTCATCGGCGGGCACGTAGTGTAGTTAGTCAATCCATCATTGCCTACACTATGTCACTTGGTGCAGAAAATTCAGTTGCTTACTCTAATAACTCTATTGCCATACCCATAAATTTTACTATTAGTGTTACCACAGAAATTCTACCAGTGTCTATGACCAAGACATCAGTAGATTGTACAATGTACATTTGTGGTGATTCAACTGAATGCAGCAATCTTTTGTTGCAATATGGCAGTTTTTGTACACAATTAAACCGTGCTTTAACTGGAATAGCTGTTGAACAAGACAAAAACACCCAAGAAGTTTTTGCACAAGTCAAACAAATTTACAAAACACCACCAATTAAAGATTTTGGTGGTTTTAATTTTTCACAAATATTACCAGATCCATCAAAACCAAGCAAGAGGTCATTTATTGAAGATTTACTTTTCAACAAAGTGACACTTGCAGATGCTGGCTTCATCAAACAATATGGTGATTGCCTTGGTGATATTGCTGCTAGAGACCTCATTTGTGCACAAAAGTTTAACGGCCTTACTGTTTTGCCACCTTTGCTCACAGATGAAATGATTGCTCAATACACTTCTGCACTGTTAGCGGGTACAATCACTTCTGGTTGGACCTTTGGTGCAGGTGCTGCATTACAAATACCATTTGCTATGCAAATGGCTTATAGGTTTAATGGTATTGGAGTTACACAGAATGTTCTCTATGAGAACCAAAAATTGATTGCCAACCAATTTAATAGTGCTATTGGCAAAATTCAAGACTCACTTTCTTCCACAGCAAGTGCACTTGGAAAACTTCAAGATGTGGTCAACCAAAATGCACAAGCTTTAAACACGCTTGTTAAACAACTTAGCTCCAATTTTGGTGCAATTTCAAGTGTTTTAAATGATATCCTTGCACGTCTTGACAAAGTTGAGGCTGAAGTGCAAATTGATAGGTTGATCACAGGCAGACTTCAAAGTTTGCAGACATATGTGACTCAACAATTAATTAGAGCTGCAGAAATCAGAGCTTCTGCTAATCTTGCTGCTACTAAAATGTCAGAGTGTGTACTTGGACAATCAAAAAGAGTTGATTTTTGTGGAAAGGGCTATCATCTTATGTCCTTCCCTCAGTCAGCACCTCATGGTGTAGTCTTCTTGCATGTGACTTATGTCCCTGCACAAGAAAAGAACTTCACAACTGCTCCTGCCATTTGTCATGATGGAAAAGCACACTTTCCTCGTGAAGGTGTCTTTGTTTCAAATGGCACACACTGGTTTGTAACACAAAGGAATTTTTATGAACCACAAATCATTACTACACACAACACATTTGTGTCTGGTAACTGTGATGTTGTAATAGGAATTGTCAACAACACAGTTTATGATCCTTTGCAACCTGAATTAGACTCATTCAAGGAGGAGTTAGATAAATATTTTAAGAATCATACATCACCAGATGTTGATTTAGGTGACATCTCTGGCATTAATGCTTCAGTTGTAAACATTCAAAAAGAAATTGACCGCCTCAATGAGGTTGCCAAGAATTTAAATGAATCTCTCATCGATCTCCAAGAACTTGGAAAGTATGAGCAGTATATAAAATGGCCATGGTACATTTGGCTAGGTTTTATAGCTGGCTTGATTGCCATAGTAATGGTGACAATTATGCTTTGCTGTATGACCAGTTGCTGTAGTTGTCTCAAGGGCTGTTGTTCTTGTGGATCCTGCTGCAAATTTGATGAAGACGACTCTGAGCCAGTGCTCAAAGGAGTCAAATTACATTACACATAAACGAACTTATGGATTTGTTTATGAGAATCTTCACAATTGGAACTGTAACTTTGAAGCAAGGTGAAATCAAGGATGCTACTCCTTCAGATTTTGTTCGCGCTACTGCAACGATACCGATACAAGCCTCACTCCCTTTCGGATGGCTTATTGTTGGCGTTGCACTTCTTGCTGTTTTTCAGAGCGCTTCCAAAATCATAACCCTCAAAAAGAGATGGCAACTAGCACTCTCCAAGGGTGTTCACTTTGTTTGCAACTTGCTGTTGTTGTTTGTAACAGTTTACTCACACCTTTTGCTCGTTGCTGCTGGCCTTGAAGCCCCTTTTCTCTATCTTTATGCTTTAGTCTACTTCTTGCAGAGTATAAACTTTGTAAGAATAATAATGAGGCTTTGGCTTTGCTGGAAATGCCGTTCCAAAAACCCATTACTTTATGATGCCAACTATTTTCTTTGCTGGCATACTAATTGTTACGACTATTGTATACCTTACAATAGTGTAACTTCTTCAATTGTCATTACTTCAGGTGATGGCACAACAAGTCCTATTTCTGAACATGACTACCAGATTGGTGGTTATACTGAAAAATGGGAATCTGGAGTAAAAGACTGTGTTGTATTACACAGTTACTTCACTTCAGACTATTACCAGCTGTACTCAACTCAATTGAGTACAGACACTGGTGTTGAACATGTTACCTTCTTCATCTACAATAAAATTGTTGATGAGCCTGAAGAACATGTCCAAATTCACACAATCGACGGTTCATCCGGAGTTGTTAATCCAGTAATGGAACCAATTTATGATGAACCGACGACGACTACTAGCGTGCCTTTGTAAGCACAAGCTGATGAGTACGAACTTATGTACTCATTCGTTTCGGAAGAGACAGGTACGTTAATAGTTAATAGCGTACTTCTTTTTCTTGCTTTCGTGGTATTCTTGCTAGTTACACTAGCCATCCTTACTGCGCTTCGATTGTGTGCGTACTGCTGCAATATTGTTAACGTGAGTCTTGTAAAACCTTCTTTTTACGTTTACTCTCGTGTTAAAAATCTGAATTCTTCTAGAGTTCCTGATCTTCTGGTCTAAACGAACTAAATATTATATTAGTTTTTCTGTTTGGAACTTTAATTTTAGCCATGGCAGATTCCAACGGTACTATTACCGTTGAAGAGCTTAAAAAGCTCCTTGAACAATGGAACCTAGTAATAGGTTTCCTATTCCTTACATGGATTTGTCTTCTACAATTTGCCTATGCCAACAGGAATAGGTTTTTGTATATAATTAAGTTAATTTTCCTCTGGCTGTTATGGCCAGTAACTTTAGCTTGTTTTGTGCTTGCTGCTGTTTACAGAATAAATTGGATCACCGGTGGAATTGCTATCGCAATGGCTTGTCTTGTAGGCTTGATGTGGCTCAGCTACTTCATTGCTTCTTTCAGACTGTTTGCGCGTACGCGTTCCATGTGGTCATTCAATCCAGAAACTAACATTCTTCTCAACGTGCCACTCCATGGCACTATTCTGACCAGACCGCTTCTAGAAAGTGAACTCGTAATCGGAGCTGTGATCCTTCGTGGACATCTTCGTATTGCTGGACACCATCTAGGACGCTGTGACATCAAGGACCTGCCTAAAGAAATCACTGTTGCTACATCACGAACGCTTTCTTATTACAAATTGGGAGCTTCGCAGCGTGTAGCAGGTGACTCAGGTTTTGCTGCATACAGTCGCTACAGGATTGGCAACTATAAATTAAACACAGACCATTCCAGTAGCAGTGACAATATTGCTTTGCTTGTACAGTAAGTGACAACAGATGTTTCATCTCGTTGACTTTCAGGTTACTATAGCAGAGATATTACTAATTATTATGAGGACTTTTAAAGTTTCCATTTGGAATCTTGATTACATCATAAACCTCATAATTAAAAATTTATCTAAGTCACTAACTGAGAATAAATATTCTCAATTAGATGAAGAGCAACCAATGGAGATTGATTAAACGAACATGAAAATTATTCTTTTCTTGGCACTGATAACACTCGCTACTTGTGAGCTTTATCACTACCAAGAGTGTGTTAGAGGTACAACAGTACTTTTAAAAGAACCTTGCTCTTCTGGAACATACGAGGGCAATTCACCATTTCATCCTCTAGCTGATAACAAATTTGCACTGACTTGCTTTAGCACTCAATTTGCTTTTGCTTGTCCTGACGGCGTAAAACACGTCTATCAGTTACGTGCCAGATCAGTTTCACCTAAACTGTTCATCAGACAAGAGGAAGTTCAAGAACTTTACTCTCCAATTTTTCTTATTGTTGCGGCAATAGTGTTTATAACACTTTGCTTCACACTCAAAAGAAAGACAGAATGATTGAACTTTCATTAATTGACTTCTATTTGTGCTTTTTAGCCTTTCTGCTATTCCTTGTTTTAATTATGCTTATTATCTTTTGGTTCTCACTTGAACTGCAAGATCATAATGAAACTTGTCACGCCTAAACGAACATGAAATTTCTTGTTTTCTTAGGAATCATCACAACTGTAGCTGCATTTCACCAAGAATGTAGTTTACAGTCATGTACTTAACATCAACCATATGTAGTTGATGACCCGTGTCCTATTCACTTCTATTCTAAATGGTATATTAGAGTAGGAGCTATAAAATCAGCACCTTTAATTGAATTGTGCGTGGATGAGGCTGGTTCTAAATCACCCATTCAGTGCATCGATATCGGTAATTATACAGTTTCCTGTTTACCTTTTACAATTAATTGCCAGGAACCTAAATTGGGTAGTCTTGTAGTGCGTTGTTCGTTCTATGAAGACTTTTTAGAGTATCATGACGTTCGTGTTGTTTTAGATTTCATCTAAACGAACAAACTAAATGTCTCTAAATGGACCCCAAAATCAGCGAAATGCACCCCGCATTACGTTTGGTGGACCCTCAGATTCAACTGGCAGTAACCAGAATGGAGAACGCAGTGGGGCGCGATCAAAACAACGTCGGCCCCAAGGTTTACCCAATAATACTGCGTCTTGGTTCACCGCTCTCACTCAACATGGCAAGGAAGACCTTAAATTCCCTCGAGGACAAGGCGTTCCAATTAACACCAATAGCAGTCCAGATGACCAAATTGGCTACTACCGAAGAGCTACCAGACGAATTCGTGGTGGTGACGGTAAAATGAAAGATCTCAGTCCAAGATGGTATTTCTACTACCTAGGAACTGGGCCAGAAGCTGGACTTCCCTATGGTGCTAACAAAGACGGCATCATATGGGTTGCAACTGAGGGAGCCTTGAATACACCAAAAGATCACATTGGCACCCGCAATCCTGCTAACAATGCTGCAATCGTGCTACAACTTCCTCAAGGAACAACATTGCCAAAAGGCTTCTACGCAGAAGGGAGCAGAGGCGGCAGTCAAGCCTCTTCTCGTTCCTCATCACGTAGTCGCAACAGTTCAAGAAATTCAACTCCAGGCAGCAGTAAACGAACTTCTCCTGCTAGAATGGCTGGCAATGGCGGTGATGCTGCTCTTGCTTTGCTGCTGCTTGACAGATTGAACCAGCTTGAGAGCAAAATGTTTGGTAAAGGCCAACAACAACAAGGCCAAACTGTCACTAAGAAATCTGCTGCTGAGGCTTCTAAGAAGCCTCGGCAAAAACGTACTGCCACTAAAGCATACAATGTAACACAAGCTTTCGGCAGACGTGGTCCAGAACAAACTCAAGGAAATTTTGGGGACCAGGAACTAATCAGACAAGGAACTGATTACAAACATTGGCCGCAAATTGCACAATTTGCCCCCAGCGCTTCAGCGTTCTTCGGAATGTCGCGCATTGGCATGGAAGTCACACCTTCGGGAACGTGGTTGACCTACACAGGTGCCATCAAATTGGATGACAAAGATCCAAATTTCAAAGATCAAGTCATTTTGCTGAATAAGCATATTGACGCATACAAAACATTCCCACCAACAGAGCCTAAAAAGGACAAAAAGAAGAAGGCTGATGAAACTCAAGCCTTACCGCAGAGACAGAAGAAACAGCAAACTGTGACTCTTCTTCCTGCTGCAGATTTGGATGATTTCTCCAAACAATTGCAACAATCCATGAGCAGTGCTGACTCAACTCAGGCCTAAACTCATGCAGACCACACAAGGCAGATGGGCTATATAAACGTTTTCGCTTTTCCGTTTACGATATATAGTCTACTCTTGTGCAGAATGAATTCTCGTAACTACATAGCACAAGTAGATGTAGTTAACTTTAATCTCACATAGCAATCTTTAATCAGTGTGTAACATTAGGGAGGACTTGAAAGAGCCACCACATTTTCACCGAGGCACGATCGAGTGTACAGTGAACAATGCTAGGGAGAGCTGCCTATATGGAAGAGCCCTAATGTGTAAAATTAATTTTAGTAGTGCTATCCNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\r\n"
     ]
    }
   ],
   "source": [
    "! head practice_fastas/NORW-220299_S80.primertrimmed.consensus.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36c7f8a6",
   "metadata": {},
   "source": [
    "Check how many As, Cs, Ts, Gs and Ns there are, and how long the sequence is:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "fe16ca97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file                                                         format  type     num_seqs  sum_len  min_len  avg_len  max_len\r\n",
      "practice_fastas/COG111-E-NC_S83.primertrimmed.consensus.fa   FASTA   Unlimit         1        0        0        0        0\r\n",
      "practice_fastas/COG111-NC_S89.primertrimmed.consensus.fa     FASTA   Unlimit         1        0        0        0        0\r\n",
      "practice_fastas/COG111-PC_S86.primertrimmed.consensus.fa     FASTA   DNA             1   29,903   29,903   29,903   29,903\r\n",
      "practice_fastas/COG112-PC_S114.primertrimmed.consensus.fa    FASTA   DNA             1   29,903   29,903   29,903   29,903\r\n",
      "practice_fastas/NORW-220299_S80.primertrimmed.consensus.fa   FASTA   DNA             1   29,874   29,874   29,874   29,874\r\n",
      "practice_fastas/NORW-22A50F_S96.primertrimmed.consensus.fa   FASTA   DNA             1   29,886   29,886   29,886   29,886\r\n",
      "practice_fastas/NORW-22AE40_S97.primertrimmed.consensus.fa   FASTA   DNA             1   29,874   29,874   29,874   29,874\r\n",
      "practice_fastas/NORW-22D87C_S81.primertrimmed.consensus.fa   FASTA   Unlimit         1        0        0        0        0\r\n",
      "practice_fastas/NORW-22DDC8_S95.primertrimmed.consensus.fa   FASTA   DNA             1   29,874   29,874   29,874   29,874\r\n",
      "practice_fastas/NORW-22DFFF_S94.primertrimmed.consensus.fa   FASTA   DNA             1   29,874   29,874   29,874   29,874\r\n",
      "practice_fastas/NORW-230B31_S87.primertrimmed.consensus.fa   FASTA   Unlimit         1        0        0        0        0\r\n",
      "practice_fastas/NORW-23264D_S112.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-23267A_S106.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-232689_S137.primertrimmed.consensus.fa  FASTA   Unlimit         1        0        0        0        0\r\n",
      "practice_fastas/NORW-2326A7_S131.primertrimmed.consensus.fa  FASTA   DNA             1   29,903   29,903   29,903   29,903\r\n",
      "practice_fastas/NORW-2326C5_S143.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-2326F2_S128.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-232829_S109.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-232847_S115.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-232883_S142.primertrimmed.consensus.fa  FASTA   DNA             1   29,883   29,883   29,883   29,883\r\n",
      "practice_fastas/NORW-232892_S118.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-2328DE_S133.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-232935_S135.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-232971_S125.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-232980_S124.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-23299F_S140.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n",
      "practice_fastas/NORW-232BD5_S138.primertrimmed.consensus.fa  FASTA   DNA             1   29,884   29,884   29,884   29,884\r\n"
     ]
    }
   ],
   "source": [
    "! seqkit stats practice_fastas/*.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88d0c0db",
   "metadata": {},
   "source": [
    "So, these are all about 29,900 bp long (right for SARS-CoV-2!), except that five files are empty except for the header, like the first one:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "c06f39da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Consensus_COG111-E-NC_S83.primertrimmed.consensus_threshold_0.75_quality_20\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! head -10 practice_fastas/COG111-E-NC_S83.primertrimmed.consensus.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad6f2ab",
   "metadata": {},
   "source": [
    "The empty files should not be included in the analysis; I will update `filter_fa_qc.sh` to make sure the files each contain at least one sequence.  Alternatively: just do the concatenation/filtering step as part of the bash script.  (Update: this has been done)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "952d9e01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "#testing out ways of how to do this\n",
    "\n",
    "#count lines of text in fasta with one sequence\n",
    "! cat practice_fastas/NORW-232BD5_S138.primertrimmed.consensus.fa | sed '/^\\s*$/d' | wc -l\n",
    "#count lines of text in empty fasta\n",
    "! cat practice_fastas/NORW-22D87C_S81.primertrimmed.consensus.fa | sed '/^\\s*$/d' | wc -l\n",
    "\n",
    "#code snippet credit: https://stackoverflow.com/questions/114814/count-non-blank-lines-of-code-in-bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db03b665",
   "metadata": {},
   "source": [
    "# Merging fastas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ad9d3a9",
   "metadata": {},
   "source": [
    "MAFFT takes as input a single fasta file with as many sequences as need to be aligned.  As each of my files contains at most 1 sequence, I'll have to append them all together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "9af9b503",
   "metadata": {},
   "outputs": [],
   "source": [
    "!  cat practice_fastas/*.fa > practice_fastas/merged-fasta.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3da8d572",
   "metadata": {},
   "source": [
    "Check that all 27 fastas made it in:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "9d73a4fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27\r\n"
     ]
    }
   ],
   "source": [
    "! grep -c \">\" practice_fastas/merged-fasta.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d40daf0",
   "metadata": {},
   "source": [
    "Remove the empty records and check how many sequences there are now:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "967a080b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22\n",
      ">Consensus_COG111-PC_S86.primertrimmed.consensus_threshold_0.75_quality_20\n"
     ]
    }
   ],
   "source": [
    "! awk 'BEGIN {RS = \">\" ; FS = \"\\n\" ; ORS = \"\"} $2 {print \">\"$0}' practice_fastas/merged-fasta.fa > practice_fastas/merged-fasta-nonempty.fa\n",
    "#code snippet credit: https://www.biostars.org/p/78786/\n",
    "\n",
    "! grep -c \">\" practice_fastas/merged-fasta-nonempty.fa\n",
    "! head -1 practice_fastas/merged-fasta-nonempty.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65a1c242",
   "metadata": {},
   "source": [
    "Trees and things will be easier to read if the record names are shortened to just \"COG111-PC_S86\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "ececcd25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">COG111-PC_S86\r\n",
      ">COG112-PC_S114\r\n",
      ">NORW-220299_S80\r\n",
      ">NORW-22A50F_S96\r\n",
      ">NORW-22AE40_S97\r\n",
      ">NORW-22DDC8_S95\r\n",
      ">NORW-22DFFF_S94\r\n",
      ">NORW-23264D_S112\r\n",
      ">NORW-23267A_S106\r\n",
      ">NORW-2326A7_S131\r\n",
      ">NORW-2326C5_S143\r\n",
      ">NORW-2326F2_S128\r\n",
      ">NORW-232829_S109\r\n",
      ">NORW-232847_S115\r\n",
      ">NORW-232883_S142\r\n",
      ">NORW-232892_S118\r\n",
      ">NORW-2328DE_S133\r\n",
      ">NORW-232935_S135\r\n",
      ">NORW-232971_S125\r\n",
      ">NORW-232980_S124\r\n",
      ">NORW-23299F_S140\r\n",
      ">NORW-232BD5_S138\r\n"
     ]
    }
   ],
   "source": [
    "! sed -i -e 's/Consensus_//g' practice_fastas/merged-fasta-nonempty.fa\n",
    "! sed -i -e 's/.primertrimmed.consensus_threshold_0.75_quality_20//g' practice_fastas/merged-fasta-nonempty.fa\n",
    "! grep '>' practice_fastas/merged-fasta-nonempty.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fe96d90",
   "metadata": {},
   "source": [
    "# Playing with MAFFT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80facabd",
   "metadata": {},
   "source": [
    "Now try running MAFFT on autopilot, preserving the input order of sequences:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "19281f6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 10552 ambiguous characters.\n",
      "    1 / 22\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      "   20 / 22\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP    21 / 21 \n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      "    0 / 22\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   20 / 22\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP    21 / 21 \n",
      "done.\n",
      "\n",
      "disttbfast (nuc) Version 7.490\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-2 (Fast but rough)\n",
      " Progressive method (guide trees were built 2 times.)\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! mafft --inputorder practice_fastas/merged-fasta-nonempty.fa > practice_fastas/auto.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a825a05a",
   "metadata": {},
   "source": [
    "The output of MAFFT is a fasta file with the sequences aligned to each other."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "1c410620",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">COG111-PC_S86\r\n",
      "nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnctttcgatctcttgtagatct\r\n",
      "gttctctaaacgaactnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngnnnnnn\r\n",
      "nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn\r\n",
      "nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn\r\n",
      "nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn\r\n",
      "nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngctcgtacgtggctttgg\r\n",
      "agactccgtggaggaggtcttatcagaggcacgtcaacatcttaaagatggcacttgtgg\r\n",
      "cttagtagaagttgaaaaaggcgttttgcctcaacttgaacagccctatgtgttcatcaa\r\n",
      "acgttcggatgctcgaactgcacctcatggtcatgttatggttgagctggtagcagaact\r\n"
     ]
    }
   ],
   "source": [
    "! head practice_fastas/auto.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0041692a",
   "metadata": {},
   "source": [
    "# Multiple alignment to a reference sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4f884d1",
   "metadata": {},
   "source": [
    "To align the sequences to the reference Wuhan-Wu-1, follow the syntax below (link: https://mafft.cbrc.jp/alignment/software/closelyrelatedviralgenomes.html).  The reference sequence for Wuhan-Wu-1 is available at: https://www.ncbi.nlm.nih.gov/nuccore/1798174254"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "74c65a69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mafft --6merpair --addfragments othersequences referencesequence > output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "3a903534",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nadd = 22\n",
      "ppenalty_ex = -10\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 1\n",
      "njobc = 2\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 10552 ambiguous characters\n",
      "    1 / 1\n",
      "done.\n",
      "\n",
      "nambiguous = 0 / 29903, ratio=0.000000\n",
      "fTEP 21 / 22                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (nuc) Version 7.490\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-fragment (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! mafft --6merpair --addfragments practice_fastas/merged-fasta-nonempty.fa Wuhan-Wu-1.fasta > output.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "6434bd81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NC_045512.2 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\r\n",
      "attaaaggtttataccttcccaggtaacaaaccaaccaactttcgatctcttgtagatct\r\n",
      "gttctctaaacgaactttaaaatctgtgtggctgtcactcggctgcatgcttagtgcact\r\n",
      "cacgcagtataattaataactaattactgtcgttgacaggacacgagtaactcgtctatc\r\n",
      "ttctgcaggctgcttacggtttcgtccgtgttgcagccgatcatcagcacatctaggttt\r\n",
      "cgtccgggtgtgaccgaaaggtaagatggagagccttgtccctggtttcaacgagaaaac\r\n",
      "acacgtccaactcagtttgcctgttttacaggttcgcgacgtgctcgtacgtggctttgg\r\n",
      "agactccgtggaggaggtcttatcagaggcacgtcaacatcttaaagatggcacttgtgg\r\n",
      "cttagtagaagttgaaaaaggcgttttgcctcaacttgaacagccctatgtgttcatcaa\r\n",
      "acgttcggatgctcgaactgcacctcatggtcatgttatggttgagctggtagcagaact\r\n"
     ]
    }
   ],
   "source": [
    "! head output.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a48cdf51",
   "metadata": {},
   "source": [
    "# Maximum likelihood tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0124a1df",
   "metadata": {},
   "source": [
    "Now that you have the MSA, you can create a tree using IQTREE.  IQTREE exists as a [web server ](http://iqtree.cibiv.univie.ac.at/), and there is also a command line option.  I'm starting with the [Beginner's Tutorial](http://www.iqtree.org/doc/Tutorial).  Using the default parameters:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "48fd1788",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 1.6.12 for Linux 64-bit built Mar 23 2020\n",
      "Developed by Bui Quang Minh, Nguyen Lam Tung, Olga Chernomor,\n",
      "Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    tamarisk (AVX2, FMA3, 9 GB RAM)\n",
      "Command: iqtree -s output.fa -redo\n",
      "Seed:    169559 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Fri Feb  4 12:11:10 2022\n",
      "Kernel:  AVX+FMA - 1 threads (8 CPU cores detected)\n",
      "\n",
      "HINT: Use -nt option to specify number of threads because your CPU has 8 cores!\n",
      "HINT: -nt AUTO will automatically determine the best number of threads to use.\n",
      "\n",
      "Reading alignment file output.fa ... Fasta format detected\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "WARNING: 2 sites contain only gaps or ambiguous characters.\n",
      "Alignment has 23 sequences with 29905 columns, 244 distinct patterns\n",
      "50 parsimony-informative, 57 singleton sites, 29798 constant sites\n",
      "                  Gap/Ambiguity  Composition  p-value\n",
      "   1  NC_045512.2         0.01%    passed     99.37%\n",
      "   2  COG111-PC_S86       2.38%    passed    100.00%\n",
      "   3  COG112-PC_S114      0.36%    passed     99.58%\n",
      "   4  NORW-220299_S80     1.06%    passed     99.67%\n",
      "   5  NORW-22A50F_S96     7.62%    passed     91.37%\n",
      "   6  NORW-22AE40_S97     0.51%    passed     99.98%\n",
      "   7  NORW-22DDC8_S95     1.04%    passed     99.90%\n",
      "   8  NORW-22DFFF_S94     0.67%    passed     99.96%\n",
      "   9  NORW-23264D_S112    0.42%    passed     99.84%\n",
      "  10  NORW-23267A_S106    0.42%    passed     99.84%\n",
      "  11  NORW-2326A7_S131    1.98%    passed     99.98%\n",
      "  12  NORW-2326C5_S143    0.45%    passed     99.80%\n",
      "  13  NORW-2326F2_S128    2.22%    passed     99.93%\n",
      "  14  NORW-232829_S109    1.50%    passed     99.88%\n",
      "  15  NORW-232847_S115    2.20%    passed     99.98%\n",
      "  16  NORW-232883_S142    0.43%    passed     99.78%\n",
      "  17  NORW-232892_S118    0.75%    passed     99.98%\n",
      "  18  NORW-2328DE_S133    0.43%    passed     99.71%\n",
      "  19  NORW-232935_S135    6.34%    passed     94.45%\n",
      "  20  NORW-232971_S125    0.78%    passed     99.98%\n",
      "  21  NORW-232980_S124    0.42%    passed     99.80%\n",
      "  22  NORW-23299F_S140    2.17%    passed     99.87%\n",
      "  23  NORW-232BD5_S138    2.60%    passed     99.82%\n",
      "****  TOTAL               1.60%  0 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "\n",
      "NOTE: Restoring information from model checkpoint file output.fa.model.gz\n",
      "\n",
      "CHECKPOINT: Initial tree restored\n",
      "NOTE: ModelFinder requires 1 MB RAM!\n",
      "ModelFinder will test 286 DNA models (sample size: 29905) ...\n",
      " No. Model         -LnL         df  AIC          AICc         BIC\n",
      "  1  JC            42560.521    43  85207.041    85207.168    85564.190\n",
      "  2  JC+I          42560.521    44  85209.041    85209.174    85574.496\n",
      "  3  JC+G4         42560.521    44  85209.042    85209.174    85574.496\n",
      "  4  JC+I+G4       42560.689    45  85211.379    85211.517    85585.139\n",
      "  5  JC+R2         42560.588    45  85211.176    85211.314    85584.936\n",
      "  6  JC+R3         42560.563    47  85215.127    85215.278    85605.498\n",
      " 14  F81+F         41653.755    46  83399.511    83399.656    83781.577\n",
      " 15  F81+F+I       41653.755    47  83401.511    83401.662    83791.882\n",
      " 16  F81+F+G4      41653.756    47  83401.511    83401.662    83791.883\n",
      " 17  F81+F+I+G4    41653.942    48  83403.884    83404.041    83802.561\n",
      " 18  F81+F+R2      41653.831    48  83403.661    83403.819    83802.339\n",
      " 19  F81+F+R3      41653.818    50  83407.636    83407.807    83822.925\n",
      " 27  K2P           42531.888    44  85151.775    85151.908    85517.229\n",
      " 28  K2P+I         42531.887    45  85153.775    85153.914    85527.535\n",
      " 29  K2P+G4        42531.888    45  85153.775    85153.914    85527.535\n",
      " 30  K2P+I+G4      42532.054    46  85156.108    85156.252    85538.174\n",
      " 31  K2P+R2        42531.953    46  85155.906    85156.051    85537.972\n",
      " 32  K2P+R3        42531.941    48  85159.881    85160.039    85558.559\n",
      " 40  HKY+F         41622.756    47  83339.511    83339.662    83729.883\n",
      " 41  HKY+F+I       41622.755    48  83341.510    83341.668    83740.188\n",
      " 42  HKY+F+G4      41622.755    48  83341.511    83341.668    83740.188\n",
      " 43  HKY+F+I+G4    41622.958    49  83343.916    83344.080    83750.899\n",
      " 44  HKY+F+R2      41622.838    49  83343.675    83343.839    83750.658\n",
      " 45  HKY+F+R3      41622.826    51  83347.653    83347.831    83771.248\n",
      " 53  TNe           42515.411    45  85120.822    85120.961    85494.582\n",
      " 54  TNe+I         42515.402    46  85122.804    85122.949    85504.870\n",
      " 55  TNe+G4        42515.396    46  85122.793    85122.937    85504.859\n",
      " 56  TNe+I+G4      42515.568    47  85125.135    85125.286    85515.507\n",
      " 57  TNe+R2        42515.458    47  85124.916    85125.067    85515.287\n",
      " 58  TNe+R3        42515.444    49  85128.889    85129.053    85535.872\n",
      " 66  TN+F          41605.913    48  83307.826    83307.984    83706.504\n",
      " 67  TN+F+I        41605.901    49  83309.803    83309.967    83716.786\n",
      " 68  TN+F+G4       41605.899    49  83309.797    83309.961    83716.781\n",
      " 69  TN+F+I+G4     41606.121    50  83312.243    83312.414    83727.532\n",
      " 70  TN+F+R2       41605.988    50  83311.975    83312.146    83727.264\n",
      " 71  TN+F+R3       41605.972    52  83315.944    83316.129    83747.845\n",
      " 79  K3P           42529.933    45  85149.865    85150.004    85523.625\n",
      " 80  K3P+I         42529.933    46  85151.865    85152.010    85533.931\n",
      " 81  K3P+G4        42529.933    46  85151.865    85152.010    85533.931\n",
      " 82  K3P+I+G4      42530.097    47  85154.194    85154.345    85544.565\n",
      " 83  K3P+R2        42529.995    47  85153.991    85154.142    85544.362\n",
      " 84  K3P+R3        42529.985    49  85157.969    85158.133    85564.952\n",
      " 92  K3Pu+F        41620.131    48  83336.262    83336.419    83734.939\n",
      " 93  K3Pu+F+I      41620.125    49  83338.250    83338.414    83745.233\n",
      " 94  K3Pu+F+G4     41620.125    49  83338.250    83338.414    83745.233\n",
      " 95  K3Pu+F+I+G4   41620.329    50  83340.659    83340.830    83755.948\n",
      " 96  K3Pu+F+R2     41620.206    50  83340.412    83340.583    83755.701\n",
      " 97  K3Pu+F+R3     41620.191    52  83344.383    83344.568    83776.284\n",
      "105  TPM2+F        41617.898    48  83331.796    83331.954    83730.474\n",
      "106  TPM2+F+I      41617.886    49  83333.772    83333.936    83740.755\n",
      "107  TPM2+F+G4     41617.874    49  83333.748    83333.912    83740.732\n",
      "108  TPM2+F+I+G4   41618.068    50  83336.135    83336.306    83751.424\n",
      "109  TPM2+F+R2     41617.931    50  83335.863    83336.033    83751.152\n",
      "110  TPM2+F+R3     41617.909    52  83339.818    83340.002    83771.718\n",
      "118  TPM2u+F       41617.815    48  83331.630    83331.787    83730.307\n",
      "119  TPM2u+F+I     41617.806    49  83333.612    83333.776    83740.595\n",
      "120  TPM2u+F+G4    41617.787    49  83333.574    83333.738    83740.557\n",
      "121  TPM2u+F+I+G4  41617.987    50  83335.975    83336.145    83751.264\n",
      "122  TPM2u+F+R2    41617.856    50  83335.713    83335.883    83751.002\n",
      "123  TPM2u+F+R3    41617.837    52  83339.674    83339.858    83771.574\n",
      "131  TPM3+F        41622.374    48  83340.748    83340.906    83739.426\n",
      "132  TPM3+F+I      41622.370    49  83342.740    83342.904    83749.723\n",
      "133  TPM3+F+G4     41622.367    49  83342.734    83342.899    83749.718\n",
      "134  TPM3+F+I+G4   41622.562    50  83345.124    83345.295    83760.413\n",
      "135  TPM3+F+R2     41622.437    50  83344.875    83345.046    83760.164\n",
      "136  TPM3+F+R3     41622.424    52  83348.848    83349.032    83780.748\n",
      "144  TPM3u+F       41622.344    48  83340.689    83340.846    83739.366\n",
      "145  TPM3u+F+I     41622.343    49  83342.687    83342.851    83749.670\n",
      "146  TPM3u+F+G4    41622.343    49  83342.686    83342.850    83749.669\n",
      "147  TPM3u+F+I+G4  41622.539    50  83345.078    83345.249    83760.368\n",
      "148  TPM3u+F+R2    41622.417    50  83344.834    83345.005    83760.123\n",
      "149  TPM3u+F+R3    41622.402    52  83348.804    83348.988    83780.704\n",
      "157  TIMe          42513.575    46  85119.149    85119.294    85501.215\n",
      "158  TIMe+I        42513.558    47  85121.115    85121.267    85511.487\n",
      "159  TIMe+G4       42513.545    47  85121.090    85121.241    85511.462\n",
      "160  TIMe+I+G4     42513.709    48  85123.418    85123.576    85522.096\n",
      "161  TIMe+R2       42513.587    48  85123.174    85123.331    85521.851\n",
      "162  TIMe+R3       42513.567    50  85127.133    85127.304    85542.422\n",
      "170  TIM+F         41603.363    49  83304.726    83304.890    83711.709\n",
      "171  TIM+F+I       41603.354    50  83306.709    83306.879    83721.998\n",
      "172  TIM+F+G4      41603.348    50  83306.695    83306.866    83721.984\n",
      "173  TIM+F+I+G4    41603.569    51  83309.137    83309.315    83732.732\n",
      "174  TIM+F+R2      41603.420    51  83308.839    83309.017    83732.434\n",
      "175  TIM+F+R3      41603.398    53  83312.796    83312.988    83753.003\n",
      "183  TIM2e         42512.744    46  85117.488    85117.633    85499.554\n",
      "184  TIM2e+I       42512.737    47  85119.474    85119.625    85509.846\n",
      "185  TIM2e+G4      42512.672    47  85119.344    85119.496    85509.716\n",
      "186  TIM2e+I+G4    42512.304    48  85120.608    85120.765    85519.285\n",
      "187  TIM2e+R2      42512.157    48  85120.313    85120.471    85518.991\n",
      "188  TIM2e+R3      42512.127    50  85124.253    85124.424    85539.542\n",
      "196  TIM2+F        41600.946    49  83299.892    83300.056    83706.875\n",
      "197  TIM2+F+I      41600.927    50  83301.853    83302.024    83717.142\n",
      "198  TIM2+F+G4     41600.909    50  83301.818    83301.989    83717.107\n",
      "199  TIM2+F+I+G4   41601.137    51  83304.274    83304.452    83727.869\n",
      "200  TIM2+F+R2     41600.985    51  83303.970    83304.148    83727.565\n",
      "201  TIM2+F+R3     41600.966    53  83307.933    83308.125    83748.139\n",
      "209  TIM3e         42514.528    46  85121.056    85121.201    85503.122\n",
      "210  TIM3e+I       42514.485    47  85122.969    85123.120    85513.341\n",
      "211  TIM3e+G4      42514.466    47  85122.932    85123.083    85513.304\n",
      "212  TIM3e+I+G4    42514.604    48  85125.208    85125.365    85523.885\n",
      "213  TIM3e+R2      42514.482    48  85124.963    85125.121    85523.641\n",
      "214  TIM3e+R3      42514.445    50  85128.890    85129.061    85544.179\n",
      "222  TIM3+F        41605.432    49  83308.863    83309.028    83715.847\n",
      "223  TIM3+F+I      41605.431    50  83310.861    83311.032    83726.150\n",
      "224  TIM3+F+G4     41605.430    50  83310.861    83311.032    83726.150\n",
      "225  TIM3+F+I+G4   41605.649    51  83313.299    83313.476    83736.893\n",
      "226  TIM3+F+R2     41605.508    51  83313.017    83313.195    83736.612\n",
      "227  TIM3+F+R3     41605.493    53  83316.986    83317.178    83757.193\n",
      "235  TVMe          42526.610    47  85147.220    85147.371    85537.591\n",
      "236  TVMe+I        42526.602    48  85149.203    85149.361    85547.881\n",
      "237  TVMe+G4       42526.596    48  85149.192    85149.349    85547.869\n",
      "238  TVMe+I+G4     42526.749    49  85151.497    85151.661    85558.480\n",
      "239  TVMe+R2       42526.639    49  85151.278    85151.443    85558.262\n",
      "240  TVMe+R3       42526.626    51  85155.251    85155.429    85578.846\n",
      "248  TVM+F         41616.483    50  83332.966    83333.136    83748.255\n",
      "249  TVM+F+I       41616.444    51  83334.888    83335.065    83758.483\n",
      "250  TVM+F+G4      41616.441    51  83334.882    83335.059    83758.476\n",
      "251  TVM+F+I+G4    41616.637    52  83337.274    83337.459    83769.175\n",
      "252  TVM+F+R2      41616.507    52  83337.014    83337.199    83768.915\n",
      "253  TVM+F+R3      41616.495    54  83340.990    83341.189    83789.502\n",
      "261  SYM           42510.102    48  85116.205    85116.362    85514.882\n",
      "262  SYM+I         42510.101    49  85118.202    85118.366    85525.185\n",
      "263  SYM+G4        42510.100    49  85118.200    85118.364    85525.184\n",
      "264  SYM+I+G4      42510.260    50  85120.519    85120.690    85535.808\n",
      "265  SYM+R2        42510.149    50  85120.298    85120.469    85535.587\n",
      "266  SYM+R3        42510.136    52  85124.272    85124.457    85556.172\n",
      "274  GTR+F         41599.665    51  83301.330    83301.508    83724.925\n",
      "275  GTR+F+I       41599.651    52  83303.303    83303.487    83735.203\n",
      "276  GTR+F+G4      41599.643    52  83303.287    83303.471    83735.187\n",
      "277  GTR+F+I+G4    41599.860    53  83305.720    83305.912    83745.927\n",
      "278  GTR+F+R2      41599.711    53  83305.422    83305.613    83745.628\n",
      "279  GTR+F+R3      41599.696    55  83309.392    83309.598    83766.210\n",
      "Akaike Information Criterion:           TIM2+F\n",
      "Corrected Akaike Information Criterion: TIM2+F\n",
      "Bayesian Information Criterion:         TN+F\n",
      "Best-fit model: TN+F chosen according to BIC\n",
      "\n",
      "All model information printed to output.fa.model.gz\n",
      "CPU time for ModelFinder: 0.018 seconds (0h:0m:0s)\n",
      "Wall-clock time for ModelFinder: 0.022 seconds (0h:0m:0s)\n",
      "\n",
      "NOTE: 0 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -41653.666\n",
      "2. Current log-likelihood: -41616.152\n",
      "3. Current log-likelihood: -41606.061\n",
      "Optimal log-likelihood: -41605.984\n",
      "Rate parameters:  A-C: 1.00000  A-G: 1.78033  A-T: 1.00000  C-G: 1.00000  C-T: 8.61790  G-T: 1.00000\n",
      "Base frequencies:  A: 0.299  C: 0.183  G: 0.196  T: 0.321\n",
      "Parameters optimization took 3 rounds (0.006 sec)\n",
      "Computing ML distances based on estimated model parameters... 0.001 sec\n",
      "Computing BIONJ tree...\n",
      "0.000 seconds\n",
      "Log-likelihood of BIONJ tree: -41605.981\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating 98 parsimony trees... 0.062 second\n",
      "Computing log-likelihood of 98 initial trees ... 0.111 seconds\n",
      "Current best score: -41605.981\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -41605.922\n",
      "BETTER TREE FOUND at iteration 2: -41605.913\n",
      "Iteration 10 / LogL: -41606.024 / Time: 0h:0m:0s\n",
      "Iteration 20 / LogL: -41606.049 / Time: 0h:0m:0s\n",
      "Finish initializing candidate tree set (20)\n",
      "Current best tree score: -41605.913 / CPU time: 0.267\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -41605.876\n",
      "Iteration 30 / LogL: -41605.987 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 40 / LogL: -41605.960 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 43: -41605.839\n",
      "UPDATE BEST LOG-LIKELIHOOD: -41605.837\n",
      "Iteration 50 / LogL: -41605.873 / Time: 0h:0m:0s (0h:0m:1s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 52: -41605.820\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 54: -41605.799\n",
      "Iteration 60 / LogL: -41605.957 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 70 / LogL: -41605.902 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 80 / LogL: -41605.833 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 90 / LogL: -41605.826 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 100 / LogL: -41612.354 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 110 / LogL: -41605.951 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -41605.798\n",
      "Iteration 120 / LogL: -41608.369 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 130 / LogL: -41605.799 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 140 / LogL: -41605.825 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 150 / LogL: -41605.829 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "TREE SEARCH COMPLETED AFTER 155 ITERATIONS / Time: 0h:0m:1s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -41605.798\n",
      "Optimal log-likelihood: -41605.796\n",
      "Rate parameters:  A-C: 1.00000  A-G: 1.67317  A-T: 1.00000  C-G: 1.00000  C-T: 7.99672  G-T: 1.00000\n",
      "Base frequencies:  A: 0.299  C: 0.183  G: 0.196  T: 0.321\n",
      "Parameters optimization took 1 rounds (0.001 sec)\n",
      "BEST SCORE FOUND : -41605.796\n",
      "Total tree length: 0.004\n",
      "\n",
      "Total number of iterations: 155\n",
      "CPU time used for tree search: 1.308 sec (0h:0m:1s)\n",
      "Wall-clock time used for tree search: 1.362 sec (0h:0m:1s)\n",
      "Total CPU time used: 1.324 sec (0h:0m:1s)\n",
      "Total wall-clock time used: 1.378 sec (0h:0m:1s)\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                output.fa.iqtree\n",
      "  Maximum-likelihood tree:       output.fa.treefile\n",
      "  Likelihood distances:          output.fa.mldist\n",
      "  Screen log file:               output.fa.log\n",
      "\n",
      "Date and Time: Fri Feb  4 12:11:11 2022\n"
     ]
    }
   ],
   "source": [
    "! iqtree -s output.fa -redo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7f9c0f6",
   "metadata": {},
   "source": [
    "The maximum likelihood tree can be found as a drawing and in Newick format at the end of the .iqtree output file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "df7ea619",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE 1.6.12 built Mar 23 2020\r\n",
      "\r\n",
      "Input file name: output.fa\r\n",
      "Type of analysis: ModelFinder + tree reconstruction\r\n",
      "Random seed number: 169559\r\n",
      "\r\n",
      "REFERENCES\r\n",
      "----------\r\n",
      "\r\n",
      "To cite ModelFinder please use: \r\n",
      "\r\n",
      "Subha Kalyaanamoorthy, Bui Quang Minh, Thomas KF Wong, Arndt von Haeseler,\r\n",
      "and Lars S Jermiin (2017) ModelFinder: Fast model selection for\r\n",
      "accurate phylogenetic estimates. Nature Methods, 14:587–589.\r\n",
      "https://doi.org/10.1038/nmeth.4285\r\n",
      "\r\n",
      "To cite IQ-TREE please use:\r\n",
      "\r\n",
      "Lam-Tung Nguyen, Heiko A. Schmidt, Arndt von Haeseler, and Bui Quang Minh\r\n",
      "(2015) IQ-TREE: A fast and effective stochastic algorithm for estimating\r\n",
      "maximum likelihood phylogenies. Mol Biol Evol, 32:268-274.\r\n",
      "https://doi.org/10.1093/molbev/msu300\r\n",
      "\r\n",
      "SEQUENCE ALIGNMENT\r\n",
      "------------------\r\n",
      "\r\n",
      "Input data: 23 sequences with 29905 nucleotide sites\r\n",
      "Number of constant sites: 29798 (= 99.6422% of all sites)\r\n",
      "Number of invariant (constant or ambiguous constant) sites: 29798 (= 99.6422% of all sites)\r\n",
      "Number of parsimony informative sites: 50\r\n",
      "Number of distinct site patterns: 244\r\n",
      "\r\n",
      "ModelFinder\r\n",
      "-----------\r\n",
      "\r\n",
      "Best-fit model according to BIC: TN+F\r\n",
      "\r\n",
      "List of models sorted by BIC scores: \r\n",
      "\r\n",
      "Model             LogL          AIC      w-AIC      AICc     w-AICc       BIC      w-BIC\r\n",
      "TN+F            -41605.9132  83307.8264 - 0.0059  83307.9839 - 0.0059  83706.5039 + 0.5163\r\n",
      "TIM2+F          -41600.9461  83299.8922 + 0.3093  83300.0563 + 0.3106  83706.8754 + 0.4288\r\n",
      "TIM+F           -41603.3630  83304.7260 - 0.0276  83304.8901 - 0.0277  83711.7092 - 0.0382\r\n",
      "TIM3+F          -41605.4317  83308.8635 - 0.0035  83309.0276 - 0.0035  83715.8467 - 0.0048\r\n",
      "TN+F+G4         -41605.8987  83309.7973 - 0.0022  83309.9615 - 0.0022  83716.7806 - 0.0030\r\n",
      "TN+F+I          -41605.9013  83309.8025 - 0.0022  83309.9667 - 0.0022  83716.7858 - 0.0030\r\n",
      "TIM2+F+G4       -41600.9092  83301.8184 + 0.1180  83301.9892 + 0.1182  83717.1074 - 0.0026\r\n",
      "TIM2+F+I        -41600.9267  83301.8534 + 0.1160  83302.0242 + 0.1161  83717.1424 - 0.0025\r\n",
      "TIM+F+G4        -41603.3475  83306.6951 - 0.0103  83306.8659 - 0.0103  83721.9841 - 0.0002\r\n",
      "TIM+F+I         -41603.3543  83306.7086 - 0.0102  83306.8795 - 0.0102  83721.9977 - 0.0002\r\n",
      "GTR+F           -41599.6650  83301.3299 + 0.1507  83301.5076 + 0.1503  83724.9248 - 0.0001\r\n",
      "TIM3+F+G4       -41605.4303  83310.8607 - 0.0013  83311.0315 - 0.0013  83726.1497 - 0.0000\r\n",
      "TIM3+F+I        -41605.4306  83310.8612 - 0.0013  83311.0321 - 0.0013  83726.1503 - 0.0000\r\n",
      "TN+F+R2         -41605.9877  83311.9754 - 0.0007  83312.1462 - 0.0007  83727.2644 - 0.0000\r\n",
      "TN+F+I+G4       -41606.1215  83312.2430 - 0.0006  83312.4138 - 0.0006  83727.5320 - 0.0000\r\n",
      "TIM2+F+R2       -41600.9852  83303.9704 - 0.0402  83304.1481 - 0.0401  83727.5652 - 0.0000\r\n",
      "TIM2+F+I+G4     -41601.1372  83304.2743 - 0.0346  83304.4520 - 0.0345  83727.8691 - 0.0000\r\n",
      "HKY+F           -41622.7556  83339.5112 - 0.0000  83339.6623 - 0.0000  83729.8829 - 0.0000\r\n",
      "TPM2u+F         -41617.8149  83331.6299 - 0.0000  83331.7875 - 0.0000  83730.3074 - 0.0000\r\n",
      "TPM2+F          -41617.8982  83331.7963 - 0.0000  83331.9539 - 0.0000  83730.4738 - 0.0000\r\n",
      "TIM+F+R2        -41603.4196  83308.8391 - 0.0035  83309.0168 - 0.0035  83732.4339 - 0.0000\r\n",
      "TIM+F+I+G4      -41603.5687  83309.1374 - 0.0030  83309.3151 - 0.0030  83732.7322 - 0.0000\r\n",
      "K3Pu+F          -41620.1309  83336.2619 - 0.0000  83336.4195 - 0.0000  83734.9394 - 0.0000\r\n",
      "GTR+F+G4        -41599.6434  83303.2868 + 0.0567  83303.4714 + 0.0563  83735.1874 - 0.0000\r\n",
      "GTR+F+I         -41599.6514  83303.3027 + 0.0562  83303.4874 + 0.0559  83735.2034 - 0.0000\r\n",
      "TIM3+F+R2       -41605.5084  83313.0168 - 0.0004  83313.1945 - 0.0004  83736.6117 - 0.0000\r\n",
      "TIM3+F+I+G4     -41605.6493  83313.2986 - 0.0004  83313.4763 - 0.0004  83736.8934 - 0.0000\r\n",
      "TPM3u+F         -41622.3443  83340.6885 - 0.0000  83340.8461 - 0.0000  83739.3660 - 0.0000\r\n",
      "TPM3+F          -41622.3741  83340.7482 - 0.0000  83340.9057 - 0.0000  83739.4257 - 0.0000\r\n",
      "HKY+F+I         -41622.7551  83341.5102 - 0.0000  83341.6678 - 0.0000  83740.1877 - 0.0000\r\n",
      "HKY+F+G4        -41622.7553  83341.5106 - 0.0000  83341.6682 - 0.0000  83740.1881 - 0.0000\r\n",
      "TPM2u+F+G4      -41617.7870  83333.5741 - 0.0000  83333.7382 - 0.0000  83740.5573 - 0.0000\r\n",
      "TPM2u+F+I       -41617.8059  83333.6118 - 0.0000  83333.7759 - 0.0000  83740.5951 - 0.0000\r\n",
      "TPM2+F+G4       -41617.8741  83333.7482 - 0.0000  83333.9124 - 0.0000  83740.7315 - 0.0000\r\n",
      "TPM2+F+I        -41617.8858  83333.7717 - 0.0000  83333.9358 - 0.0000  83740.7549 - 0.0000\r\n",
      "K3Pu+F+G4       -41620.1249  83338.2498 - 0.0000  83338.4139 - 0.0000  83745.2330 - 0.0000\r\n",
      "K3Pu+F+I        -41620.1249  83338.2498 - 0.0000  83338.4139 - 0.0000  83745.2330 - 0.0000\r\n",
      "GTR+F+R2        -41599.7108  83305.4215 - 0.0195  83305.6133 - 0.0193  83745.6279 - 0.0000\r\n",
      "GTR+F+I+G4      -41599.8602  83305.7203 - 0.0168  83305.9121 - 0.0166  83745.9267 - 0.0000\r\n",
      "TN+F+R3         -41605.9721  83315.9442 - 0.0001  83316.1288 - 0.0001  83747.8448 - 0.0000\r\n",
      "TIM2+F+R3       -41600.9665  83307.9330 - 0.0056  83308.1247 - 0.0055  83748.1394 - 0.0000\r\n",
      "TVM+F           -41616.4828  83332.9656 - 0.0000  83333.1364 - 0.0000  83748.2546 - 0.0000\r\n",
      "TPM3u+F+G4      -41622.3428  83342.6855 - 0.0000  83342.8496 - 0.0000  83749.6688 - 0.0000\r\n",
      "TPM3u+F+I       -41622.3433  83342.6866 - 0.0000  83342.8507 - 0.0000  83749.6699 - 0.0000\r\n",
      "TPM3+F+G4       -41622.3672  83342.7344 - 0.0000  83342.8985 - 0.0000  83749.7176 - 0.0000\r\n",
      "TPM3+F+I        -41622.3700  83342.7401 - 0.0000  83342.9042 - 0.0000  83749.7233 - 0.0000\r\n",
      "HKY+F+R2        -41622.8375  83343.6750 - 0.0000  83343.8391 - 0.0000  83750.6583 - 0.0000\r\n",
      "HKY+F+I+G4      -41622.9580  83343.9159 - 0.0000  83344.0800 - 0.0000  83750.8992 - 0.0000\r\n",
      "TPM2u+F+R2      -41617.8563  83335.7126 - 0.0000  83335.8834 - 0.0000  83751.0016 - 0.0000\r\n",
      "TPM2+F+R2       -41617.9313  83335.8626 - 0.0000  83336.0334 - 0.0000  83751.1516 - 0.0000\r\n",
      "TPM2u+F+I+G4    -41617.9873  83335.9746 - 0.0000  83336.1455 - 0.0000  83751.2637 - 0.0000\r\n",
      "TPM2+F+I+G4     -41618.0675  83336.1350 - 0.0000  83336.3059 - 0.0000  83751.4241 - 0.0000\r\n",
      "TIM+F+R3        -41603.3981  83312.7962 - 0.0005  83312.9879 - 0.0005  83753.0026 - 0.0000\r\n",
      "K3Pu+F+R2       -41620.2062  83340.4124 - 0.0000  83340.5832 - 0.0000  83755.7014 - 0.0000\r\n",
      "K3Pu+F+I+G4     -41620.3293  83340.6587 - 0.0000  83340.8295 - 0.0000  83755.9477 - 0.0000\r\n",
      "TIM3+F+R3       -41605.4931  83316.9862 - 0.0001  83317.1780 - 0.0001  83757.1926 - 0.0000\r\n",
      "TVM+F+G4        -41616.4408  83334.8816 - 0.0000  83335.0592 - 0.0000  83758.4764 - 0.0000\r\n",
      "TVM+F+I         -41616.4439  83334.8878 - 0.0000  83335.0654 - 0.0000  83758.4826 - 0.0000\r\n",
      "TPM3u+F+R2      -41622.4168  83344.8337 - 0.0000  83345.0045 - 0.0000  83760.1227 - 0.0000\r\n",
      "TPM3+F+R2       -41622.4374  83344.8748 - 0.0000  83345.0457 - 0.0000  83760.1639 - 0.0000\r\n",
      "TPM3u+F+I+G4    -41622.5392  83345.0785 - 0.0000  83345.2493 - 0.0000  83760.3675 - 0.0000\r\n",
      "TPM3+F+I+G4     -41622.5622  83345.1243 - 0.0000  83345.2952 - 0.0000  83760.4134 - 0.0000\r\n",
      "GTR+F+R3        -41599.6959  83309.3919 - 0.0027  83309.5983 - 0.0026  83766.2098 - 0.0000\r\n",
      "TVM+F+R2        -41616.5072  83337.0145 - 0.0000  83337.1991 - 0.0000  83768.9151 - 0.0000\r\n",
      "TVM+F+I+G4      -41616.6370  83337.2741 - 0.0000  83337.4587 - 0.0000  83769.1747 - 0.0000\r\n",
      "HKY+F+R3        -41622.8264  83347.6528 - 0.0000  83347.8305 - 0.0000  83771.2477 - 0.0000\r\n",
      "TPM2u+F+R3      -41617.8368  83339.6736 - 0.0000  83339.8582 - 0.0000  83771.5742 - 0.0000\r\n",
      "TPM2+F+R3       -41617.9088  83339.8177 - 0.0000  83340.0023 - 0.0000  83771.7183 - 0.0000\r\n",
      "K3Pu+F+R3       -41620.1915  83344.3829 - 0.0000  83344.5676 - 0.0000  83776.2835 - 0.0000\r\n",
      "TPM3u+F+R3      -41622.4018  83348.8036 - 0.0000  83348.9883 - 0.0000  83780.7043 - 0.0000\r\n",
      "TPM3+F+R3       -41622.4239  83348.8477 - 0.0000  83349.0324 - 0.0000  83780.7484 - 0.0000\r\n",
      "F81+F           -41653.7554  83399.5108 - 0.0000  83399.6556 - 0.0000  83781.5767 - 0.0000\r\n",
      "TVM+F+R3        -41616.4950  83340.9901 - 0.0000  83341.1891 - 0.0000  83789.5023 - 0.0000\r\n",
      "F81+F+I         -41653.7554  83401.5108 - 0.0000  83401.6619 - 0.0000  83791.8825 - 0.0000\r\n",
      "F81+F+G4        -41653.7555  83401.5111 - 0.0000  83401.6622 - 0.0000  83791.8828 - 0.0000\r\n",
      "F81+F+R2        -41653.8307  83403.6614 - 0.0000  83403.8190 - 0.0000  83802.3389 - 0.0000\r\n",
      "F81+F+I+G4      -41653.9419  83403.8837 - 0.0000  83404.0413 - 0.0000  83802.5612 - 0.0000\r\n",
      "F81+F+R3        -41653.8182  83407.6364 - 0.0000  83407.8072 - 0.0000  83822.9254 - 0.0000\r\n",
      "TNe             -42515.4109  85120.8219 - 0.0000  85120.9605 - 0.0000  85494.5820 - 0.0000\r\n",
      "TIM2e           -42512.7440  85117.4879 - 0.0000  85117.6327 - 0.0000  85499.5538 - 0.0000\r\n",
      "TIMe            -42513.5746  85119.1493 - 0.0000  85119.2941 - 0.0000  85501.2152 - 0.0000\r\n",
      "TIM3e           -42514.5282  85121.0564 - 0.0000  85121.2012 - 0.0000  85503.1223 - 0.0000\r\n",
      "TNe+G4          -42515.3963  85122.7926 - 0.0000  85122.9374 - 0.0000  85504.8585 - 0.0000\r\n",
      "TNe+I           -42515.4022  85122.8045 - 0.0000  85122.9493 - 0.0000  85504.8704 - 0.0000\r\n",
      "TIM2e+G4        -42512.6722  85119.3444 - 0.0000  85119.4955 - 0.0000  85509.7161 - 0.0000\r\n",
      "TIM2e+I         -42512.7369  85119.4739 - 0.0000  85119.6250 - 0.0000  85509.8456 - 0.0000\r\n",
      "TIMe+G4         -42513.5451  85121.0902 - 0.0000  85121.2413 - 0.0000  85511.4619 - 0.0000\r\n",
      "TIMe+I          -42513.5577  85121.1154 - 0.0000  85121.2666 - 0.0000  85511.4871 - 0.0000\r\n",
      "TIM3e+G4        -42514.4660  85122.9321 - 0.0000  85123.0832 - 0.0000  85513.3038 - 0.0000\r\n",
      "TIM3e+I         -42514.4846  85122.9692 - 0.0000  85123.1204 - 0.0000  85513.3409 - 0.0000\r\n",
      "SYM             -42510.1023  85116.2046 - 0.0000  85116.3622 - 0.0000  85514.8821 - 0.0000\r\n",
      "TNe+R2          -42515.4578  85124.9156 - 0.0000  85125.0667 - 0.0000  85515.2873 - 0.0000\r\n",
      "TNe+I+G4        -42515.5676  85125.1353 - 0.0000  85125.2864 - 0.0000  85515.5070 - 0.0000\r\n",
      "K2P             -42531.8875  85151.7751 - 0.0000  85151.9077 - 0.0000  85517.2295 - 0.0000\r\n",
      "TIM2e+R2        -42512.1567  85120.3133 - 0.0000  85120.4709 - 0.0000  85518.9908 - 0.0000\r\n",
      "TIM2e+I+G4      -42512.3038  85120.6075 - 0.0000  85120.7651 - 0.0000  85519.2850 - 0.0000\r\n",
      "TIMe+R2         -42513.5868  85123.1737 - 0.0000  85123.3313 - 0.0000  85521.8512 - 0.0000\r\n",
      "TIMe+I+G4       -42513.7092  85123.4184 - 0.0000  85123.5760 - 0.0000  85522.0959 - 0.0000\r\n",
      "K3P             -42529.9327  85149.8653 - 0.0000  85150.0040 - 0.0000  85523.6255 - 0.0000\r\n",
      "TIM3e+R2        -42514.4816  85124.9632 - 0.0000  85125.1207 - 0.0000  85523.6406 - 0.0000\r\n",
      "TIM3e+I+G4      -42514.6038  85125.2077 - 0.0000  85125.3653 - 0.0000  85523.8852 - 0.0000\r\n",
      "SYM+G4          -42510.1001  85118.2003 - 0.0000  85118.3644 - 0.0000  85525.1835 - 0.0000\r\n",
      "SYM+I           -42510.1010  85118.2021 - 0.0000  85118.3662 - 0.0000  85525.1853 - 0.0000\r\n",
      "K2P+I           -42531.8875  85153.7750 - 0.0000  85153.9137 - 0.0000  85527.5351 - 0.0000\r\n",
      "K2P+G4          -42531.8876  85153.7753 - 0.0000  85153.9139 - 0.0000  85527.5354 - 0.0000\r\n",
      "K3P+G4          -42529.9325  85151.8651 - 0.0000  85152.0099 - 0.0000  85533.9310 - 0.0000\r\n",
      "K3P+I           -42529.9326  85151.8652 - 0.0000  85152.0100 - 0.0000  85533.9311 - 0.0000\r\n",
      "SYM+R2          -42510.1492  85120.2983 - 0.0000  85120.4692 - 0.0000  85535.5874 - 0.0000\r\n",
      "SYM+I+G4        -42510.2597  85120.5193 - 0.0000  85120.6902 - 0.0000  85535.8084 - 0.0000\r\n",
      "TNe+R3          -42515.4445  85128.8890 - 0.0000  85129.0531 - 0.0000  85535.8722 - 0.0000\r\n",
      "TVMe            -42526.6099  85147.2198 - 0.0000  85147.3709 - 0.0000  85537.5915 - 0.0000\r\n",
      "K2P+R2          -42531.9529  85155.9058 - 0.0000  85156.0506 - 0.0000  85537.9717 - 0.0000\r\n",
      "K2P+I+G4        -42532.0538  85156.1076 - 0.0000  85156.2524 - 0.0000  85538.1735 - 0.0000\r\n",
      "TIM2e+R3        -42512.1267  85124.2533 - 0.0000  85124.4242 - 0.0000  85539.5424 - 0.0000\r\n",
      "TIMe+R3         -42513.5666  85127.1332 - 0.0000  85127.3040 - 0.0000  85542.4222 - 0.0000\r\n",
      "TIM3e+R3        -42514.4449  85128.8898 - 0.0000  85129.0606 - 0.0000  85544.1788 - 0.0000\r\n",
      "K3P+R2          -42529.9953  85153.9905 - 0.0000  85154.1416 - 0.0000  85544.3622 - 0.0000\r\n",
      "K3P+I+G4        -42530.0969  85154.1937 - 0.0000  85154.3449 - 0.0000  85544.5654 - 0.0000\r\n",
      "TVMe+G4         -42526.5960  85149.1919 - 0.0000  85149.3495 - 0.0000  85547.8694 - 0.0000\r\n",
      "TVMe+I          -42526.6015  85149.2030 - 0.0000  85149.3606 - 0.0000  85547.8805 - 0.0000\r\n",
      "SYM+R3          -42510.1359  85124.2719 - 0.0000  85124.4565 - 0.0000  85556.1725 - 0.0000\r\n",
      "TVMe+R2         -42526.6392  85151.2784 - 0.0000  85151.4425 - 0.0000  85558.2617 - 0.0000\r\n",
      "TVMe+I+G4       -42526.7486  85151.4971 - 0.0000  85151.6612 - 0.0000  85558.4804 - 0.0000\r\n",
      "K2P+R3          -42531.9407  85159.8814 - 0.0000  85160.0389 - 0.0000  85558.5589 - 0.0000\r\n",
      "JC              -42560.5207  85207.0414 - 0.0000  85207.1681 - 0.0000  85564.1900 - 0.0000\r\n",
      "K3P+R3          -42529.9846  85157.9691 - 0.0000  85158.1332 - 0.0000  85564.9524 - 0.0000\r\n",
      "JC+I            -42560.5207  85209.0414 - 0.0000  85209.1740 - 0.0000  85574.4957 - 0.0000\r\n",
      "JC+G4           -42560.5208  85209.0417 - 0.0000  85209.1743 - 0.0000  85574.4960 - 0.0000\r\n",
      "TVMe+R3         -42526.6255  85155.2511 - 0.0000  85155.4288 - 0.0000  85578.8459 - 0.0000\r\n",
      "JC+R2           -42560.5879  85211.1758 - 0.0000  85211.3144 - 0.0000  85584.9359 - 0.0000\r\n",
      "JC+I+G4         -42560.6893  85211.3786 - 0.0000  85211.5173 - 0.0000  85585.1387 - 0.0000\r\n",
      "JC+R3           -42560.5633  85215.1266 - 0.0000  85215.2778 - 0.0000  85605.4983 - 0.0000\r\n",
      "\r\n",
      "AIC, w-AIC   : Akaike information criterion scores and weights.\r\n",
      "AICc, w-AICc : Corrected AIC scores and weights.\r\n",
      "BIC, w-BIC   : Bayesian information criterion scores and weights.\r\n",
      "\r\n",
      "Plus signs denote the 95% confidence sets.\r\n",
      "Minus signs denote significant exclusion.\r\n",
      "\r\n",
      "SUBSTITUTION PROCESS\r\n",
      "--------------------\r\n",
      "\r\n",
      "Model of substitution: TN+F\r\n",
      "\r\n",
      "Rate parameter R:\r\n",
      "\r\n",
      "  A-C: 1.0000\r\n",
      "  A-G: 1.6732\r\n",
      "  A-T: 1.0000\r\n",
      "  C-G: 1.0000\r\n",
      "  C-T: 7.9967\r\n",
      "  G-T: 1.0000\r\n",
      "\r\n",
      "State frequencies: (empirical counts from alignment)\r\n",
      "\r\n",
      "  pi(A) = 0.2991\r\n",
      "  pi(C) = 0.1832\r\n",
      "  pi(G) = 0.1962\r\n",
      "  pi(T) = 0.3215\r\n",
      "\r\n",
      "Rate matrix Q:\r\n",
      "\r\n",
      "  A   -0.5084    0.1118    0.2004    0.1962\r\n",
      "  C    0.1826    -1.871    0.1198     1.569\r\n",
      "  G    0.3055    0.1118   -0.6135    0.1962\r\n",
      "  T    0.1826    0.8943    0.1198    -1.197\r\n",
      "\r\n",
      "Model of rate heterogeneity: Uniform\r\n",
      "\r\n",
      "MAXIMUM LIKELIHOOD TREE\r\n",
      "-----------------------\r\n",
      "\r\n",
      "Log-likelihood of the tree: -41605.7956 (s.e. 102.1628)\r\n",
      "Unconstrained log-likelihood (without tree): -58428.3435\r\n",
      "Number of free parameters (#branches + #model parameters): 48\r\n",
      "Akaike information criterion (AIC) score: 83307.5912\r\n",
      "Corrected Akaike information criterion (AICc) score: 83307.7488\r\n",
      "Bayesian information criterion (BIC) score: 83706.2687\r\n",
      "\r\n",
      "Total tree length (sum of branch lengths): 0.0037\r\n",
      "Sum of internal branch lengths: 0.0014 (38.7310% of tree length)\r\n",
      "\r\n",
      "WARNING: 12 near-zero internal branches (<0.0000) should be treated with caution\r\n",
      "         Such branches are denoted by '**' in the figure below\r\n",
      "\r\n",
      "NOTE: Tree is UNROOTED although outgroup taxon 'NC_045512.2' is drawn at root\r\n",
      "\r\n",
      "+-----NC_045512.2\r\n",
      "|\r\n",
      "|              +**COG111-PC_S86\r\n",
      "|  +-----------|\r\n",
      "|  |           +**COG112-PC_S114\r\n",
      "+**|\r\n",
      "|  |                                                          +**NORW-220299_S80\r\n",
      "|  |                                                       +**|\r\n",
      "|  |                                                       |  +**NORW-22DDC8_S95\r\n",
      "|  |                                              +--------|\r\n",
      "|  |                                              |        |  +**NORW-22AE40_S97\r\n",
      "|  |                                              |        +--|\r\n",
      "|  |                                              |           +**NORW-22DFFF_S94\r\n",
      "|  |                                           +--|\r\n",
      "|  |                                           |  +--NORW-232980_S124\r\n",
      "|  |                                        +**|\r\n",
      "|  |                                        |  +-------------NORW-2326C5_S143\r\n",
      "|  |                                     +**|\r\n",
      "|  |                                     |  |        +-------------------NORW-22A50F_S96\r\n",
      "|  |                                     |  |     +**|\r\n",
      "|  |                                     |  |     |  |              +**NORW-23264D_S112\r\n",
      "|  |                                     |  |     |  |           +**|\r\n",
      "|  |                                     |  |     |  |           |  |  +--NORW-232892_S118\r\n",
      "|  |                                     |  |     |  |           |  +**|\r\n",
      "|  |                                     |  |     |  |           |     |     +**NORW-232935_S135\r\n",
      "|  |                                     |  |     |  |           |     |  +**|\r\n",
      "|  |                                     |  |     |  |           |     |  |  +**NORW-232BD5_S138\r\n",
      "|  |                                     |  |     |  |           |     +**|\r\n",
      "|  |                                     |  |     |  |           |        +**NORW-232971_S125\r\n",
      "|  |                                     |  |     |  +-----------|\r\n",
      "|  |                                     |  |     |              +--NORW-23267A_S106\r\n",
      "|  |                                     |  |  +**|\r\n",
      "|  |                                     |  |  |  +-------NORW-2326F2_S128\r\n",
      "|  |                                     |  +**|\r\n",
      "|  |                                     |     +--NORW-23299F_S140\r\n",
      "|  +-------------------------------------|\r\n",
      "|                                        |     +---NORW-232829_S109\r\n",
      "|                                        |  +**|\r\n",
      "|                                        |  |  |  +---NORW-232847_S115\r\n",
      "|                                        |  |  +--|\r\n",
      "|                                        |  |     +--NORW-2328DE_S133\r\n",
      "|                                        +--|\r\n",
      "|                                           +--NORW-232883_S142\r\n",
      "|\r\n",
      "+----------------NORW-2326A7_S131\r\n",
      "\r\n",
      "Tree in newick format:\r\n",
      "\r\n",
      "(NC_045512.2:0.0001334687,((COG111-PC_S86:0.0000020137,COG112-PC_S114:0.0000020016):0.0002680557,((((((NORW-220299_S80:0.0000025025,NORW-22DDC8_S95:0.0000020011):0.0000021551,(NORW-22AE40_S97:0.0000010000,NORW-22DFFF_S94:0.0000010000):0.0000334989):0.0002012528,NORW-232980_S124:0.0000338312):0.0000335226,NORW-2326C5_S143:0.0003022161):0.0000012900,(((NORW-22A50F_S96:0.0004354599,((NORW-23264D_S112:0.0000010000,(NORW-232892_S118:0.0000336300,((NORW-232935_S135:0.0000020137,NORW-232BD5_S138:0.0000020137):0.0000025713,NORW-232971_S125:0.0000010000):0.0000020300):0.0000010000):0.0000333970,NORW-23267A_S106:0.0000669636):0.0002670747):0.0000020008,NORW-2326F2_S128:0.0001708848):0.0000020092,NORW-23299F_S140:0.0000336503):0.0000020086):0.0000020350,((NORW-232829_S109:0.0001018665,(NORW-232847_S115:0.0001025700,NORW-2328DE_S133:0.0000671156):0.0000334755):0.0000010000,NORW-232883_S142:0.0000672216):0.0000334789):0.0008063101):0.0000020037,NORW-2326A7_S131:0.0003755238);\r\n",
      "\r\n",
      "TIME STAMP\r\n",
      "----------\r\n",
      "\r\n",
      "Date and time: Fri Feb  4 12:11:11 2022\r\n",
      "Total CPU time used: 1.32442 seconds (0h:0m:1s)\r\n",
      "Total wall-clock time used: 1.37982 seconds (0h:0m:1s)\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! cat output.fa.iqtree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ded8ba1",
   "metadata": {},
   "source": [
    "In the output file for IQTREE, there is this note:\n",
    "    \n",
    "    WARNING: 12 near-zero internal branches (<0.0000) should be treated with caution\n",
    "         Such branches are denoted by '**' in the figure below\n",
    "         \n",
    "This means that there are [polytomies](https://en.wikipedia.org/wiki/Polytomy) present in the data.  To (hopefully) resolve these, use TreeTime!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8f08062",
   "metadata": {},
   "source": [
    "# TreeTime"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "623f0619",
   "metadata": {},
   "source": [
    "TreeTime optimizes branch lengths and resolves polytomies by incorporating sampling date information into the tree building process.  The sample dates can be found in ________________.  The treetime source is [here](https://github.com/neherlab/treetime) and here's the [documentation](https://treetime.readthedocs.io/en/latest/index.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "53e2892d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: TreeTime: Maximum Likelihood Phylodynamics\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  {homoplasy,ancestral,mugration,clock,version}\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help            show this help message and exit\r\n",
      "  --tree TREE           Name of file containing the tree in newick, nexus, or\r\n",
      "                        phylip format. If none is provided, treetime will\r\n",
      "                        attempt to build a tree from the alignment using\r\n",
      "                        fasttree, iqtree, or raxml (assuming they are\r\n",
      "                        installed)\r\n",
      "  --sequence-length SEQUENCE_LENGTH\r\n",
      "                        length of the sequence, used to calculate expected\r\n",
      "                        variation in branch length. Not required if alignment\r\n",
      "                        is provided.\r\n",
      "  --aln ALN             alignment file (fasta)\r\n",
      "  --vcf-reference VCF_REFERENCE\r\n",
      "                        only for vcf input: fasta file of the sequence the VCF\r\n",
      "                        was mapped to.\r\n",
      "  --dates DATES         csv file with dates for nodes with 'node_name, date'\r\n",
      "                        where date is float (as in 2012.15)\r\n",
      "  --name-column NAME_COLUMN\r\n",
      "                        label of the column to be used as taxon name\r\n",
      "  --date-column DATE_COLUMN\r\n",
      "                        label of the column to be used as sampling date\r\n",
      "  --clock-filter CLOCK_FILTER\r\n",
      "                        ignore tips that don't follow a loose clock, 'clock-\r\n",
      "                        filter=number of interquartile ranges from\r\n",
      "                        regression'. Default=3.0, set to 0 to switch off.\r\n",
      "  --reroot REROOT [REROOT ...]\r\n",
      "                        Reroot the tree using root-to-tip regression. Valid\r\n",
      "                        choices are 'min_dev', 'least-squares', and 'oldest'.\r\n",
      "                        'least-squares' adjusts the root to minimize residuals\r\n",
      "                        of the root-to-tip vs sampling time regression,\r\n",
      "                        'min_dev' minimizes variance of root-to-tip distances.\r\n",
      "                        'least-squares' can be combined with --covariation to\r\n",
      "                        account for shared ancestry. Alternatively, you can\r\n",
      "                        specify a node name or a list of node names to be used\r\n",
      "                        as outgroup or use 'oldest' to reroot to the oldest\r\n",
      "                        node. By default, TreeTime will reroot using 'least-\r\n",
      "                        squares'. Use --keep-root to keep the current root.\r\n",
      "  --keep-root           don't reroot the tree. Otherwise, reroot to minimize\r\n",
      "                        the the residual of the regression of root-to-tip\r\n",
      "                        distance and sampling time\r\n",
      "  --tip-slack TIP_SLACK\r\n",
      "                        excess variance associated with terminal nodes\r\n",
      "                        accounting for overdisperion of the molecular clock\r\n",
      "  --covariation         Account for covariation when estimating rates or\r\n",
      "                        rerooting using root-to-tip regression, default False.\r\n",
      "  --gtr GTR             GTR model to use. '--gtr infer' will infer a model\r\n",
      "                        from the data. Alternatively, specify the model type.\r\n",
      "                        If the specified model requires additional options,\r\n",
      "                        use '--gtr-params' to specify those.\r\n",
      "  --gtr-params GTR_PARAMS [GTR_PARAMS ...]\r\n",
      "                        GTR parameters for the model specified by the --gtr\r\n",
      "                        argument. The parameters should be feed as 'key=value'\r\n",
      "                        list of parameters. Example: '--gtr K80 --gtr-params\r\n",
      "                        kappa=0.2 pis=0.25,0.25,0.25,0.25'. See the exact\r\n",
      "                        definitions of the parameters in the GTR creation\r\n",
      "                        methods in treetime/nuc_models.py or\r\n",
      "                        treetime/aa_models.py\r\n",
      "  --aa                  use aminoacid alphabet\r\n",
      "  --clock-rate CLOCK_RATE\r\n",
      "                        if specified, the rate of the molecular clock won't be\r\n",
      "                        optimized.\r\n",
      "  --clock-std-dev CLOCK_STD_DEV\r\n",
      "                        standard deviation of the provided clock rate estimate\r\n",
      "  --branch-length-mode {auto,input,joint,marginal}\r\n",
      "                        If set to 'input', the provided branch length will be\r\n",
      "                        used without modification. Note that branch lengths\r\n",
      "                        optimized by treetime are only accurate at short\r\n",
      "                        evolutionary distances.\r\n",
      "  --confidence          estimate confidence intervals of divergence times.\r\n",
      "  --keep-polytomies     Don't resolve polytomies using temporal information.\r\n",
      "  --relax RELAX RELAX   use an autocorrelated molecular clock. Strength of the\r\n",
      "                        gaussian priors on branch specific rate deviation and\r\n",
      "                        the coupling of parent and offspring rates can be\r\n",
      "                        specified e.g. as --relax 1.0 0.5. Values around 1.0\r\n",
      "                        correspond to weak priors, larger values constrain\r\n",
      "                        rate deviations more strongly. Coupling 0 (--relax 1.0\r\n",
      "                        0) corresponds to an un-correlated clock.\r\n",
      "  --max-iter MAX_ITER   maximal number of iterations the inference cycle is\r\n",
      "                        run. Note that for polytomy resolution and coalescence\r\n",
      "                        models max_iter should be at least 2\r\n",
      "  --coalescent COALESCENT\r\n",
      "                        coalescent time scale -- sensible values are on the\r\n",
      "                        order of the average hamming distance of\r\n",
      "                        contemporaneous sequences. In addition, 'opt'\r\n",
      "                        'skyline' are valid options and estimate a constant\r\n",
      "                        coalescent rate or a piecewise linear coalescent rate\r\n",
      "                        history\r\n",
      "  --n-skyline N_SKYLINE\r\n",
      "                        number of grid points in skyline coalescent model\r\n",
      "  --plot-tree PLOT_TREE\r\n",
      "                        filename to save the plot to. Suffix will determine\r\n",
      "                        format (choices pdf, png, svg, default=pdf)\r\n",
      "  --plot-rtt PLOT_RTT   filename to save the plot to. Suffix will determine\r\n",
      "                        format (choices pdf, png, svg, default=pdf)\r\n",
      "  --tip-labels          add tip labels (default for small trees with <30\r\n",
      "                        leaves)\r\n",
      "  --no-tip-labels       don't show tip labels (default for small trees with\r\n",
      "                        >=30 leaves)\r\n",
      "  --keep-overhangs      do not fill terminal gaps\r\n",
      "  --zero-based          zero based mutation indexing\r\n",
      "  --reconstruct-tip-states\r\n",
      "                        overwrite ambiguous states on tips with the most\r\n",
      "                        likely inferred state\r\n",
      "  --report-ambiguous    include transitions involving ambiguous states\r\n",
      "  --verbose VERBOSE     verbosity of output 0-6\r\n",
      "  --outdir OUTDIR       directory to write the output to\r\n",
      "  --version             show program's version number and exit\r\n"
     ]
    }
   ],
   "source": [
    "! treetime -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8155e8e5",
   "metadata": {},
   "source": [
    "The inputs `treetime` requires are: a .csv of the sampling dates, the Newick tree from IQTREE, an outdir, and optionally the alignment from MAFFT.  Various formats of inputs can be used (see documentation).  The simplest implementation of treetime is:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85d5074d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#! treetime --tree output.fa.treefile --dates data/h3n2_na/h3n2_na_20.metadata.csv --aln output.fa --outdir 20210513_timetree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42525224",
   "metadata": {},
   "source": [
    "The output is a \"GTR model, a molecular clock model, and a time-stamped phylogeny\"."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b43c742",
   "metadata": {},
   "source": [
    "If the evolutionary rate is hard to estimate from the data, you can specify the evolutionary rate using `--clock-rate <rate>`.  In the paper _________ the clock rate was _____."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef8fdefd",
   "metadata": {},
   "source": [
    "**Coalescent priors in TreeTime**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7b6bcb3",
   "metadata": {},
   "source": [
    "A coalescent prior is optional, and can be added using `--coalescent <arg>`.  A coalescent prior is a time scale in units of time to divergence (time to coalescence, when viewed in reverse), given as a floating point number.  \n",
    "\n",
    "According to [Wikipedia](https://en.wikipedia.org/wiki/Coalescent_theory), coalescent theory is about using models to estimate the time to coalescence (time backwards from two lineages to their most recent common ancestor) in evolutionary trees, including aspects like recombination, natural selection, gene flow, and other factors.  The time to coalescence is the number of preceding generations to coalescence (how many generations beforehand the most recent common ancestor was found).  In general, the expected time to coalescence, in units of generations, is equal to the population size (# copies of each locus).  A \"coalescent event\" is the place where two branches meet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "54b961e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! treetime --tree data/ebola/ebola.nwk --dates data/ebola/ebola.metadata.csv --aln data/ebola/ebola.fasta --outdir ebola  --coalescent skyline"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
