{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "044214b9",
   "metadata": {},
   "source": [
    "**Back to TreeTime**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6f40ca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import libraries\n",
    "from IPython.display import IFrame\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daa126a6",
   "metadata": {},
   "source": [
    "**Rerooting Phylogenetic Trees**\n",
    "\n",
    "It's important to never trust the rooting from IQTREE or similar: the output is always unrooted, no matter how rooted it may look!  This means that `--keep-root` in TreeTime won't work, since there is no root.  However, you can define the root in TreeTime using `--reroot 'oldest'`, if you know the oldest sample, or one of the other rerooting options."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4fbe32b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! treetime -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a43667e0",
   "metadata": {},
   "source": [
    "**Molecular Clock**\n",
    "\n",
    "Usually, the molecular clock rate is set to `0.008` for SARS-CoV-2.  This is the average mutation rate estimate, but it doesn't work in general for epidemiology.  Using a hard clock will cause TreeTime to ignore some tips so it can get the clock right.  This is noticeable especially when you have clusters from different lineages, where the jump in mutation counts between lineages is bigger than the clock would estimate (like the jump to Omicron).  In these cases, if you want a regression plot like TreeTime's, you can view the plot in Tempest and make a regression line within each cluster separately.\n",
    "\n",
    "For a more realistic measure, you can use the **relaxed molecular clock**, which says that each cluster can have its own mutation rate independent of the other clusters.  Leo recommends setting `--relax 1.0 0`.  It's good to include a molecular clock rate as well, just so that all the clusters together have to average out to `0.008`.  Once you've done both of these, TreeTime's regression plot kind of means nothing.\n",
    "\n",
    "The number used for the rate is the one used in publications in the past, but with Omicron being super transmissible it might not work anymore.  Proceed with caution if your samples are ~90% Omicron, then.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "456911a4",
   "metadata": {},
   "source": [
    "**Wuhan-Hu-1**\n",
    "\n",
    "Wuhan-Wu-1 (MN908947) is one of the two earliest samples we have for SARS-CoV-2, and it's the one most often used as the reference genome for SARS-CoV-2 (in this lab, always).  It comes from a patient sample taken on December 26, 2019, and is from the B lineage.  The other sample, Wuhan-Hu-4, is from the A lineage and was sampled in January 2020.  For more on SARS-CoV-2 nomenclature, see [this paper](https://www.nature.com/articles/s41564-020-0770-5).\n",
    "\n",
    "Since we only have access to the A and B lineage references but not their common ancestor, it can happen that you'll build a tree with Wuhan-Hu-1 in it only to see that it's not a common ancestor for all the samples in the tree.\n",
    "\n",
    "In general, it's difficult to infer a good tree using samples all clustered near each other in time plus one sample from years ago--like the other day, when I tried to build a tree using 200 samples from winter 2021 + Wuhan-Wu-1.  If you want to include the reference, a good way to do it is to build the tree using just the 200 samples, and then to add the reference sequence afterwards.  In IQTREE you can do this with `-g`.  **Question: Is this done before TreeTime or after?  I think after.  See if it can be done.**\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "982d27fa",
   "metadata": {},
   "source": [
    "**Visualizing Nexus trees**\n",
    "\n",
    "Nexus trees can be viewed using `figtree`: `figtree mytree.nexus`.  Here's one of mine from the other day:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e826c4ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "javax.swing.UIManager$LookAndFeelInfo[Metal javax.swing.plaf.metal.MetalLookAndFeel]\r\n",
      "javax.swing.UIManager$LookAndFeelInfo[Nimbus javax.swing.plaf.nimbus.NimbusLookAndFeel]\r\n",
      "javax.swing.UIManager$LookAndFeelInfo[CDE/Motif com.sun.java.swing.plaf.motif.MotifLookAndFeel]\r\n",
      "javax.swing.UIManager$LookAndFeelInfo[GTK+ com.sun.java.swing.plaf.gtk.GTKLookAndFeel]\r\n"
     ]
    }
   ],
   "source": [
    "#this will open the file in a new window\n",
    "! figtree COG-UK-NORW/results/timetree-tiplabels/timetree.nexus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1726fcd3",
   "metadata": {},
   "source": [
    "**Uniform vs. Coalescent Trees**\n",
    "\n",
    "In a uniform tree, the branch lengths between nodes is constant.  This is good for humans, cows and bacteria: things with discrete generations.  For viral transmission dynamics, it makes more sense to use a coalescent model, in which the branch lengths get shorter as time goes on, representing viral spread.  **Q: clarify this part.**  TreeTime has a few coalescent models, including a skyline model: test these out and see what they do."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfc714a6",
   "metadata": {},
   "source": [
    "**Testing some treetime models**\n",
    "\n",
    "These are going to be tested on the 200 NORW samples, without Wuhan-Hu-1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ca084734",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 161919 ambiguous characters.\n",
      "  101 / 200\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      "  190 / 200\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP   195 / 199 \n",
      "Reallocating..done. *alloclen = 60816\n",
      "STEP   199 / 199 \n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      "  100 / 200\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "  190 / 200\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP   199 / 199 \n",
      "done.\n",
      "\n",
      "disttbfast (nuc) Version 7.490\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-2 (Fast but rough)\n",
      " Progressive method (guide trees were built 2 times.)\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "ERROR: Could not open ../data/COG-UK-NORW/results/20220210/iqtree/NORW_200_aln.fa.log for logging\n"
     ]
    }
   ],
   "source": [
    "! mkdir -p COG-UK-NORW/results/20220210\n",
    "! mafft COG-UK-NORW/NORW_200.fasta > COG-UK-NORW/results/20220210/NORW_200_aln.fa #align NORW samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b496f00d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 1.6.12 for Linux 64-bit built Mar 23 2020\n",
      "Developed by Bui Quang Minh, Nguyen Lam Tung, Olga Chernomor,\n",
      "Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    tamarisk (AVX2, FMA3, 9 GB RAM)\n",
      "Command: iqtree -s ../data/COG-UK-NORW/results/20220210/NORW_200_aln.fa -pre ../data/COG-UK-NORW/results/20220210/NORW_200_aln.fa\n",
      "Seed:    628004 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Mon Feb 14 15:42:47 2022\n",
      "Kernel:  AVX+FMA - 1 threads (8 CPU cores detected)\n",
      "\n",
      "HINT: Use -nt option to specify number of threads because your CPU has 8 cores!\n",
      "HINT: -nt AUTO will automatically determine the best number of threads to use.\n",
      "\n",
      "Reading alignment file ../data/COG-UK-NORW/results/20220210/NORW_200_aln.fa ... Fasta format detected\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "WARNING: 506 sites contain only gaps or ambiguous characters.\n",
      "Alignment has 200 sequences with 29914 columns, 1567 distinct patterns\n",
      "174 parsimony-informative, 238 singleton sites, 29502 constant sites\n",
      "WARNING: Some sequence names are changed as follows:\n",
      "England/NORW-22D223/2021 -> England_NORW-22D223_2021\n",
      "England/NORW-22D1CC/2021 -> England_NORW-22D1CC_2021\n",
      "England/NORW-22D162/2021 -> England_NORW-22D162_2021\n",
      "England/NORW-22B37D/2021 -> England_NORW-22B37D_2021\n",
      "England/NORW-22D2C9/2021 -> England_NORW-22D2C9_2021\n",
      "England/NORW-22D1BD/2021 -> England_NORW-22D1BD_2021\n",
      "England/NORW-22D36C/2021 -> England_NORW-22D36C_2021\n",
      "England/NORW-22463F/2021 -> England_NORW-22463F_2021\n",
      "England/NORW-22465D/2021 -> England_NORW-22465D_2021\n",
      "England/NORW-22D205/2021 -> England_NORW-22D205_2021\n",
      "England/NORW-22B4C5/2021 -> England_NORW-22B4C5_2021\n",
      "England/NORW-22D27E/2021 -> England_NORW-22D27E_2021\n",
      "England/NORW-22D180/2021 -> England_NORW-22D180_2021\n",
      "England/NORW-22466C/2021 -> England_NORW-22466C_2021\n",
      "England/NORW-22464E/2021 -> England_NORW-22464E_2021\n",
      "England/NORW-22B085/2021 -> England_NORW-22B085_2021\n",
      "England/NORW-22D302/2021 -> England_NORW-22D302_2021\n",
      "England/NORW-22D1AE/2021 -> England_NORW-22D1AE_2021\n",
      "England/NORW-22D34E/2021 -> England_NORW-22D34E_2021\n",
      "England/NORW-22B00D/2021 -> England_NORW-22B00D_2021\n",
      "England/NORW-22B304/2021 -> England_NORW-22B304_2021\n",
      "England/NORW-22D214/2021 -> England_NORW-22D214_2021\n",
      "England/NORW-22D250/2021 -> England_NORW-22D250_2021\n",
      "England/NORW-22468A/2021 -> England_NORW-22468A_2021\n",
      "England/NORW-22D171/2021 -> England_NORW-22D171_2021\n",
      "England/NORW-22D26F/2021 -> England_NORW-22D26F_2021\n",
      "England/NORW-22D0DE/2021 -> England_NORW-22D0DE_2021\n",
      "England/NORW-22D3D5/2021 -> England_NORW-22D3D5_2021\n",
      "England/NORW-22D2F6/2021 -> England_NORW-22D2F6_2021\n",
      "England/NORW-22B2F8/2021 -> England_NORW-22B2F8_2021\n",
      "England/NORW-22D2BA/2021 -> England_NORW-22D2BA_2021\n",
      "England/NORW-22AF4D/2021 -> England_NORW-22AF4D_2021\n",
      "England/NORW-22D19F/2021 -> England_NORW-22D19F_2021\n",
      "England/NORW-22D108/2021 -> England_NORW-22D108_2021\n",
      "England/NORW-22D38A/2021 -> England_NORW-22D38A_2021\n",
      "England/NORW-22D2E7/2021 -> England_NORW-22D2E7_2021\n",
      "England/NORW-22D1F9/2021 -> England_NORW-22D1F9_2021\n",
      "England/NORW-22D3B7/2021 -> England_NORW-22D3B7_2021\n",
      "England/NORW-22D399/2021 -> England_NORW-22D399_2021\n",
      "England/NORW-22B3E6/2021 -> England_NORW-22B3E6_2021\n",
      "England/NORW-22B8AB/2021 -> England_NORW-22B8AB_2021\n",
      "England/NORW-22B38C/2021 -> England_NORW-22B38C_2021\n",
      "England/NORW-22B54A/2021 -> England_NORW-22B54A_2021\n",
      "England/NORW-22BD51/2021 -> England_NORW-22BD51_2021\n",
      "England/NORW-22CC17/2021 -> England_NORW-22CC17_2021\n",
      "England/NORW-21FC08/2021 -> England_NORW-21FC08_2021\n",
      "England/NORW-22BBA2/2021 -> England_NORW-22BBA2_2021\n",
      "England/NORW-22BA0F/2021 -> England_NORW-22BA0F_2021\n",
      "England/NORW-22B225/2021 -> England_NORW-22B225_2021\n",
      "England/NORW-22D539/2021 -> England_NORW-22D539_2021\n",
      "England/NORW-22BB39/2021 -> England_NORW-22BB39_2021\n",
      "England/NORW-22B216/2021 -> England_NORW-22B216_2021\n",
      "England/NORW-22BB0C/2021 -> England_NORW-22BB0C_2021\n",
      "England/NORW-22C1CD/2021 -> England_NORW-22C1CD_2021\n",
      "England/NORW-22BE30/2021 -> England_NORW-22BE30_2021\n",
      "England/NORW-22BDF7/2021 -> England_NORW-22BDF7_2021\n",
      "England/NORW-22BB93/2021 -> England_NORW-22BB93_2021\n",
      "England/NORW-22BE6D/2021 -> England_NORW-22BE6D_2021\n",
      "England/NORW-22B2CB/2021 -> England_NORW-22B2CB_2021\n",
      "England/NORW-22BCDC/2021 -> England_NORW-22BCDC_2021\n",
      "England/NORW-22CE5D/2021 -> England_NORW-22CE5D_2021\n",
      "England/NORW-22BFE2/2021 -> England_NORW-22BFE2_2021\n",
      "England/NORW-22C172/2021 -> England_NORW-22C172_2021\n",
      "England/NORW-22B902/2021 -> England_NORW-22B902_2021\n",
      "England/NORW-22BCCD/2021 -> England_NORW-22BCCD_2021\n",
      "England/NORW-22D469/2021 -> England_NORW-22D469_2021\n",
      "England/NORW-22BF79/2021 -> England_NORW-22BF79_2021\n",
      "England/NORW-22B9F3/2021 -> England_NORW-22B9F3_2021\n",
      "England/NORW-22BD8E/2021 -> England_NORW-22BD8E_2021\n",
      "England/NORW-22BDBB/2021 -> England_NORW-22BDBB_2021\n",
      "England/NORW-22CD14/2021 -> England_NORW-22CD14_2021\n",
      "England/NORW-22D593/2021 -> England_NORW-22D593_2021\n",
      "England/NORW-22BF6A/2021 -> England_NORW-22BF6A_2021\n",
      "England/NORW-22CD9C/2021 -> England_NORW-22CD9C_2021\n",
      "England/NORW-21FC62/2021 -> England_NORW-21FC62_2021\n",
      "England/NORW-22BD42/2021 -> England_NORW-22BD42_2021\n",
      "England/NORW-22CD41/2021 -> England_NORW-22CD41_2021\n",
      "England/NORW-22D548/2021 -> England_NORW-22D548_2021\n",
      "England/NORW-22D43C/2021 -> England_NORW-22D43C_2021\n",
      "England/NORW-22B42F/2021 -> England_NORW-22B42F_2021\n",
      "England/NORW-22BC45/2021 -> England_NORW-22BC45_2021\n",
      "England/NORW-22CDC9/2021 -> England_NORW-22CDC9_2021\n",
      "England/NORW-22C206/2021 -> England_NORW-22C206_2021\n",
      "England/NORW-22B2BC/2021 -> England_NORW-22B2BC_2021\n",
      "England/NORW-22BD7F/2021 -> England_NORW-22BD7F_2021\n",
      "England/NORW-22BEC7/2021 -> England_NORW-22BEC7_2021\n",
      "England/NORW-22BEA9/2021 -> England_NORW-22BEA9_2021\n",
      "England/NORW-22D487/2021 -> England_NORW-22D487_2021\n",
      "England/NORW-22BBB1/2021 -> England_NORW-22BBB1_2021\n",
      "England/NORW-22B9B7/2021 -> England_NORW-22B9B7_2021\n",
      "England/NORW-22CD6F/2021 -> England_NORW-22CD6F_2021\n",
      "England/NORW-22BE4F/2021 -> England_NORW-22BE4F_2021\n",
      "England/NORW-22CC53/2021 -> England_NORW-22CC53_2021\n",
      "England/NORW-22CD7E/2021 -> England_NORW-22CD7E_2021\n",
      "England/NORW-22B43E/2021 -> England_NORW-22B43E_2021\n",
      "England/NORW-22C145/2021 -> England_NORW-22C145_2021\n",
      "England/NORW-22B313/2021 -> England_NORW-22B313_2021\n",
      "England/NORW-22D29C/2021 -> England_NORW-22D29C_2021\n",
      "England/NORW-22BD15/2021 -> England_NORW-22BD15_2021\n",
      "England/NORW-22B2AD/2021 -> England_NORW-22B2AD_2021\n",
      "England/NORW-22BDD9/2021 -> England_NORW-22BDD9_2021\n",
      "England/NORW-22B52C/2021 -> England_NORW-22B52C_2021\n",
      "England/NORW-22B39B/2021 -> England_NORW-22B39B_2021\n",
      "England/NORW-22B243/2021 -> England_NORW-22B243_2021\n",
      "England/NORW-22B261/2021 -> England_NORW-22B261_2021\n",
      "England/NORW-22BC81/2021 -> England_NORW-22BC81_2021\n",
      "England/NORW-22BD06/2021 -> England_NORW-22BD06_2021\n",
      "England/NORW-22BAF0/2021 -> England_NORW-22BAF0_2021\n",
      "England/NORW-22BDE8/2021 -> England_NORW-22BDE8_2021\n",
      "England/NORW-22B45C/2021 -> England_NORW-22B45C_2021\n",
      "England/NORW-22BAD2/2021 -> England_NORW-22BAD2_2021\n",
      "England/NORW-22B51D/2021 -> England_NORW-22B51D_2021\n",
      "England/NORW-22C039/2021 -> England_NORW-22C039_2021\n",
      "England/NORW-22B252/2021 -> England_NORW-22B252_2021\n",
      "England/NORW-22BDAC/2021 -> England_NORW-22BDAC_2021\n",
      "England/NORW-22B35F/2021 -> England_NORW-22B35F_2021\n",
      "England/NORW-22C1DC/2021 -> England_NORW-22C1DC_2021\n",
      "England/NORW-22B3B9/2021 -> England_NORW-22B3B9_2021\n",
      "England/NORW-22BE9A/2021 -> England_NORW-22BE9A_2021\n",
      "England/NORW-22D496/2021 -> England_NORW-22D496_2021\n",
      "England/NORW-22BF5B/2021 -> England_NORW-22BF5B_2021\n",
      "England/NORW-22BE8B/2021 -> England_NORW-22BE8B_2021\n",
      "England/NORW-22C057/2021 -> England_NORW-22C057_2021\n",
      "England/NORW-22B9E4/2021 -> England_NORW-22B9E4_2021\n",
      "England/NORW-21FB29/2021 -> England_NORW-21FB29_2021\n",
      "England/NORW-22B2DA/2021 -> England_NORW-22B2DA_2021\n",
      "England/NORW-22BE7C/2021 -> England_NORW-22BE7C_2021\n",
      "England/NORW-22CB65/2021 -> England_NORW-22CB65_2021\n",
      "England/NORW-22BB75/2021 -> England_NORW-22BB75_2021\n",
      "England/NORW-22CC62/2021 -> England_NORW-22CC62_2021\n",
      "England/NORW-228F7C/2021 -> England_NORW-228F7C_2021\n",
      "England/NORW-229719/2021 -> England_NORW-229719_2021\n",
      "England/NORW-2273F9/2021 -> England_NORW-2273F9_2021\n",
      "England/NORW-22970A/2021 -> England_NORW-22970A_2021\n",
      "England/NORW-2297A0/2021 -> England_NORW-2297A0_2021\n",
      "England/NORW-22898D/2021 -> England_NORW-22898D_2021\n",
      "England/NORW-229C0B/2021 -> England_NORW-229C0B_2021\n",
      "England/NORW-2291DF/2021 -> England_NORW-2291DF_2021\n",
      "England/NORW-228747/2021 -> England_NORW-228747_2021\n",
      "England/NORW-228FE5/2021 -> England_NORW-228FE5_2021\n",
      "England/NORW-228DDC/2021 -> England_NORW-228DDC_2021\n",
      "England/NORW-22A077/2021 -> England_NORW-22A077_2021\n",
      "England/NORW-22899C/2021 -> England_NORW-22899C_2021\n",
      "England/NORW-227E07/2021 -> England_NORW-227E07_2021\n",
      "England/NORW-229281/2021 -> England_NORW-229281_2021\n",
      "England/NORW-22A772/2021 -> England_NORW-22A772_2021\n",
      "England/NORW-2245F6/2021 -> England_NORW-2245F6_2021\n",
      "England/NORW-22A45D/2021 -> England_NORW-22A45D_2021\n",
      "England/NORW-229597/2020 -> England_NORW-229597_2020\n",
      "England/NORW-2284F5/2021 -> England_NORW-2284F5_2021\n",
      "England/NORW-229D08/2021 -> England_NORW-229D08_2021\n",
      "England/NORW-227265/2021 -> England_NORW-227265_2021\n",
      "England/NORW-228176/2021 -> England_NORW-228176_2021\n",
      "England/NORW-22A96D/2021 -> England_NORW-22A96D_2021\n",
      "England/NORW-229139/2021 -> England_NORW-229139_2021\n",
      "England/NORW-2283E9/2020 -> England_NORW-2283E9_2020\n",
      "England/NORW-2294A9/2021 -> England_NORW-2294A9_2021\n",
      "England/NORW-2295C4/2021 -> England_NORW-2295C4_2021\n",
      "England/NORW-22AAD3/2021 -> England_NORW-22AAD3_2021\n",
      "England/NORW-22951F/2021 -> England_NORW-22951F_2021\n",
      "England/NORW-22A332/2021 -> England_NORW-22A332_2021\n",
      "England/NORW-22897E/2021 -> England_NORW-22897E_2021\n",
      "England/NORW-22A97C/2021 -> England_NORW-22A97C_2021\n",
      "England/NORW-229B1D/2021 -> England_NORW-229B1D_2021\n",
      "England/NORW-228B0F/2021 -> England_NORW-228B0F_2021\n",
      "England/NORW-228307/2021 -> England_NORW-228307_2021\n",
      "England/NORW-227F7D/2021 -> England_NORW-227F7D_2021\n",
      "England/NORW-228000/2021 -> England_NORW-228000_2021\n",
      "England/NORW-2245E7/2021 -> England_NORW-2245E7_2021\n",
      "England/NORW-229B86/2020 -> England_NORW-229B86_2020\n",
      "England/NORW-228B87/2021 -> England_NORW-228B87_2021\n",
      "England/NORW-224611/2021 -> England_NORW-224611_2021\n",
      "England/NORW-229096/2021 -> England_NORW-229096_2021\n",
      "England/NORW-227FE6/2021 -> England_NORW-227FE6_2021\n",
      "England/NORW-2292CD/2021 -> England_NORW-2292CD_2021\n",
      "England/NORW-228DFA/2021 -> England_NORW-228DFA_2021\n",
      "England/NORW-22911B/2021 -> England_NORW-22911B_2021\n",
      "England/NORW-22A253/2021 -> England_NORW-22A253_2021\n",
      "England/NORW-22A9F4/2021 -> England_NORW-22A9F4_2021\n",
      "England/NORW-228871/2021 -> England_NORW-228871_2021\n",
      "England/NORW-22862C/2021 -> England_NORW-22862C_2021\n",
      "England/NORW-2296C1/2021 -> England_NORW-2296C1_2021\n",
      "England/NORW-22756C/2021 -> England_NORW-22756C_2021\n",
      "England/NORW-22960D/2021 -> England_NORW-22960D_2021\n",
      "England/NORW-227F8C/2021 -> England_NORW-227F8C_2021\n",
      "England/NORW-22961C/2021 -> England_NORW-22961C_2021\n",
      "England/NORW-228E15/2021 -> England_NORW-228E15_2021\n",
      "England/NORW-229A01/2020 -> England_NORW-229A01_2020\n",
      "England/NORW-228CEE/2021 -> England_NORW-228CEE_2021\n",
      "England/NORW-229658/2021 -> England_NORW-229658_2021\n",
      "England/NORW-227247/2021 -> England_NORW-227247_2021\n",
      "England/NORW-22AEAA/2021 -> England_NORW-22AEAA_2021\n",
      "England/NORW-227F31/2021 -> England_NORW-227F31_2021\n",
      "England/NORW-2291A2/2021 -> England_NORW-2291A2_2021\n",
      "England/NORW-228352/2021 -> England_NORW-228352_2021\n",
      "England/NORW-227FB9/2021 -> England_NORW-227FB9_2021\n",
      "England/NORW-229AF2/2020 -> England_NORW-229AF2_2020\n",
      "England/NORW-228B78/2021 -> England_NORW-228B78_2021\n",
      "England/NORW-22810D/2020 -> England_NORW-22810D_2020\n",
      "England/NORW-228C57/2021 -> England_NORW-228C57_2021\n",
      "\n",
      "                          Gap/Ambiguity  Composition  p-value\n",
      "   1  England_NORW-22D223_2021    1.76%    passed    100.00%\n",
      "   2  England_NORW-22D1CC_2021    1.76%    passed    100.00%\n",
      "   3  England_NORW-22D162_2021    9.55%    passed     98.45%\n",
      "   4  England_NORW-22B37D_2021    1.76%    passed     99.99%\n",
      "   5  England_NORW-22D2C9_2021    1.76%    passed     99.95%\n",
      "   6  England_NORW-22D1BD_2021    1.76%    passed     99.96%\n",
      "   7  England_NORW-22D36C_2021    1.76%    passed     99.97%\n",
      "   8  England_NORW-22463F_2021    1.76%    passed    100.00%\n",
      "   9  England_NORW-22465D_2021    1.76%    passed     99.99%\n",
      "  10  England_NORW-22D205_2021    1.76%    passed    100.00%\n",
      "  11  England_NORW-22B4C5_2021    1.69%    passed     99.97%\n",
      "  12  England_NORW-22D27E_2021    1.76%    passed     99.98%\n",
      "  13  England_NORW-22D180_2021    2.81%    passed     99.85%\n",
      "  14  England_NORW-22466C_2021    9.70%    passed     99.78%\n",
      "  15  England_NORW-22464E_2021    1.76%    passed     99.98%\n",
      "  16  England_NORW-22B085_2021    1.76%    passed     99.97%\n",
      "  17  England_NORW-22D302_2021    1.93%    passed    100.00%\n",
      "  18  England_NORW-22D1AE_2021    2.83%    passed     99.86%\n",
      "  19  England_NORW-22D34E_2021    1.76%    passed     99.98%\n",
      "  20  England_NORW-22B00D_2021    1.76%    passed     99.97%\n",
      "  21  England_NORW-22B304_2021    2.21%    passed     99.94%\n",
      "  22  England_NORW-22D214_2021    7.70%    passed     96.88%\n",
      "  23  England_NORW-22D250_2021    2.18%    passed     99.98%\n",
      "  24  England_NORW-22468A_2021    1.76%    passed     99.99%\n",
      "  25  England_NORW-22D171_2021    1.76%    passed    100.00%\n",
      "  26  England_NORW-22D26F_2021    1.76%    passed     99.97%\n",
      "  27  England_NORW-22D0DE_2021    1.76%    passed    100.00%\n",
      "  28  England_NORW-22D3D5_2021    2.31%    passed     99.90%\n",
      "  29  England_NORW-22D2F6_2021    1.76%    passed    100.00%\n",
      "  30  England_NORW-22B2F8_2021    1.69%    passed     99.98%\n",
      "  31  England_NORW-22D2BA_2021    1.76%    passed     99.98%\n",
      "  32  England_NORW-22AF4D_2021    8.29%    passed     99.69%\n",
      "  33  England_NORW-22D19F_2021    8.21%    passed     98.30%\n",
      "  34  England_NORW-22D108_2021    1.76%    passed     99.99%\n",
      "  35  England_NORW-22D38A_2021    1.76%    passed     99.99%\n",
      "  36  England_NORW-22D2E7_2021    4.04%    passed     99.98%\n",
      "  37  England_NORW-22D1F9_2021    8.97%    passed     97.81%\n",
      "  38  England_NORW-22D3B7_2021    1.76%    passed     99.99%\n",
      "  39  England_NORW-22D399_2021    8.93%    passed     98.92%\n",
      "  40  England_NORW-22B3E6_2021    1.78%    passed     99.98%\n",
      "  41  England_NORW-22B8AB_2021    1.76%    passed     99.98%\n",
      "  42  England_NORW-22B38C_2021    1.76%    passed    100.00%\n",
      "  43  England_NORW-22B54A_2021    1.76%    passed     99.99%\n",
      "  44  England_NORW-22BD51_2021    5.33%    passed     99.07%\n",
      "  45  England_NORW-22CC17_2021    1.76%    passed     99.98%\n",
      "  46  England_NORW-21FC08_2021    1.76%    passed     99.99%\n",
      "  47  England_NORW-22BBA2_2021    1.76%    passed     99.99%\n",
      "  48  England_NORW-22BA0F_2021    2.08%    passed     99.97%\n",
      "  49  England_NORW-22B225_2021    2.05%    passed    100.00%\n",
      "  50  England_NORW-22D539_2021    7.26%    passed     99.58%\n",
      "  51  England_NORW-22BB39_2021    1.77%    passed     99.97%\n",
      "  52  England_NORW-22B216_2021    1.76%    passed     99.98%\n",
      "  53  England_NORW-22BB0C_2021    1.76%    passed     99.98%\n",
      "  54  England_NORW-22C1CD_2021    1.76%    passed     99.99%\n",
      "  55  England_NORW-22BE30_2021    1.76%    passed     99.96%\n",
      "  56  England_NORW-22BDF7_2021    1.76%    passed     99.99%\n",
      "  57  England_NORW-22BB93_2021    1.76%    passed     99.99%\n",
      "  58  England_NORW-22BE6D_2021    1.80%    passed     99.97%\n",
      "  59  England_NORW-22B2CB_2021    1.76%    passed     99.99%\n",
      "  60  England_NORW-22BCDC_2021    2.35%    passed     99.84%\n",
      "  61  England_NORW-22CE5D_2021    1.76%    passed     99.98%\n",
      "  62  England_NORW-22BFE2_2021    1.76%    passed     99.94%\n",
      "  63  England_NORW-22C172_2021    2.05%    passed     99.99%\n",
      "  64  England_NORW-22B902_2021    1.76%    passed     99.96%\n",
      "  65  England_NORW-22BCCD_2021    1.76%    passed     99.96%\n",
      "  66  England_NORW-22D469_2021    1.76%    passed     99.99%\n",
      "  67  England_NORW-22BF79_2021    7.57%    passed     99.11%\n",
      "  68  England_NORW-22B9F3_2021    1.76%    passed     99.99%\n",
      "  69  England_NORW-22BD8E_2021    1.76%    passed     99.99%\n",
      "  70  England_NORW-22BDBB_2021    5.68%    passed     99.73%\n",
      "  71  England_NORW-22CD14_2021    1.76%    passed     99.97%\n",
      "  72  England_NORW-22D593_2021    2.67%    passed     99.63%\n",
      "  73  England_NORW-22BF6A_2021    2.77%    passed     99.88%\n",
      "  74  England_NORW-22CD9C_2021    1.76%    passed     99.99%\n",
      "  75  England_NORW-21FC62_2021    4.42%    passed     99.80%\n",
      "  76  England_NORW-22BD42_2021    1.76%    passed     99.98%\n",
      "  77  England_NORW-22CD41_2021    1.76%    passed     99.98%\n",
      "  78  England_NORW-22D548_2021    1.76%    passed     99.97%\n",
      "  79  England_NORW-22D43C_2021    4.60%    passed     98.04%\n",
      "  80  England_NORW-22B42F_2021    1.80%    passed     99.97%\n",
      "  81  England_NORW-22BC45_2021    3.74%    passed     99.46%\n",
      "  82  England_NORW-22CDC9_2021    2.31%    passed     99.99%\n",
      "  83  England_NORW-22C206_2021    1.81%    passed     99.98%\n",
      "  84  England_NORW-22B2BC_2021    1.80%    passed     99.97%\n",
      "  85  England_NORW-22BD7F_2021    1.76%    passed     99.96%\n",
      "  86  England_NORW-22BEC7_2021    1.76%    passed     99.98%\n",
      "  87  England_NORW-22BEA9_2021    7.86%    passed     99.52%\n",
      "  88  England_NORW-22D487_2021    2.45%    passed     99.98%\n",
      "  89  England_NORW-22BBB1_2021    1.76%    passed     99.97%\n",
      "  90  England_NORW-22B9B7_2021    3.80%    passed     99.87%\n",
      "  91  England_NORW-22CD6F_2021    2.23%    passed    100.00%\n",
      "  92  England_NORW-22BE4F_2021    1.69%    passed     99.91%\n",
      "  93  England_NORW-22CC53_2021    1.76%    passed     99.99%\n",
      "  94  England_NORW-22CD7E_2021    1.85%    passed     99.95%\n",
      "  95  England_NORW-22B43E_2021    1.80%    passed     99.98%\n",
      "  96  England_NORW-22C145_2021    2.19%    passed     99.86%\n",
      "  97  England_NORW-22B313_2021    1.76%    passed     99.99%\n",
      "  98  England_NORW-22D29C_2021    7.52%    passed     98.58%\n",
      "  99  England_NORW-22BD15_2021    1.89%    passed    100.00%\n",
      " 100  England_NORW-22B2AD_2021    1.76%    passed     99.98%\n",
      " 101  England_NORW-22BDD9_2021    4.64%    passed     99.89%\n",
      " 102  England_NORW-22B52C_2021    1.76%    passed     99.97%\n",
      " 103  England_NORW-22B39B_2021    5.44%    passed     96.36%\n",
      " 104  England_NORW-22B243_2021    1.76%    passed     99.98%\n",
      " 105  England_NORW-22B261_2021    1.77%    passed     99.97%\n",
      " 106  England_NORW-22BC81_2021    1.76%    passed    100.00%\n",
      " 107  England_NORW-22BD06_2021    1.91%    passed    100.00%\n",
      " 108  England_NORW-22BAF0_2021    1.76%    passed     99.96%\n",
      " 109  England_NORW-22BDE8_2021    1.76%    passed     99.99%\n",
      " 110  England_NORW-22B45C_2021    1.76%    passed     99.99%\n",
      " 111  England_NORW-22BAD2_2021    1.76%    passed     99.95%\n",
      " 112  England_NORW-22B51D_2021    1.69%    passed     99.93%\n",
      " 113  England_NORW-22C039_2021    1.82%    passed    100.00%\n",
      " 114  England_NORW-22B252_2021    2.51%    passed     99.98%\n",
      " 115  England_NORW-22BDAC_2021    1.76%    passed     99.97%\n",
      " 116  England_NORW-22B35F_2021    3.27%    passed     99.87%\n",
      " 117  England_NORW-22C1DC_2021    1.76%    passed     99.93%\n",
      " 118  England_NORW-22B3B9_2021    1.76%    passed     99.95%\n",
      " 119  England_NORW-22BE9A_2021    1.76%    passed     99.96%\n",
      " 120  England_NORW-22D496_2021    2.19%    passed     99.96%\n",
      " 121  England_NORW-22BF5B_2021    1.69%    passed     99.96%\n",
      " 122  England_NORW-22BE8B_2021    9.51%    passed     98.79%\n",
      " 123  England_NORW-22C057_2021    1.76%    passed     99.97%\n",
      " 124  England_NORW-22B9E4_2021    1.76%    passed     99.95%\n",
      " 125  England_NORW-21FB29_2021    1.76%    passed     99.99%\n",
      " 126  England_NORW-22B2DA_2021    1.76%    passed     99.97%\n",
      " 127  England_NORW-22BE7C_2021    1.89%    passed     99.99%\n",
      " 128  England_NORW-22CB65_2021    1.76%    passed     99.98%\n",
      " 129  England_NORW-22BB75_2021    1.76%    passed     99.95%\n",
      " 130  England_NORW-22CC62_2021    1.76%    passed     99.99%\n",
      " 131  England_NORW-228F7C_2021    1.99%    passed     99.99%\n",
      " 132  England_NORW-229719_2021    4.18%    passed     99.42%\n",
      " 133  England_NORW-2273F9_2021    1.80%    passed     99.99%\n",
      " 134  England_NORW-22970A_2021    5.20%    passed     97.74%\n",
      " 135  England_NORW-2297A0_2021    1.75%    passed     99.91%\n",
      " 136  England_NORW-22898D_2021    3.24%    passed     99.99%\n",
      " 137  England_NORW-229C0B_2021    1.76%    passed     99.95%\n",
      " 138  England_NORW-2291DF_2021    2.51%    passed     99.98%\n",
      " 139  England_NORW-228747_2021    1.76%    passed     99.93%\n",
      " 140  England_NORW-228FE5_2021    3.76%    passed     99.87%\n",
      " 141  England_NORW-228DDC_2021    1.88%    passed    100.00%\n",
      " 142  England_NORW-22A077_2021    7.18%    passed     97.49%\n",
      " 143  England_NORW-22899C_2021    1.76%    passed     99.99%\n",
      " 144  England_NORW-227E07_2021    1.76%    passed     99.96%\n",
      " 145  England_NORW-229281_2021    3.46%    passed     99.99%\n",
      " 146  England_NORW-22A772_2021    1.79%    passed     99.98%\n",
      " 147  England_NORW-2245F6_2021    1.76%    passed     99.99%\n",
      " 148  England_NORW-22A45D_2021    6.30%    passed     97.87%\n",
      " 149  England_NORW-229597_2020    1.69%    passed     99.89%\n",
      " 150  England_NORW-2284F5_2021    1.70%    passed     99.94%\n",
      " 151  England_NORW-229D08_2021    8.45%    passed     93.69%\n",
      " 152  England_NORW-227265_2021    2.48%    passed     99.97%\n",
      " 153  England_NORW-228176_2021    3.28%    passed     99.91%\n",
      " 154  England_NORW-22A96D_2021    4.73%    passed     98.78%\n",
      " 155  England_NORW-229139_2021    1.76%    passed     99.98%\n",
      " 156  England_NORW-2283E9_2020    1.69%    passed     99.97%\n",
      " 157  England_NORW-2294A9_2021    1.76%    passed     99.97%\n",
      " 158  England_NORW-2295C4_2021    8.81%    passed     88.81%\n",
      " 159  England_NORW-22AAD3_2021    1.79%    passed     99.98%\n",
      " 160  England_NORW-22951F_2021    2.86%    passed     99.98%\n",
      " 161  England_NORW-22A332_2021    2.41%    passed     99.99%\n",
      " 162  England_NORW-22897E_2021    1.88%    passed    100.00%\n",
      " 163  England_NORW-22A97C_2021    1.78%    passed     99.95%\n",
      " 164  England_NORW-229B1D_2021    2.00%    passed    100.00%\n",
      " 165  England_NORW-228B0F_2021    1.69%    passed     99.95%\n",
      " 166  England_NORW-228307_2021    1.76%    passed     99.98%\n",
      " 167  England_NORW-227F7D_2021    9.58%    passed     92.08%\n",
      " 168  England_NORW-228000_2021    1.76%    passed     99.94%\n",
      " 169  England_NORW-2245E7_2021    1.76%    passed     99.99%\n",
      " 170  England_NORW-229B86_2020    4.38%    passed     99.52%\n",
      " 171  England_NORW-228B87_2021    1.76%    passed     99.97%\n",
      " 172  England_NORW-224611_2021    1.76%    passed     99.99%\n",
      " 173  England_NORW-229096_2021    2.94%    passed     99.94%\n",
      " 174  England_NORW-227FE6_2021    2.97%    passed     99.99%\n",
      " 175  England_NORW-2292CD_2021    1.69%    passed     99.89%\n",
      " 176  England_NORW-228DFA_2021    2.06%    passed     99.98%\n",
      " 177  England_NORW-22911B_2021    4.61%    passed     99.04%\n",
      " 178  England_NORW-22A253_2021    1.83%    passed     99.92%\n",
      " 179  England_NORW-22A9F4_2021    1.76%    passed     99.99%\n",
      " 180  England_NORW-228871_2021    2.76%    passed     99.97%\n",
      " 181  England_NORW-22862C_2021    1.80%    passed     99.97%\n",
      " 182  England_NORW-2296C1_2021    1.70%    passed     99.91%\n",
      " 183  England_NORW-22756C_2021    4.80%    passed     98.49%\n",
      " 184  England_NORW-22960D_2021    2.02%    passed    100.00%\n",
      " 185  England_NORW-227F8C_2021    1.76%    passed     99.97%\n",
      " 186  England_NORW-22961C_2021    1.76%    passed     99.99%\n",
      " 187  England_NORW-228E15_2021    5.16%    passed     98.32%\n",
      " 188  England_NORW-229A01_2020    6.29%    passed     97.71%\n",
      " 189  England_NORW-228CEE_2021    2.59%    passed     99.92%\n",
      " 190  England_NORW-229658_2021    1.78%    passed     99.97%\n",
      " 191  England_NORW-227247_2021    1.76%    passed     99.98%\n",
      " 192  England_NORW-22AEAA_2021    1.76%    passed     99.99%\n",
      " 193  England_NORW-227F31_2021    2.70%    passed     99.96%\n",
      " 194  England_NORW-2291A2_2021    3.31%    passed     99.98%\n",
      " 195  England_NORW-228352_2021    1.69%    passed     99.91%\n",
      " 196  England_NORW-227FB9_2021    3.68%    passed     99.92%\n",
      " 197  England_NORW-229AF2_2020    2.74%    passed     99.20%\n",
      " 198  England_NORW-228B78_2021    6.86%    passed     96.50%\n",
      " 199  England_NORW-22810D_2020    2.70%    passed     99.96%\n",
      " 200  England_NORW-228C57_2021    9.05%    passed     95.69%\n",
      "****  TOTAL                       2.80%  0 sequences failed composition chi2 test (p-value<5%; df=3)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOTE: England_NORW-22D0DE_2021 is identical to England_NORW-22D205_2021 but kept for subsequent analysis\n",
      "NOTE: England_NORW-22D2BA_2021 is identical to England_NORW-22D27E_2021 but kept for subsequent analysis\n",
      "NOTE: England_NORW-22BEC7_2021 is identical to England_NORW-22B216_2021 but kept for subsequent analysis\n",
      "NOTE: England_NORW-22899C_2021 is identical to England_NORW-22C1CD_2021 but kept for subsequent analysis\n",
      "NOTE: England_NORW-22B9F3_2021 is identical to England_NORW-22BDF7_2021 but kept for subsequent analysis\n",
      "NOTE: England_NORW-22B42F_2021 is identical to England_NORW-22BE6D_2021 but kept for subsequent analysis\n",
      "NOTE: England_NORW-22BCCD_2021 is identical to England_NORW-22B902_2021 but kept for subsequent analysis\n",
      "NOTE: England_NORW-22BE9A_2021 is identical to England_NORW-22BD7F_2021 but kept for subsequent analysis\n",
      "NOTE: England_NORW-22B2DA_2021 is identical to England_NORW-22C057_2021 but kept for subsequent analysis\n",
      "NOTE: England_NORW-224611_2021 is identical to England_NORW-2245F6_2021 but kept for subsequent analysis\n",
      "NOTE: England_NORW-227F8C_2021 is identical to England_NORW-228B87_2021 but kept for subsequent analysis\n",
      "NOTE: 4 identical sequences (see below) will be ignored for subsequent analysis\n",
      "NOTE: England_NORW-22961C_2021 (identical to England_NORW-22C1CD_2021) is ignored but added at the end\n",
      "NOTE: England_NORW-22CC53_2021 (identical to England_NORW-22BDF7_2021) is ignored but added at the end\n",
      "NOTE: England_NORW-22B2BC_2021 (identical to England_NORW-22BE6D_2021) is ignored but added at the end\n",
      "NOTE: England_NORW-22862C_2021 (identical to England_NORW-22BE6D_2021) is ignored but added at the end\n",
      "\n",
      "For your convenience alignment with unique sequences printed to ../data/COG-UK-NORW/results/20220210/NORW_200_aln.fa.uniqueseq.phy\n",
      "\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.041 seconds\n",
      "NOTE: ModelFinder requires 99 MB RAM!\n",
      "ModelFinder will test 286 DNA models (sample size: 29914) ...\n",
      " No. Model         -LnL         df  AIC          AICc         BIC\n",
      "  1  JC            45617.347    389 92012.694    92022.971    95243.760\n",
      "  2  JC+I          45592.296    390 91964.593    91974.923    95203.965\n",
      "  3  JC+G4         45595.583    390 91971.166    91981.496    95210.538\n",
      "  4  JC+I+G4       45592.322    391 91966.645    91977.029    95214.323\n",
      "  5  JC+R2         45610.036    391 92002.071    92012.455    95249.749\n",
      "  6  JC+R3         45608.040    393 92002.079    92012.570    95266.370\n",
      " 14  F81+F         44676.129    392 90136.259    90146.696    93392.243\n",
      " 15  F81+F+I       44653.068    393 90092.136    90102.626    93356.426\n",
      " 16  F81+F+G4      44655.387    393 90096.775    90107.266    93361.065\n",
      " 17  F81+F+I+G4    44653.125    394 90094.250    90104.795    93366.846\n",
      " 18  F81+F+R2      44669.049    394 90126.097    90136.642    93398.693\n",
      " 19  F81+F+R3      44667.076    396 90126.151    90136.803    93415.360\n",
      " 27  K2P           45434.096    390 91648.191    91658.522    94887.563\n",
      " 28  K2P+I         45408.957    391 91599.914    91610.297    94847.592\n",
      " 29  K2P+G4        45412.285    391 91606.570    91616.954    94854.248\n",
      " 30  K2P+I+G4      45408.981    392 91601.962    91612.399    94857.946\n",
      " 31  K2P+R2        45426.887    392 91637.774    91648.211    94893.758\n",
      " 32  K2P+R3        45424.780    394 91637.559    91648.104    94910.155\n",
      " 40  HKY+F         44480.713    393 89747.425    89757.916    93011.715\n",
      " 41  HKY+F+I       44459.889    394 89707.777    89718.322    92980.373\n",
      " 42  HKY+F+G4      44461.271    394 89710.543    89721.087    92983.139\n",
      " 43  HKY+F+I+G4    44459.973    395 89709.945    89720.544    92990.848\n",
      " 44  HKY+F+R2      44473.949    395 89737.898    89748.496    93018.800\n",
      " 45  HKY+F+R3      44471.950    397 89737.900    89748.607    93035.415\n",
      " 53  TNe           45383.507    391 91549.014    91559.398    94796.692\n",
      " 54  TNe+I         45360.374    392 91504.748    91515.185    94760.733\n",
      " 55  TNe+G4        45362.643    392 91509.285    91519.722    94765.269\n",
      " 56  TNe+I+G4      45360.411    393 91506.822    91517.313    94771.112\n",
      " 57  TNe+R2        45376.527    393 91539.053    91549.544    94803.343\n",
      " 58  TNe+R3        45374.399    395 91538.799    91549.397    94819.701\n",
      " 66  TN+F          44428.809    394 89645.618    89656.162    92918.214\n",
      " 67  TN+F+I        44410.984    395 89611.968    89622.566    92892.870\n",
      " 68  TN+F+G4       44411.416    395 89612.832    89623.430    92893.735\n",
      " 69  TN+F+I+G4     44411.068    396 89614.136    89624.789    92903.345\n",
      " 70  TN+F+R2       44422.431    396 89636.862    89647.514    92926.070\n",
      " 71  TN+F+R3       44420.458    398 89636.916    89647.677    92942.737\n",
      " 79  K3P           45428.327    391 91638.655    91649.039    94886.333\n",
      " 80  K3P+I         45403.185    392 91590.370    91600.807    94846.354\n",
      " 81  K3P+G4        45406.512    392 91597.024    91607.462    94853.009\n",
      " 82  K3P+I+G4      45403.197    393 91592.394    91602.884    94856.684\n",
      " 83  K3P+R2        45421.331    393 91628.661    91639.152    94892.951\n",
      " 84  K3P+R3        45419.034    395 91628.068    91638.666    94908.970\n",
      " 92  K3Pu+F        44473.067    394 89734.134    89744.679    93006.730\n",
      " 93  K3Pu+F+I      44452.456    395 89694.913    89705.511    92975.815\n",
      " 94  K3Pu+F+G4     44453.755    395 89697.509    89708.107    92978.412\n",
      " 95  K3Pu+F+I+G4   44452.542    396 89697.084    89707.737    92986.293\n",
      " 96  K3Pu+F+R2     44466.523    396 89725.046    89735.698    93014.254\n",
      " 97  K3Pu+F+R3     44464.335    398 89724.670    89735.431    93030.491\n",
      "105  TPM2+F        44472.016    394 89732.032    89742.576    93004.628\n",
      "106  TPM2+F+I      44451.141    395 89692.282    89702.880    92973.185\n",
      "107  TPM2+F+G4     44452.531    395 89695.062    89705.660    92975.964\n",
      "108  TPM2+F+I+G4   44451.221    396 89694.443    89705.095    92983.651\n",
      "109  TPM2+F+R2     44465.457    396 89722.913    89733.565    93012.121\n",
      "110  TPM2+F+R3     44463.205    398 89722.411    89733.171    93028.231\n",
      "118  TPM2u+F       44471.863    394 89731.726    89742.270    93004.322\n",
      "119  TPM2u+F+I     44450.987    395 89691.975    89702.573    92972.877\n",
      "120  TPM2u+F+G4    44452.381    395 89694.762    89705.360    92975.664\n",
      "121  TPM2u+F+I+G4  44451.071    396 89694.142    89704.795    92983.351\n",
      "122  TPM2u+F+R2    44465.363    396 89722.726    89733.378    93011.934\n",
      "123  TPM2u+F+R3    44463.081    398 89722.162    89732.923    93027.983\n",
      "131  TPM3+F        44479.714    394 89747.427    89757.972    93020.024\n",
      "132  TPM3+F+I      44458.622    395 89707.243    89717.841    92988.145\n",
      "133  TPM3+F+G4     44460.109    395 89710.219    89720.817    92991.121\n",
      "134  TPM3+F+I+G4   44458.703    396 89709.407    89720.059    92998.615\n",
      "135  TPM3+F+R2     44473.247    396 89738.495    89749.147    93027.703\n",
      "136  TPM3+F+R3     44470.916    398 89737.831    89748.592    93043.652\n",
      "144  TPM3u+F       44479.632    394 89747.264    89757.809    93019.861\n",
      "145  TPM3u+F+I     44458.547    395 89707.094    89717.692    92987.996\n",
      "146  TPM3u+F+G4    44460.037    395 89710.073    89720.672    92990.976\n",
      "147  TPM3u+F+I+G4  44458.633    396 89709.266    89719.918    92998.475\n",
      "148  TPM3u+F+R2    44473.215    396 89738.430    89749.082    93027.638\n",
      "149  TPM3u+F+R3    44470.828    398 89737.657    89748.417    93043.477\n",
      "157  TIMe          45377.488    392 91538.976    91549.413    94794.960\n",
      "158  TIMe+I        45354.385    393 91494.771    91505.261    94759.061\n",
      "159  TIMe+G4       45356.656    393 91499.311    91509.802    94763.601\n",
      "160  TIMe+I+G4     45354.421    394 91496.842    91507.387    94769.438\n",
      "161  TIMe+R2       45370.844    394 91529.689    91540.233    94802.285\n",
      "162  TIMe+R3       45368.280    396 91528.561    91539.213    94817.769\n",
      "170  TIM+F         44420.850    395 89631.699    89642.298    92912.602\n",
      "171  TIM+F+I       44403.219    396 89598.438    89609.090    92887.646\n",
      "172  TIM+F+G4      44403.606    396 89599.213    89609.865    92888.421\n",
      "173  TIM+F+I+G4    44403.303    397 89600.606    89611.312    92898.120\n",
      "174  TIM+F+R2      44414.807    397 89623.613    89634.320    92921.128\n",
      "175  TIM+F+R3      44412.488    399 89622.977    89633.792    92937.103\n",
      "183  TIM2e         45377.947    392 91539.894    91550.331    94795.878\n",
      "184  TIM2e+I       45354.983    393 91495.965    91506.456    94760.256\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "185  TIM2e+G4      45357.188    393 91500.377    91510.867    94764.667\n",
      "186  TIM2e+I+G4    45355.018    394 91498.035    91508.580    94770.631\n",
      "187  TIM2e+R2      45371.436    394 91530.871    91541.416    94803.468\n",
      "188  TIM2e+R3      45368.796    396 91529.592    91540.244    94818.800\n",
      "196  TIM2+F        44419.918    395 89629.836    89640.435    92910.739\n",
      "197  TIM2+F+I      44402.174    396 89596.348    89607.001    92885.557\n",
      "198  TIM2+F+G4     44402.584    396 89597.167    89607.819    92886.375\n",
      "199  TIM2+F+I+G4   44402.245    397 89598.491    89609.197    92896.005\n",
      "200  TIM2+F+R2     44413.927    397 89621.855    89632.561    92919.369\n",
      "201  TIM2+F+R3     44411.473    399 89620.947    89631.762    92935.073\n",
      "209  TIM3e         45380.630    392 91545.261    91555.698    94801.245\n",
      "210  TIM3e+I       45356.680    393 91499.361    91509.851    94763.651\n",
      "211  TIM3e+G4      45359.357    393 91504.714    91515.204    94769.004\n",
      "212  TIM3e+I+G4    45356.687    394 91501.375    91511.919    94773.971\n",
      "213  TIM3e+R2      45374.106    394 91536.211    91546.756    94808.807\n",
      "214  TIM3e+R3      45371.223    396 91534.446    91545.098    94823.654\n",
      "222  TIM3+F        44427.584    395 89645.169    89655.767    92926.071\n",
      "223  TIM3+F+I      44409.655    396 89611.310    89621.962    92900.519\n",
      "224  TIM3+F+G4     44410.115    396 89612.230    89622.883    92901.439\n",
      "225  TIM3+F+I+G4   44409.735    397 89613.469    89624.176    92910.984\n",
      "226  TIM3+F+R2     44421.634    397 89637.268    89647.975    92934.783\n",
      "227  TIM3+F+R3     44419.001    399 89636.002    89646.817    92950.129\n",
      "235  TVMe          45423.005    393 91632.011    91642.502    94896.301\n",
      "236  TVMe+I        45396.943    394 91581.887    91592.431    94854.483\n",
      "237  TVMe+G4       45400.788    394 91589.576    91600.121    94862.172\n",
      "238  TVMe+I+G4     45396.955    395 91583.910    91594.509    94864.813\n",
      "239  TVMe+R2       45416.451    395 91622.902    91633.500    94903.804\n",
      "240  TVMe+R3       45413.357    397 91620.714    91631.421    94918.229\n",
      "248  TVM+F         44467.323    396 89726.646    89737.298    93015.855\n",
      "249  TVM+F+I       44446.126    397 89686.251    89696.958    92983.766\n",
      "250  TVM+F+G4      44447.644    397 89689.288    89699.994    92986.802\n",
      "251  TVM+F+I+G4    44446.202    398 89688.404    89699.165    92994.224\n",
      "252  TVM+F+R2      44461.150    398 89718.300    89729.060    93024.120\n",
      "253  TVM+F+R3      44458.212    400 89716.425    89727.295    93038.858\n",
      "261  SYM           45372.614    394 91533.229    91543.773    94805.825\n",
      "262  SYM+I         45348.518    395 91487.036    91497.635    94767.939\n",
      "263  SYM+G4        45351.264    395 91492.528    91503.126    94773.430\n",
      "264  SYM+I+G4      45348.540    396 91489.079    91499.732    94778.288\n",
      "265  SYM+R2        45366.210    396 91524.421    91535.073    94813.629\n",
      "266  SYM+R3        45363.032    398 91522.064    91532.825    94827.884\n",
      "274  GTR+F         44415.591    397 89625.183    89635.889    92922.697\n",
      "275  GTR+F+I       44397.598    398 89591.195    89601.956    92897.016\n",
      "276  GTR+F+G4      44398.071    398 89592.142    89602.903    92897.963\n",
      "277  GTR+F+I+G4    44397.681    399 89593.363    89604.178    92907.490\n",
      "278  GTR+F+R2      44409.803    399 89617.605    89628.421    92931.732\n",
      "279  GTR+F+R3      44406.906    401 89615.812    89626.737    92946.551\n",
      "Akaike Information Criterion:           GTR+F+I\n",
      "Corrected Akaike Information Criterion: GTR+F+I\n",
      "Bayesian Information Criterion:         TIM2+F+I\n",
      "Best-fit model: TIM2+F+I chosen according to BIC\n",
      "\n",
      "All model information printed to ../data/COG-UK-NORW/results/20220210/NORW_200_aln.fa.model.gz\n",
      "CPU time for ModelFinder: 63.246 seconds (0h:1m:3s)\n",
      "Wall-clock time for ModelFinder: 64.067 seconds (0h:1m:4s)\n",
      "\n",
      "NOTE: 9 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -44662.703\n",
      "2. Current log-likelihood: -44402.035\n",
      "Optimal log-likelihood: -44402.033\n",
      "Rate parameters:  A-C: 0.43457  A-G: 2.21760  A-T: 0.43457  C-G: 1.00000  C-T: 7.31249  G-T: 1.00000\n",
      "Base frequencies:  A: 0.299  C: 0.183  G: 0.196  T: 0.322\n",
      "Proportion of invariable sites: 0.795\n",
      "Parameters optimization took 2 rounds (0.601 sec)\n",
      "Computing ML distances based on estimated model parameters... 0.439 sec\n",
      "Computing BIONJ tree...\n",
      "0.022 seconds\n",
      "Log-likelihood of BIONJ tree: -44404.492\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 1.985 second\n",
      "Computing log-likelihood of 98 initial trees ... 5.777 seconds\n",
      "Current best score: -44401.635\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -44401.253\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 3: -44401.232\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 6: -44401.002\n",
      "Iteration 10 / LogL: -44402.181 / Time: 0h:0m:13s\n",
      "Iteration 20 / LogL: -44401.155 / Time: 0h:0m:17s\n",
      "Finish initializing candidate tree set (20)\n",
      "Current best tree score: -44401.002 / CPU time: 16.038\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 23: -44400.427\n",
      "Iteration 30 / LogL: -44400.490 / Time: 0h:0m:22s (0h:1m:13s left)\n",
      "Iteration 40 / LogL: -44400.783 / Time: 0h:0m:28s (0h:0m:59s left)\n",
      "Iteration 50 / LogL: -44468.581 / Time: 0h:0m:33s (0h:0m:49s left)\n",
      "Iteration 60 / LogL: -44400.564 / Time: 0h:0m:38s (0h:0m:41s left)\n",
      "Iteration 70 / LogL: -44400.518 / Time: 0h:0m:44s (0h:0m:34s left)\n",
      "Iteration 80 / LogL: -44422.235 / Time: 0h:0m:50s (0h:0m:27s left)\n",
      "Iteration 90 / LogL: -44400.590 / Time: 0h:0m:55s (0h:0m:20s left)\n",
      "Iteration 100 / LogL: -44400.629 / Time: 0h:1m:0s (0h:0m:14s left)\n",
      "Iteration 110 / LogL: -44400.733 / Time: 0h:1m:5s (0h:0m:7s left)\n",
      "Iteration 120 / LogL: -44400.601 / Time: 0h:1m:9s (0h:0m:1s left)\n",
      "TREE SEARCH COMPLETED AFTER 124 ITERATIONS / Time: 0h:1m:11s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -44400.427\n",
      "Optimal log-likelihood: -44400.424\n",
      "Rate parameters:  A-C: 0.41263  A-G: 2.22344  A-T: 0.41263  C-G: 1.00000  C-T: 7.32495  G-T: 1.00000\n",
      "Base frequencies:  A: 0.299  C: 0.183  G: 0.196  T: 0.322\n",
      "Proportion of invariable sites: 0.795\n",
      "Parameters optimization took 1 rounds (0.131 sec)\n",
      "BEST SCORE FOUND : -44400.424\n",
      "Total tree length: 0.015\n",
      "\n",
      "Total number of iterations: 124\n",
      "CPU time used for tree search: 69.553 sec (0h:1m:9s)\n",
      "Wall-clock time used for tree search: 70.473 sec (0h:1m:10s)\n",
      "Total CPU time used: 70.888 sec (0h:1m:10s)\n",
      "Total wall-clock time used: 71.826 sec (0h:1m:11s)\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                ../data/COG-UK-NORW/results/20220210/NORW_200_aln.fa.iqtree\n",
      "  Maximum-likelihood tree:       ../data/COG-UK-NORW/results/20220210/NORW_200_aln.fa.treefile\n",
      "  Likelihood distances:          ../data/COG-UK-NORW/results/20220210/NORW_200_aln.fa.mldist\n",
      "  Screen log file:               ../data/COG-UK-NORW/results/20220210/NORW_200_aln.fa.log\n",
      "\n",
      "Date and Time: Mon Feb 14 15:45:04 2022\n"
     ]
    }
   ],
   "source": [
    "! iqtree -s COG-UK-NORW/results/20220210/NORW_200_aln.fa -pre 'COG-UK-NORW/results/20220210/NORW_200_aln.fa'\n",
    "! grep '>' COG-UK-NORW/results/20220210/NORW_200_aln.fa | sed 's/>//g' > 'COG-UK-NORW/results/20220210/NORW_200_leaftips.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4c6323d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "readfile = 'COG-UK-NORW/NORW_metadata.csv'\n",
    "df = pd.read_csv(readfile, header=0, usecols=['sequence_name', 'sample_date'])\n",
    "df = df.rename(columns={'sequence_name':'name', 'sample_date':'date'}) #headers required by treetime\n",
    "\n",
    "#include only the names that match the names in the alignment\n",
    "leaf_df = pd. read_csv('COG-UK-NORW/results/20220210/NORW_200_leaftips.csv', header=None)\n",
    "leaf_df = leaf_df.rename(columns={0:'name'})\n",
    "\n",
    "#add dates to leaf_df, and format the names like in the tree\n",
    "save_df = leaf_df.merge(df, how='left')\n",
    "save_df['name'] = save_df['name'].str.replace('/', '_')\n",
    "\n",
    "save_df.to_csv('COG-UK-NORW/results/20220210/NORW_treetime_dates.csv', sep=',', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f5e67e2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "! sed -i -e 's/\\//_/g' COG-UK-NORW/results/20220210/NORW_200_aln.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f6794b0",
   "metadata": {},
   "source": [
    "1) `--relax 1.0 0`, `--clock-rate 0.008`, `--reroot 'oldest'` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "53585fc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Attempting to parse dates...\n",
      "\tUsing column 'name' as name. This needs match the taxon names in the tree!!\n",
      "\tUsing column 'date' as date.\n",
      "\n",
      "0.00\t-TreeAnc: set-up\n",
      "\n",
      "3.57\tWARNING: Previous versions of TreeTime (<0.7.0) RECONSTRUCTED sequences of\n",
      "    \ttips at positions with AMBIGUOUS bases. This resulted in unexpected\n",
      "    \tbehavior is some cases and is no longer done by default. If you want to\n",
      "    \treplace those ambiguous sites with their most likely state, rerun with\n",
      "    \t`reconstruct_tip_states=True` or `--reconstruct-tip-states`.\n",
      "\n",
      "5.02\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "5.16\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "5.96\t###TreeTime.run: INITIAL ROUND\n",
      "\n",
      "12.19\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "12.31\t###TreeTime.run: rerunning timetree after rerooting\n",
      "\n",
      "18.75\t###TreeTime.run: ITERATION 1 out of 2 iterations\n",
      "relaxed_clock {'slack': 1.0, 'coupling': 0.0}\n",
      "\n",
      "26.79\t###TreeTime.run: ITERATION 2 out of 2 iterations\n",
      "relaxed_clock {'slack': 1.0, 'coupling': 0.0}\n",
      "\n",
      "34.21\t###TreeTime.run: CONVERGED\n",
      "\n",
      "Inferred sequence evolution model (saved as ../data/COG-UK-NORW/results/20220210/timetree-1/sequence_evolution_model.txt):\n",
      "Substitution rate (mu): 1.0\n",
      "\n",
      "Equilibrium frequencies (pi_i):\n",
      "  A: 0.2964\n",
      "  C: 0.1787\n",
      "  G: 0.1938\n",
      "  T: 0.3212\n",
      "  -: 0.0099\n",
      "\n",
      "Symmetrized rates from j->i (W_ij):\n",
      "\tA\tC\tG\tT\t-\n",
      "  A\t0\t0.4073\t1.4634\t0.2335\t6.2291\n",
      "  C\t0.4073\t0\t0.4359\t4.7819\t4.6598\n",
      "  G\t1.4634\t0.4359\t0\t0.7731\t3.7078\n",
      "  T\t0.2335\t4.7819\t0.7731\t0\t7.9399\n",
      "  -\t6.2291\t4.6598\t3.7078\t7.9399\t0\n",
      "\n",
      "Actual rates from j->i (Q_ij):\n",
      "\tA\tC\tG\tT\t-\n",
      "  A\t0\t0.1207\t0.4338\t0.0692\t1.8463\n",
      "  C\t0.0728\t0\t0.0779\t0.8545\t0.8327\n",
      "  G\t0.2836\t0.0845\t0\t0.1498\t0.7185\n",
      "  T\t0.075\t1.536\t0.2483\t0\t2.5503\n",
      "  -\t0.0618\t0.0462\t0.0368\t0.0787\t0\n",
      "\n",
      "\n",
      "Inferred sequence evolution model (saved as ../data/COG-UK-NORW/results/20220210/timetree-1/molecular_clock.txt):\n",
      "Root-Tip-Regression:\n",
      " --rate:\t8.000e-03\n",
      " --r^2:  \t0.25\n",
      "\n",
      "/home/madeline/anaconda3/lib/python3.7/site-packages/Bio/Phylo/_utils.py:504: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n",
      "--- saved tree as \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-1/timetree.pdf\n",
      "\n",
      "--- root-to-tip plot saved to  \n",
      "\t../data/COG-UK-NORW/results/20220210/timetree-1/root_to_tip_regression.pdf\n",
      "--- wrote branch specific rates to\n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-1/substitution_rates.tsv\n",
      "\n",
      "\n",
      "--- alignment including ancestral nodes saved as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-1/ancestral_sequences.fasta\n",
      "\n",
      "--- saved divergence times in \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-1/dates.tsv\n",
      "\n",
      "--- tree saved in nexus format as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-1/timetree.nexus\n",
      "\n",
      "--- divergence tree saved in nexus format as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-1/divergence_tree.nexus\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! treetime --relax 1.0 0 --clock-rate 0.008 --reroot 'oldest' --tree COG-UK-NORW/results/iqtree/NORW_200_ref_aln.fa.treefile --dates COG-UK-NORW/NORW_treetime_dates.csv --aln COG-UK-NORW/results/NORW_200_ref_aln.fa --outdir COG-UK-NORW/results/20220210/timetree-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a3111c48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"600\"\n",
       "            height=\"300\"\n",
       "            src=\"../data/COG-UK-NORW/results/20220210/timetree-1/timetree.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f28bbc2d790>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IFrame(\"COG-UK-NORW/results/20220210/timetree-1/timetree.pdf\", width=600, height=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27e89fbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "2) `--relax 1.0 0`, `--clock-rate 0.008`, `--reroot 'oldest'` --coalescent 'skyline'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1039abef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Attempting to parse dates...\n",
      "\tUsing column 'name' as name. This needs match the taxon names in the tree!!\n",
      "\tUsing column 'date' as date.\n",
      "\n",
      "0.00\t-TreeAnc: set-up\n",
      "\n",
      "3.51\tWARNING: Previous versions of TreeTime (<0.7.0) RECONSTRUCTED sequences of\n",
      "    \ttips at positions with AMBIGUOUS bases. This resulted in unexpected\n",
      "    \tbehavior is some cases and is no longer done by default. If you want to\n",
      "    \treplace those ambiguous sites with their most likely state, rerun with\n",
      "    \t`reconstruct_tip_states=True` or `--reconstruct-tip-states`.\n",
      "\n",
      "5.02\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "5.16\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "5.99\t###TreeTime.run: INITIAL ROUND\n",
      "\n",
      "12.01\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "12.13\t###TreeTime.run: rerunning timetree after rerooting\n",
      "\n",
      "18.36\t###TreeTime.run: ITERATION 1 out of 2 iterations\n",
      "relaxed_clock {'slack': 1.0, 'coupling': 0.0}\n",
      "\n",
      "34.17\t###TreeTime.run: ITERATION 2 out of 2 iterations\n",
      "relaxed_clock {'slack': 1.0, 'coupling': 0.0}\n",
      "\n",
      "Inferred sequence evolution model (saved as ../data/COG-UK-NORW/results/20220210/timetree-2/sequence_evolution_model.txt):\n",
      "Substitution rate (mu): 1.0\n",
      "\n",
      "Equilibrium frequencies (pi_i):\n",
      "  A: 0.2964\n",
      "  C: 0.1787\n",
      "  G: 0.1938\n",
      "  T: 0.3212\n",
      "  -: 0.0099\n",
      "\n",
      "Symmetrized rates from j->i (W_ij):\n",
      "\tA\tC\tG\tT\t-\n",
      "  A\t0\t0.4073\t1.4634\t0.2335\t6.2291\n",
      "  C\t0.4073\t0\t0.4359\t4.7819\t4.6598\n",
      "  G\t1.4634\t0.4359\t0\t0.7731\t3.7078\n",
      "  T\t0.2335\t4.7819\t0.7731\t0\t7.9399\n",
      "  -\t6.2291\t4.6598\t3.7078\t7.9399\t0\n",
      "\n",
      "Actual rates from j->i (Q_ij):\n",
      "\tA\tC\tG\tT\t-\n",
      "  A\t0\t0.1207\t0.4338\t0.0692\t1.8463\n",
      "  C\t0.0728\t0\t0.0779\t0.8545\t0.8327\n",
      "  G\t0.2836\t0.0845\t0\t0.1498\t0.7185\n",
      "  T\t0.075\t1.536\t0.2483\t0\t2.5503\n",
      "  -\t0.0618\t0.0462\t0.0368\t0.0787\t0\n",
      "\n",
      "\n",
      "Inferred sequence evolution model (saved as ../data/COG-UK-NORW/results/20220210/timetree-2/molecular_clock.txt):\n",
      "Root-Tip-Regression:\n",
      " --rate:\t8.000e-03\n",
      " --r^2:  \t0.27\n",
      "\n",
      "Inferred coalescent model\n",
      "\tSkyline assuming 50 gen/year and approximate confidence bounds (+/- 2.000000 standard deviations of the LH)\n",
      "\tdate \tN_e \tlower \tupper\n",
      "\t2019.985\t5.943e+00\t2.419e+00\t1.460e+01\n",
      "\t2020.046\t6.450e+00\t3.180e+00\t1.308e+01\n",
      "\t2020.107\t6.718e+00\t3.313e+00\t1.363e+01\n",
      "\t2020.168\t7.062e+00\t3.482e+00\t1.432e+01\n",
      "\t2020.229\t7.486e+00\t3.691e+00\t1.518e+01\n",
      "\t2020.290\t7.970e+00\t3.930e+00\t1.616e+01\n",
      "\t2020.351\t8.506e+00\t4.194e+00\t1.725e+01\n",
      "\t2020.412\t9.119e+00\t4.496e+00\t1.849e+01\n",
      "\t2020.473\t9.781e+00\t4.823e+00\t1.984e+01\n",
      "\t2020.534\t1.034e+01\t5.099e+00\t2.097e+01\n",
      "\t2020.595\t1.069e+01\t5.273e+00\t2.169e+01\n",
      "\t2020.656\t1.070e+01\t5.277e+00\t2.171e+01\n",
      "\t2020.717\t1.073e+01\t5.292e+00\t2.177e+01\n",
      "\t2020.778\t1.088e+01\t5.362e+00\t2.206e+01\n",
      "\t2020.839\t1.114e+01\t5.535e+00\t2.241e+01\n",
      "\t2020.900\t1.174e+01\t6.270e+00\t2.199e+01\n",
      "\t2020.961\t2.069e+01\t1.302e+01\t3.290e+01\n",
      "\t2021.022\t6.827e+01\t5.398e+01\t8.635e+01\n",
      "\t2021.083\t8.796e+01\t6.500e+01\t1.190e+02\n",
      "\t2021.144\t4.585e+01\t2.241e+01\t9.381e+01\n",
      "\n",
      " --- written skyline to ../data/COG-UK-NORW/results/20220210/timetree-2/skyline.tsv\n",
      "\n",
      "/home/madeline/anaconda3/lib/python3.7/site-packages/Bio/Phylo/_utils.py:504: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n",
      "--- saved tree as \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-2/timetree.pdf\n",
      "\n",
      "--- root-to-tip plot saved to  \n",
      "\t../data/COG-UK-NORW/results/20220210/timetree-2/root_to_tip_regression.pdf\n",
      "--- wrote branch specific rates to\n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-2/substitution_rates.tsv\n",
      "\n",
      "\n",
      "--- alignment including ancestral nodes saved as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-2/ancestral_sequences.fasta\n",
      "\n",
      "--- saved divergence times in \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-2/dates.tsv\n",
      "\n",
      "--- tree saved in nexus format as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-2/timetree.nexus\n",
      "\n",
      "--- divergence tree saved in nexus format as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-2/divergence_tree.nexus\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! treetime --relax 1.0 0 --clock-rate 0.008 --reroot 'oldest' --coalescent 'skyline' --tree COG-UK-NORW/results/iqtree/NORW_200_ref_aln.fa.treefile --dates COG-UK-NORW/NORW_treetime_dates.csv --aln COG-UK-NORW/results/NORW_200_ref_aln.fa --outdir COG-UK-NORW/results/20220210/timetree-2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ce7fb4bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"600\"\n",
       "            height=\"300\"\n",
       "            src=\"../data/COG-UK-NORW/results/20220210/timetree-2/skyline.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f28bbc20f50>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IFrame(\"COG-UK-NORW/results/20220210/timetree-2/skyline.pdf\", width=600, height=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c813f3f",
   "metadata": {},
   "source": [
    "The skyline plot is a graph of the estimated \"piece-wise linear merger rate trajectory\" that TreeTime creates: a guess at the population size for different timepoints on the tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cf15ebf8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ancestral_sequences.fasta  root_to_tip_regression.pdf\t substitution_rates.tsv\r\n",
      "dates.tsv\t\t   sequence_evolution_model.txt  timetree.nexus\r\n",
      "divergence_tree.nexus\t   skyline.pdf\t\t\t timetree.pdf\r\n",
      "molecular_clock.txt\t   skyline.tsv\t\t\t trace_run.log\r\n"
     ]
    }
   ],
   "source": [
    "! ls COG-UK-NORW/results/20220210/timetree-2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71eaa12b",
   "metadata": {},
   "source": [
    "3) `--relax 1.0 0`, `--clock-rate 0.008`, `--reroot 'oldest'` --coalescent 'const'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "aa5053c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Attempting to parse dates...\n",
      "\tUsing column 'name' as name. This needs match the taxon names in the tree!!\n",
      "\tUsing column 'date' as date.\n",
      "\n",
      "0.00\t-TreeAnc: set-up\n",
      "\n",
      "3.56\tWARNING: Previous versions of TreeTime (<0.7.0) RECONSTRUCTED sequences of\n",
      "    \ttips at positions with AMBIGUOUS bases. This resulted in unexpected\n",
      "    \tbehavior is some cases and is no longer done by default. If you want to\n",
      "    \treplace those ambiguous sites with their most likely state, rerun with\n",
      "    \t`reconstruct_tip_states=True` or `--reconstruct-tip-states`.\n",
      "\n",
      "5.11\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "5.23\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "6.02\t###TreeTime.run: INITIAL ROUND\n",
      "\n",
      "12.22\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "12.34\t###TreeTime.run: rerunning timetree after rerooting\n",
      "\n",
      "18.73\t###TreeTime.run: ITERATION 1 out of 2 iterations\n",
      "relaxed_clock {'slack': 1.0, 'coupling': 0.0}\n",
      "\n",
      "34.90\t###TreeTime.run: ITERATION 2 out of 2 iterations\n",
      "relaxed_clock {'slack': 1.0, 'coupling': 0.0}\n",
      "\n",
      "Inferred sequence evolution model (saved as ../data/COG-UK-NORW/results/20220210/timetree-3/sequence_evolution_model.txt):\n",
      "Substitution rate (mu): 1.0\n",
      "\n",
      "Equilibrium frequencies (pi_i):\n",
      "  A: 0.2964\n",
      "  C: 0.1787\n",
      "  G: 0.1938\n",
      "  T: 0.3212\n",
      "  -: 0.0099\n",
      "\n",
      "Symmetrized rates from j->i (W_ij):\n",
      "\tA\tC\tG\tT\t-\n",
      "  A\t0\t0.4073\t1.4634\t0.2335\t6.2291\n",
      "  C\t0.4073\t0\t0.4359\t4.7819\t4.6598\n",
      "  G\t1.4634\t0.4359\t0\t0.7731\t3.7078\n",
      "  T\t0.2335\t4.7819\t0.7731\t0\t7.9399\n",
      "  -\t6.2291\t4.6598\t3.7078\t7.9399\t0\n",
      "\n",
      "Actual rates from j->i (Q_ij):\n",
      "\tA\tC\tG\tT\t-\n",
      "  A\t0\t0.1207\t0.4338\t0.0692\t1.8463\n",
      "  C\t0.0728\t0\t0.0779\t0.8545\t0.8327\n",
      "  G\t0.2836\t0.0845\t0\t0.1498\t0.7185\n",
      "  T\t0.075\t1.536\t0.2483\t0\t2.5503\n",
      "  -\t0.0618\t0.0462\t0.0368\t0.0787\t0\n",
      "\n",
      "\n",
      "Inferred sequence evolution model (saved as ../data/COG-UK-NORW/results/20220210/timetree-3/molecular_clock.txt):\n",
      "Root-Tip-Regression:\n",
      " --rate:\t8.000e-03\n",
      " --r^2:  \t0.27\n",
      "\n",
      "Inferred coalescent model\n",
      " --T_c: \t 1.06e-02 \toptimized inverse merger rate in units of substitutions\n",
      " --T_c: \t 1.32e+00 \toptimized inverse merger rate in years\n",
      " --N_e: \t 6.62e+01 \tcorresponding 'effective population size' assuming 50 gen/year\n",
      "\n",
      "/home/madeline/anaconda3/lib/python3.7/site-packages/Bio/Phylo/_utils.py:504: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n",
      "--- saved tree as \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-3/timetree.pdf\n",
      "\n",
      "--- root-to-tip plot saved to  \n",
      "\t../data/COG-UK-NORW/results/20220210/timetree-3/root_to_tip_regression.pdf\n",
      "--- wrote branch specific rates to\n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-3/substitution_rates.tsv\n",
      "\n",
      "\n",
      "--- alignment including ancestral nodes saved as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-3/ancestral_sequences.fasta\n",
      "\n",
      "--- saved divergence times in \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-3/dates.tsv\n",
      "\n",
      "--- tree saved in nexus format as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-3/timetree.nexus\n",
      "\n",
      "--- divergence tree saved in nexus format as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-3/divergence_tree.nexus\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! treetime --relax 1.0 0 --clock-rate 0.008 --reroot 'oldest' --coalescent 'const' --tree COG-UK-NORW/results/iqtree/NORW_200_ref_aln.fa.treefile --dates COG-UK-NORW/NORW_treetime_dates.csv --aln COG-UK-NORW/results/NORW_200_ref_aln.fa --outdir COG-UK-NORW/results/20220210/timetree-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "06bd33a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"600\"\n",
       "            height=\"300\"\n",
       "            src=\"../data/COG-UK-NORW/results/20220210/timetree-3/timetree.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f586c4577d0>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IFrame(\"COG-UK-NORW/results/20220210/timetree-3/timetree.pdf\", width=600, height=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5914a47",
   "metadata": {},
   "source": [
    "4) `--relax 1.0 0`, `--clock-rate 0.008`, `--reroot 'oldest'` --coalescent '20.0'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "661b7883",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Attempting to parse dates...\n",
      "\tUsing column 'name' as name. This needs match the taxon names in the tree!!\n",
      "\tUsing column 'date' as date.\n",
      "\n",
      "0.00\t-TreeAnc: set-up\n",
      "\n",
      "3.66\tWARNING: Previous versions of TreeTime (<0.7.0) RECONSTRUCTED sequences of\n",
      "    \ttips at positions with AMBIGUOUS bases. This resulted in unexpected\n",
      "    \tbehavior is some cases and is no longer done by default. If you want to\n",
      "    \treplace those ambiguous sites with their most likely state, rerun with\n",
      "    \t`reconstruct_tip_states=True` or `--reconstruct-tip-states`.\n",
      "\n",
      "5.12\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "5.26\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "6.06\t###TreeTime.run: INITIAL ROUND\n",
      "\n",
      "12.09\tTreeTime.reroot: with method or node: oldest\n",
      "\n",
      "12.21\t###TreeTime.run: rerunning timetree after rerooting\n",
      "\n",
      "18.43\t###TreeTime.run: ITERATION 1 out of 2 iterations\n",
      "relaxed_clock {'slack': 1.0, 'coupling': 0.0}\n",
      "\n",
      "33.48\t###TreeTime.run: ITERATION 2 out of 2 iterations\n",
      "relaxed_clock {'slack': 1.0, 'coupling': 0.0}\n",
      "\n",
      "Inferred sequence evolution model (saved as ../data/COG-UK-NORW/results/20220210/timetree-4/sequence_evolution_model.txt):\n",
      "Substitution rate (mu): 1.0\n",
      "\n",
      "Equilibrium frequencies (pi_i):\n",
      "  A: 0.2964\n",
      "  C: 0.1787\n",
      "  G: 0.1938\n",
      "  T: 0.3212\n",
      "  -: 0.0099\n",
      "\n",
      "Symmetrized rates from j->i (W_ij):\n",
      "\tA\tC\tG\tT\t-\n",
      "  A\t0\t0.4073\t1.4634\t0.2335\t6.2291\n",
      "  C\t0.4073\t0\t0.4359\t4.7819\t4.6598\n",
      "  G\t1.4634\t0.4359\t0\t0.7731\t3.7078\n",
      "  T\t0.2335\t4.7819\t0.7731\t0\t7.9399\n",
      "  -\t6.2291\t4.6598\t3.7078\t7.9399\t0\n",
      "\n",
      "Actual rates from j->i (Q_ij):\n",
      "\tA\tC\tG\tT\t-\n",
      "  A\t0\t0.1207\t0.4338\t0.0692\t1.8463\n",
      "  C\t0.0728\t0\t0.0779\t0.8545\t0.8327\n",
      "  G\t0.2836\t0.0845\t0\t0.1498\t0.7185\n",
      "  T\t0.075\t1.536\t0.2483\t0\t2.5503\n",
      "  -\t0.0618\t0.0462\t0.0368\t0.0787\t0\n",
      "\n",
      "\n",
      "Inferred sequence evolution model (saved as ../data/COG-UK-NORW/results/20220210/timetree-4/molecular_clock.txt):\n",
      "Root-Tip-Regression:\n",
      " --rate:\t8.000e-03\n",
      " --r^2:  \t0.27\n",
      "\n",
      "/home/madeline/anaconda3/lib/python3.7/site-packages/Bio/Phylo/_utils.py:504: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n",
      "--- saved tree as \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-4/timetree.pdf\n",
      "\n",
      "--- root-to-tip plot saved to  \n",
      "\t../data/COG-UK-NORW/results/20220210/timetree-4/root_to_tip_regression.pdf\n",
      "--- wrote branch specific rates to\n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-4/substitution_rates.tsv\n",
      "\n",
      "\n",
      "--- alignment including ancestral nodes saved as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-4/ancestral_sequences.fasta\n",
      "\n",
      "--- saved divergence times in \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-4/dates.tsv\n",
      "\n",
      "--- tree saved in nexus format as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-4/timetree.nexus\n",
      "\n",
      "--- divergence tree saved in nexus format as  \n",
      "\t ../data/COG-UK-NORW/results/20220210/timetree-4/divergence_tree.nexus\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! treetime --relax 1.0 0 --clock-rate 0.008 --reroot 'oldest' --coalescent '20.0' --tree COG-UK-NORW/results/iqtree/NORW_200_ref_aln.fa.treefile --dates COG-UK-NORW/NORW_treetime_dates.csv --aln COG-UK-NORW/results/NORW_200_ref_aln.fa --outdir COG-UK-NORW/results/20220210/timetree-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ae30b1aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"600\"\n",
       "            height=\"300\"\n",
       "            src=\"../data/COG-UK-NORW/results/20220210/timetree-4/timetree.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f5854c04fd0>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IFrame(\"COG-UK-NORW/results/20220210/timetree-4/timetree.pdf\", width=600, height=300)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
