{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c7851c33",
   "metadata": {},
   "source": [
    "In an attempt to learn BEAST2, I am following [this tutorial](https://taming-the-beast.org/tutorials/Introduction-to-BEAST2/)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8be2c33",
   "metadata": {},
   "source": [
    "# Opening a file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91255c18",
   "metadata": {},
   "source": [
    "First, open the `beauti2` GUI:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "234aaec7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cd ~/beast/bin\n",
    "# ./beauti"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ae5c3a6",
   "metadata": {},
   "source": [
    "Then load in a nexus file (.nex, .nxs, .nexus):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0c98dc71",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#NEXUS\r",
      "\r",
      "[!Data from:\r",
      "\tHayasaka, K., T. Gojobori, and S. Horai. 1988. Molecular phylogeny\r",
      "\t\tand evolution of primate mitochondrial DNA. Mol. Biol. Evol.\r",
      "\t\t5:626-644.\r",
      "]\r",
      "\r",
      "begin taxa;\r",
      "\tdimensions ntax=12;\r",
      "\ttaxlabels\r",
      "Lemur_catta\t\r",
      "Homo_sapiens\r",
      "Pan\r",
      "Gorilla\r",
      "Pongo\r",
      "Hylobates\r",
      "Macaca_fuscata\r",
      "M._mulatta\r",
      "M._fascicularis \r",
      "M._sylvanus\r",
      "Saimiri_sciureus\r",
      "Tarsius_syrichta\r",
      ";\r",
      "end;\r",
      "\r",
      "begin characters;\r",
      "\tdimensions nchar=898;\r",
      "\tformat missing=? gap=- matchchar=. datatype=dna;\r",
      "\toptions gapmode=missing;\r",
      "\tmatrix\r",
      "[                                                                                                                      1                                                                                                   2                                                                                                   3                                                                                                   4                                                                                                   5                                                                                                   6                                                                                                   7                                                                                                   8                                                                                                  ]\r",
      "[                            1         2         3         4         5         6         7         8         9         0         1         2         3         4         5         6         7         8         9         0         1         2         3         4         5         6         7         8         9         0         1         2         3         4         5         6         7         8         9         0         1         2         3         4         5         6         7         8         9         0         1         2         3         4         5         6         7         8         9         0         1         2         3         4         5         6         7         8         9         0         1         2         3         4         5         6         7         8         9         0         1         2         3         4         5         6         7         8         9        ]\r",
      "[                   1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678]\r",
      "\r",
      "Lemur_catta         AAGCTTCATAGGAGCAACCATTCTAATAATCGCACATGGCCTTACATCATCCATATTATTCTGTCTAGCCAACTCTAACTACGAACGAATCCATAGCCGTACAATACTACTAGCACGAGGGATCCAAACCATTCTCCCTCTTATAGCCACCTGATGACTACTCGCCAGCCTAACTAACCTAGCCCTACCCACCTCTATCAATTTAATTGGCGAACTATTCGTCACTATAGCATCCTTCTCATGATCAAACATTACAATTATCTTAATAGGCTTAAATATGCTCATCACCGCTCTCTATTCCCTCTATATATTAACTACTACACAACGAGGAAAACTCACATATCATTCGCACAACCTAAACCCATCCTTTACACGAGAAAACACCCTTATATCCATACACATACTCCCCCTTCTCCTATTTACCTTAAACCCCAAAATTATTCTAGGACCCACGTACTGTAAATATAGTTTAAA-AAAACACTAGATTGTGAATCCAGAAATAGAAGCTCAAAC-CTTCTTATTTACCGAGAAAGTAATGTATGAACTGCTAACTCTGCACTCCGTATATAAAAATACGGCTATCTCAACTTTTAAAGGATAGAAGTAATCCATTGGCCTTAGGAGCCAAAAA-ATTGGTGCAACTCCAAATAAAAGTAATAAATCTATTATCCTCTTTCACCCTTGTCACACTGATTATCCTAACTTTACCTATCATTATAAACGTTACAAACATATACAAAAACTACCCCTATGCACCATACGTAAAATCTTCTATTGCATGTGCCTTCATCACTAGCCTCATCCCAACTATATTATTTATCTCCTCAGGACAAGAAACAATCATTTCCAACTGACATTGAATAACAATCCAAACCCTAAAACTATCTATTAGCTT\r",
      "Homo_sapiens        ........CC..C...GT......C........C..C..G........C..AT..C.......C.....A.....A...........C.CT..C..T..C.TC...A.C..CT.T.A...AC.T.....TC.A.....A..A.....TTTT........T..A..A.....CG.......C...T.....C..A....T..CC..C.G..A.....C.CT..GCTAG..A.CA.G.....C........T..C..TC.CC.AC.T.C...AC.C..C..A..AG....A..C..A..C........C.....T..C..A......T.G..CTC......CC.C..CATTA.....A....A..C..A..C.................C..G.T.......C..TC....A.......CC.AT..C.C......G.C..C...ACC..GTTTT.C.CT................CC......TC............TGAC..C...G...T.-CGA.CC.................CT-CAC.A.............ATGC.C..A.G.C...C..C.T....T...................AC..CT.........T......CC.......TT............................CCA.GCACA.TA..A.A...ACCC.A..C....C.TC.....T.CCC..C...C...CC.C.C.CGTT...CCTA.......AA..T.A..CC.C.AT..T........CAT.G.C....CCA....T..T.TC..T...T....C..A.C.A....C..G.G.CT..AC......GTT..T..C..G........C...GCC....C......AACCC.G..C..CC.A.....\r",
      "Pan                 ........CC..C....TT..C..C........C..C..A........C..AT..........C.....A.....A..T..T.....C.C...C..T..C.TC...A.T..CT.C.A...AC.T.....TC.A.....A..A......TTT........C..A..A.....CG.......C.........C.TA.C..T...C.CC.A..G.....C.C...GCTAG..A.C..A.....C........T.CC..TC.CC.AC.C.C...A..C..C..A..A.....A..C..G..C........C..G..T..C..A......T....CTC......CC.C..CATTA.T...A....G..C..A..C...........T..T..C....TTT.....C..TC....A.C.....TC.AT..C.C..T..TG.T..C..CACT...TT...C.C.................CC......TC............TGAC..C...G.....-CGA.CC.................CT-.A..A..........T..ATAT.C..A.GCC.G.C..C.T....T...................AC..CC....G....T......CC.......TT............................CCA.G.ATA.TA.CA.A...ACCT.A..C..A.C.CC.T...T.C.C..C...C.C.CC.C.C.C.TT...CCTA.......AA..T.A...C.C.AT..T..G.....CAT...C..G.CCA....T....T......TT....C..A.C.A....C..A.G.CT..AC......G.T..T..C..A.....G..C...GC.....C......AACCC.G..C..CC.A.....\r",
      "Gorilla             ........CC..C...GTTG....T.....T..C..C..A...........AT..........C.....A.....A.............C...C.....C.TC...A.T..CT.T.A...AC........C.A.....A..A......CTT........T..G..A.....CG.C.....C...T.....C..A.C..T..CC..C.A..A..G..C.C...ACTAG..A.CA.A.....C..........CC..CC..T.AC.T.C...A.CT..C..A..A..T..A..C..G..C.....T........T..C..A......T....CCC.......C.C..CATCAC....A....A..C..A................T...C....T...G...C..TC....A.C.....CC.AT..C.C......G.T.....CACC..GTT...C.C.................CC......TC............TGAT..C...G.....-CAA.CC.................CT-C...A..G..........ATAC.C....GCT.G.C..C.T....T...................AC..CT.........T.......C.......TT............................C.A.G.ACG.TA.CA.A...ACCT.AG.C..A.C.TC.T...T.CCC......C...CC.C.T.C.TC..TCCTA.......AAG.T.A..CC.C.AT............ATCG.C....CCA....T....TC......T....C..A.C.A.....C.A.G.CT..AC......G.T..T..C..A.G......C...GC.....C......AA.TC....C..CC.A.....\r",
      "Pongo               ........CC..C.......CC..C..G..T..C.....A..C.....C...C..C.G.....C.....A.....A.............C...C.....C.TC...A.C..CT.T.A...CC.T.....TC.A.....C..A......CT.........T..A..A.....C........T........AC..A.C.....CC.TC.A..A.....C.C...ACTA.....CATA.....T.....T.....C..C..CC.AC...C...AC.C..C..A..A.....AA.C..A..C..T...........C..C..A...........T.C..C....C.C..CATCA.....A....A..T..T..C.....C.....T.....C..GCT.......C..TC....A.C.....C..AT..C.C.......GC..C..CGCT..GTT.G.C...................CC......T.............T.AT.....GGC.C..-CAA.CC.................CT-CAC.A..............CACT-..A.G.G.G.C..C.T....T.....G.............AC..CT....C....T.......T.......TT.........................C.GCCA.G..TA..A.CA.A..TGCCC....CT.A.C.TC.....TCCCC..C..T.CCGCT.C.C.C.TT...CCCA.......AA....A..CC.C.AC..T......A.GG.C..C....CC.....T.CT.TC.....T........A.C.A.........G.CT............C...G.CA.A......TGC....C...C.C...G..A...C....C..AC.A.....\r",
      "Hylobates           ......T.C...T......G.C..C........C..C..A..A..C..T...C.GC.......C..T..A.....A.............CT..C.....C.TC...A.C...T.T......C.....G..T.A.....A..G......TT.........CGCA..A.....CG.......C.........C..A....T..CC.CC.A..T.....C.....ACTA..G..C........C..GG......C...T....CAC.C.CC..GC.C..CG.A..A.....G..C..A..C.....T..C.....T.TC.TA...........C.C...T...C.C..CATTA.A...A....A..C..AC.C.............TAT.A...CTT..G...C.CT.......C.....CC.A...C.C.....T..C..C...ACT..CTTT..TCC.......C.........TC......T.............T.AC......G....G-.AA.C....GC............CC-CAC.A.............ACT.TC..A.G...G.C..C.T....T...................AC..CT.........T.......C.......TT...........................GCAA.G.ACA..A.CA.AG..A..C.A..G..A.CCTC.....T.CCC..C..T.CAGCC.C.C...TT...CCCA.T.....GA..TTA..CC.G.AC.........ATGA.C.....C.C.A....T..A.TC.....AT.T..C..A...A....C..G.G.A...AC........C..T.....A........C....CTG...C......G...G....C..CC.A.....\r",
      "Macaca_fuscata      ......TTCC..C........C..T..G.....T..C..A..C..C..T.......AT.....C........T..A.....T.....C.CT..C.A......C........GT.C.....AC.T....T.C.A..T..A..A.C....TTT......T..ACA..A.....T........T.........C..A........C..C.A..T.....C..T..A.TCGC.A.C..A.....C.....CC.T..C..C.....GC...C...AC.T..C..AT.A..T..G..C.....C..T...C.C.....C.....A.............C.......C....CATAAT....A....G..CC....C..............AT.A....T.......C.CGCT..AA..A....TC.AT..C.C........C..C..C..G..GTTT..C.C.....G...........CT....................T.ACC.....GA....-C.A.C.................ACT-C.C.A.G.........C.ATGTAC.....CC.....T.....T.T...................AC..CT........A.........T......C.............................TCA.GCACA..C.CA...TTA.AACA..C..T..CTC.......C.C..A..TT..GCC.C.C.C.TC...CCT........ACGT..A..CC..GAT.........A.AA.CG.AAT..A...T.......TC......CC.T....A.CT.....C....T....AAC..........C.....GG.G..............TG.C......A...G.C...A.GC.A.....\r",
      "M._mulatta          ......TTCT..C........C..C..G..T..T..C..A..C..C..T.......AT.....C........T..A.....T.....C.CT..C.A......C........GT.C..G..AC.T....T.C.A..T..A..A.C...TTT.......T..ACA..A.....T........T.........C..A.......CC..C.A..T.....C..T..A.TCGCGA.C..A.....C..G..CC.T..C..C.....A....C...A..T..C..A..A..T..G..C.....C......C.C.....C..C..A............GC.......C....CATAAT....A....A..CC....C.............TAT.A....T.......C.CGCT..AA.CA....CC.AT.TC.C........C..C..C..G..GTTT..T.C.....G...........CT......T.............T.ACC.....GA..T.-C.A.C.................ACT-C.CGA.G.........C.ATGTA......CC.....T.....T.T...................AC..CT........A.........T......T.............................TCA.GCACA..C..A...TAA.AACA..C..T..CTC.......C.C..A..TT..GCC.C.C.C.TC...CCT........ACGT..A..CC..GAT.........A.AA.CG.AAT..A...T.......TC......CC.T....A.CT.....C....T....AAC..........C.....GA.G..............T..C......A...G.C...A.AC.A.....\r",
      "M._fascicularis     .......TCC..C.......CC..T........C..C..G..C..C..T.....G.AT.....CT.G.....T..A.....T..G..C.CT....A......C.........T.C.....AC.T....TTC.A..T..AT.G.C....TT.........CACA..A.....T........T.........C..A....T...C..C.A........C..T..A.TC.C.A.T..A..T..C.....CC.T..C..C...G.G....CG...C.T.....A..A.....A..C.....C..T...C.C..G..C.T...AGT...........C.......C.C..CATAAT...TA....A..CC....C.............TAT.A....T.......C.CGCT..AA..A....TC.AT.TC.C........C..C..C..G..GTTT..C.C.................CT......T.............T.ACT.....G..CT.-C.A...................ACT-C.C.A.G........TC.ATGC.......CT.....C.....T.TC..................AC..CT........A.........T......C.............................TCA.GCACA..C.CA...TAA.AACA..C..C..CTC...G..CC.T..A..TT..GCC.C.C.C..C...CCC..T.....ACGTT.A..CC..GAC.........A.AA.CG.AAT..A...T..T..T..C..T...CC.T....A.CCC....C...CT....AAC..........C.....GG.GT................C......AT..G.C...A.AC.A.....\r",
      "M._sylvanus         .......TCC..T.....T..C..T...G.T..C.....A..C..C..T.......AC.....CT.G........A...........C.C...C.....C.TC.........T.C.....AC......T.C.A.....A..A.C....TT.......T.CACA..A.....T.....T..T..T......T..A....T...C..C.G........C.....A.TCGC.A.C..A..T..C.....CC....C..C..C..AC...C...AC.G..C..A..A..T..A..C.....C..T..TC.C.....C..C..A............GCG......C.C..CATAATT...A....A...C.T..C.............TAT.A...CT.......C.CGCT..AA..A.T..TC.AT.TC.T........C..C.........TTT..T.C.................TT......T....C........T.ACT.........T.-C.A...................ACT-..C.A.G..C.....TC.ACAC.......CT.....C.....T.T...................AC..CT..................T......T.............................TCA.G.ATA..C.CA...TAA.AACA..T..C..CTC.......C.T..A..TT.CGCT.C.C...TC...CCCA.......AC...TA...C..AAC.........A.AG.CG.AAT..A...T.....T..C......TCTT....A.CT....A...A.T..T.AAC..............C.GA.G...G..C.......T..C......A....GC...A.AT.A.....\r",
      "Saimiri_sciureus    ........CC..C....TG..C...........T..C..GT....T..G..T..GC.......C.....A.....A..T...........T..C.....A......AC.T.TA.T......C.......AC.AT....G.......G.CT.........C..A..A.AT..CG.......C...........AG....T...C..G.A..A...T..C..ACA.TCG..T.T........T.....C...T....T.....A..C.C...AC.T.....A..A..T..A..A.....C..A..TC....G.ATG.CT.......G.....TCC...T.....CAGCA.CAG...TA....A...ATA..............T..G..A....TT.....T...ACA..AA.C.....CC.......G.G......GG.A...A........T.ACCT.....T.........GCT......T.............T.AT........AAT.-TAA.......A.............G-C.C.A..........T..ATGCTC..AAGAC...C..CTT....TC...................T...T.........T...............C............................---A..CAC.T...CA....T..AA.A.....A....G....CTAGCG..A...C.AGCT.C.C.C.TT...CCTA......G.ACA.TA..CC.GTAC.........CTAG.C..CATC.AC...C....T..C..TAC.T.AT.T.TA.....C......CTTA....C......T....A.....A........C........T.........A.C........CC.A.....\r",
      "Tarsius_syrichta    ...T.....T.....C....C...T.....T..C........C..C..C...C.......T..C.....A..TA.A............G....C..T..A......GC......C..T..CC.T......C.AT.A......GC...A..A........C.........T....C.....G.....T...C.AA.A..T........C..T.....G.C...A.TA......G.A..T.....G...C..C.A..T.........G.....C.T..C.CC..T........C..A........A......C...TC.TA..T............TA..........ATCA....TA.C.TG..CC.T..C..C........T..AT.A...AT.......C..T.T...T.AA.....C.AT.TACC.........G.A...A.....A...T.....................C......T..........G..T.AT.........C....GAT............A........-..C.A.............ATGC....A.......C...GT....T...T-..........................C..T.......A..G....-...............................T....T..A..C.C..TTT.AC.AT....T.CAC..T...T.ACC..AT.T.....T.CAACA..T..A.A...TG...CAC.TG.A..CC.TTAC.........AAC.....C..C..C..A..T..A..A.....AG......TGC.CA.....C.A.A.A..AAT.......T...........................G..T..T..TA.C...T...GCC.A.....\r",
      "\t;\r",
      "end;\r",
      "\r",
      "begin assumptions;\r",
      "\tcharset noncoding = 1 458-659 897-898;\r",
      "\tcharset 1stpos = 2-457\\3 660-896\\3;\r",
      "\tcharset 2ndpos = 3-457\\3 661-896\\3;\r",
      "\tcharset 3rdpos = 4-457\\3 662-.\\3;\r\n",
      "\r",
      "\texset noncoding = 1stpos 2ndpos 3rdpos;\r",
      "\t\r",
      "\tusertype 2_1 = 4\t\t[weights transversions 2 times transitions]\r",
      "\t\ta  c  g  t\r",
      "\t[a]\t.  2  1  2\r",
      "\t[c] 2  .  2  1\r",
      "\t[g]\t1  2  .  2\r",
      "\t[t]\t2  1  2  .\r",
      "\t\t;\r",
      "\tusertype 3_1 = 4\t\t[weights transversions 3 times transitions]\r",
      "\t\ta  c  g  t\r",
      "\t[a]\t.  3  1  3\r",
      "\t[c] 3  .  3  1\r",
      "\t[g]\t1  3  .  3\r",
      "\t[t]\t3  1  3  .\r",
      "\t\t;\r",
      "\t\t\r",
      "end;\r",
      "#\ttaxset hominoids = Homo_sapiens Pan Gorilla Pongo Hylobates;\r",
      "\r",
      "begin paup;\r",
      "\tconstraints ch = ((Homo_sapiens,Pan));\r",
      "\tconstraints chg = ((Homo_sapiens,Pan,Gorilla));\r",
      "end;\r",
      "\r"
     ]
    }
   ],
   "source": [
    "! head ../data/BEAST2_tutorial/primate-mtDNA.nex"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d49ae9da",
   "metadata": {},
   "source": [
    "This nexus file doesn't look much like TreeTime's nexus trees at all; but I suppose it follows the same rules for a nexus file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "589253a7",
   "metadata": {},
   "source": [
    "The point of `beauti2` is to generate a configuration file in `.xml` format that will be the input to `BEAST`.  The `.xml` file must contain:\n",
    "\n",
    "* The data (typically a sequence alignment)\n",
    "* The model specification\n",
    "* Initial values and parameter constraints\n",
    "* Settings of the MCMC algorithm\n",
    "* Output options\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8b80a70",
   "metadata": {},
   "source": [
    "BEAST2 uses a discretized gamma site model to account for variation in the rates at which different sites in the genome mutate."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db25d2c3",
   "metadata": {},
   "source": [
    "This particular alignment file is discretized into four partitions: noncoding, codon position 1, cp2, and cp3.  The default method is for BEAST to create a separate tree and model for each of those partitions **Question: why/what is the biological motivation for doing this?**.  For this example, you're going to have it use all the partitions to create one tree.  To do this, select all 4 partitions and click \"Link Trees\" and \"Link Clock Models\".  Then, rename the trees and clock models to something more generic."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d909072d",
   "metadata": {},
   "source": [
    "The 'tip dates' panel gets filled in if your samples are dated.  The example file sequences are all taken at the same time, so skip it this time."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d193dfb",
   "metadata": {},
   "source": [
    "Next is Site Model (even though we're producing a single tree, each partition gets its own site model, with different parameters estimated for each), Clock Model, and Priors.  The priors show your prior knowledge of what the model should be: more on this later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f063c322",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?><beast beautitemplate='Standard' beautistatus='' namespace=\"beast.core:beast.evolution.alignment:beast.evolution.tree.coalescent:beast.core.util:beast.evolution.nuc:beast.evolution.operators:beast.evolution.sitemodel:beast.evolution.substitutionmodel:beast.evolution.likelihood\" required=\"\" version=\"2.6\">\r\n",
      "        \r\n",
      "    <data\r\n",
      "id=\"primate-mtDNA\"\r\n",
      "spec=\"Alignment\">\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_Lemur_catta\" spec=\"Sequence\" taxon=\"Lemur_catta\" totalcount=\"4\" value=\"AAGCTTCATAGGAGCAACCATTCTAATAATCGCACATGGCCTTACATCATCCATATTATTCTGTCTAGCCAACTCTAACTACGAACGAATCCATAGCCGTACAATACTACTAGCACGAGGGATCCAAACCATTCTCCCTCTTATAGCCACCTGATGACTACTCGCCAGCCTAACTAACCTAGCCCTACCCACCTCTATCAATTTAATTGGCGAACTATTCGTCACTATAGCATCCTTCTCATGATCAAACATTACAATTATCTTAATAGGCTTAAATATGCTCATCACCGCTCTCTATTCCCTCTATATATTAACTACTACACAACGAGGAAAACTCACATATCATTCGCACAACCTAAACCCATCCTTTACACGAGAAAACACCCTTATATCCATACACATACTCCCCCTTCTCCTATTTACCTTAAACCCCAAAATTATTCTAGGACCCACGTACTGTAAATATAGTTTAAA-AAAACACTAGATTGTGAATCCAGAAATAGAAGCTCAAAC-CTTCTTATTTACCGAGAAAGTAATGTATGAACTGCTAACTCTGCACTCCGTATATAAAAATACGGCTATCTCAACTTTTAAAGGATAGAAGTAATCCATTGGCCTTAGGAGCCAAAAA-ATTGGTGCAACTCCAAATAAAAGTAATAAATCTATTATCCTCTTTCACCCTTGTCACACTGATTATCCTAACTTTACCTATCATTATAAACGTTACAAACATATACAAAAACTACCCCTATGCACCATACGTAAAATCTTCTATTGCATGTGCCTTCATCACTAGCCTCATCCCAACTATATTATTTATCTCCTCAGGACAAGAAACAATCATTTCCAACTGACATTGAATAACAATCCAAACCCTAAAACTATCTATTAGCTT\"/>\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_Homo_sapiens\" spec=\"Sequence\" taxon=\"Homo_sapiens\" totalcount=\"4\" value=\"AAGCTTCACCGGCGCAGTCATTCTCATAATCGCCCACGGGCTTACATCCTCATTACTATTCTGCCTAGCAAACTCAAACTACGAACGCACTCACAGTCGCATCATAATCCTCTCTCAAGGACTTCAAACTCTACTCCCACTAATAGCTTTTTGATGACTTCTAGCAAGCCTCGCTAACCTCGCCTTACCCCCCACTATTAACCTACTGGGAGAACTCTCTGTGCTAGTAACCACGTTCTCCTGATCAAATATCACTCTCCTACTTACAGGACTCAACATACTAGTCACAGCCCTATACTCCCTCTACATATTTACCACAACACAATGGGGCTCACTCACCCACCACATTAACAACATAAAACCCTCATTCACACGAGAAAACACCCTCATGTTCATACACCTATCCCCCATTCTCCTCCTATCCCTCAACCCCGACATCATTACCGGGTTTTCCTCTTGTAAATATAGTTTAACCAAAACATCAGATTGTGAATCTGACAACAGAGGCTTA-CGACCCCTTATTTACCGAGAAAGCT-CACAAGAACTGCTAACTCATGCCCCCATGTCTAACAACATGGCTTTCTCAACTTTTAAAGGATAACAGCTATCCATTGGTCTTAGGCCCCAAAAATTTTGGTGCAACTCCAAATAAAAGTAATAACCATGCACACTACTATAACCACCCTAACCCTGACTTCCCTAATTCCCCCCATCCTTACCACCCTCGTTAACCCTAACAAAAAAAACTCATACCCCCATTATGTAAAATCCATTGTCGCATCCACCTTTATTATCAGTCTCTTCCCCACAACAATATTCATGTGCCTAGACCAAGAAGTTATTATCTCGAACTGACACTGAGCCACAACCCAAACAACCCAGCTCTCCCTAAGCTT\"/>\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_Pan\" spec=\"Sequence\" taxon=\"Pan\" totalcount=\"4\" value=\"AAGCTTCACCGGCGCAATTATCCTCATAATCGCCCACGGACTTACATCCTCATTATTATTCTGCCTAGCAAACTCAAATTATGAACGCACCCACAGTCGCATCATAATTCTCTCCCAAGGACTTCAAACTCTACTCCCACTAATAGCCTTTTGATGACTCCTAGCAAGCCTCGCTAACCTCGCCCTACCCCCTACCATTAATCTCCTAGGGGAACTCTCCGTGCTAGTAACCTCATTCTCCTGATCAAATACCACTCTCCTACTCACAGGATTCAACATACTAATCACAGCCCTGTACTCCCTCTACATGTTTACCACAACACAATGAGGCTCACTCACCCACCACATTAATAACATAAAGCCCTCATTCACACGAGAAAATACTCTCATATTTTTACACCTATCCCCCATCCTCCTTCTATCCCTCAATCCTGATATCATCACTGGATTCACCTCCTGTAAATATAGTTTAACCAAAACATCAGATTGTGAATCTGACAACAGAGGCTCA-CGACCCCTTATTTACCGAGAAAGCT-TATAAGAACTGCTAATTCATATCCCCATGCCTGACAACATGGCTTTCTCAACTTTTAAAGGATAACAGCCATCCGTTGGTCTTAGGCCCCAAAAATTTTGGTGCAACTCCAAATAAAAGTAATAACCATGTATACTACCATAACCACCTTAACCCTAACTCCCTTAATTCTCCCCATCCTCACCACCCTCATTAACCCTAACAAAAAAAACTCATATCCCCATTATGTGAAATCCATTATCGCGTCCACCTTTATCATTAGCCTTTTCCCCACAACAATATTCATATGCCTAGACCAAGAAGCTATTATCTCAAACTGGCACTGAGCAACAACCCAAACAACCCAGCTCTCCCTAAGCTT\"/>\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_Gorilla\" spec=\"Sequence\" taxon=\"Gorilla\" totalcount=\"4\" value=\"AAGCTTCACCGGCGCAGTTGTTCTTATAATTGCCCACGGACTTACATCATCATTATTATTCTGCCTAGCAAACTCAAACTACGAACGAACCCACAGCCGCATCATAATTCTCTCTCAAGGACTCCAAACCCTACTCCCACTAATAGCCCTTTGATGACTTCTGGCAAGCCTCGCCAACCTCGCCTTACCCCCCACCATTAACCTACTAGGAGAGCTCTCCGTACTAGTAACCACATTCTCCTGATCAAACACCACCCTTTTACTTACAGGATCTAACATACTAATTACAGCCCTGTACTCCCTTTATATATTTACCACAACACAATGAGGCCCACTCACACACCACATCACCAACATAAAACCCTCATTTACACGAGAAAACATCCTCATATTCATGCACCTATCCCCCATCCTCCTCCTATCCCTCAACCCCGATATTATCACCGGGTTCACCTCCTGTAAATATAGTTTAACCAAAACATCAGATTGTGAATCTGATAACAGAGGCTCA-CAACCCCTTATTTACCGAGAAAGCT-CGTAAGAGCTGCTAACTCATACCCCCGTGCTTGACAACATGGCTTTCTCAACTTTTAAAGGATAACAGCTATCCATTGGTCTTAGGACCCAAAAATTTTGGTGCAACTCCAAATAAAAGTAATAACTATGTACGCTACCATAACCACCTTAGCCCTAACTTCCTTAATTCCCCCTATCCTTACCACCTTCATCAATCCTAACAAAAAAAGCTCATACCCCCATTACGTAAAATCTATCGTCGCATCCACCTTTATCATCAGCCTCTTCCCCACAACAATATTTCTATGCCTAGACCAAGAAGCTATTATCTCAAGCTGACACTGAGCAACAACCCAAACAATTCAACTCTCCCTAAGCTT\"/>\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_Pongo\" spec=\"Sequence\" taxon=\"Pongo\" totalcount=\"4\" value=\"AAGCTTCACCGGCGCAACCACCCTCATGATTGCCCATGGACTCACATCCTCCCTACTGTTCTGCCTAGCAAACTCAAACTACGAACGAACCCACAGCCGCATCATAATCCTCTCTCAAGGCCTTCAAACTCTACTCCCCCTAATAGCCCTCTGATGACTTCTAGCAAGCCTCACTAACCTTGCCCTACCACCCACCATCAACCTTCTAGGAGAACTCTCCGTACTAATAGCCATATTCTCTTGATCTAACATCACCATCCTACTAACAGGACTCAACATACTAATCACAACCCTATACTCTCTCTATATATTCACCACAACACAACGAGGTACACCCACACACCACATCAACAACATAAAACCTTCTTTCACACGCGAAAATACCCTCATGCTCATACACCTATCCCCCATCCTCCTCTTATCCCTCAACCCCAGCATCATCGCTGGGTTCGCCTACTGTAAATATAGTTTAACCAAAACATTAGATTGTGAATCTAATAATAGGGCCCCA-CAACCCCTTATTTACCGAGAAAGCT-CACAAGAACTGCTAACTCTCACT-CCATGTGTGACAACATGGCTTTCTCAGCTTTTAAAGGATAACAGCTATCCCTTGGTCTTAGGATCCAAAAATTTTGGTGCAACTCCAAATAAAAGTAACAGCCATGTTTACCACCATAACTGCCCTCACCTTAACTTCCCTAATCCCCCCCATTACCGCTACCCTCATTAACCCCAACAAAAAAAACCCATACCCCCACTATGTAAAAACGGCCATCGCATCCGCCTTTACTATCAGCCTTATCCCAACAACAATATTTATCTGCCTAGGACAAGAAACCATCGTCACAAACTGATGCTGAACAACCACCCAGACACTACAACTCTCACTAAGCTT\"/>\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_Hylobates\" spec=\"Sequence\" taxon=\"Hylobates\" totalcount=\"4\" value=\"AAGCTTTACAGGTGCAACCGTCCTCATAATCGCCCACGGACTAACCTCTTCCCTGCTATTCTGCCTTGCAAACTCAAACTACGAACGAACTCACAGCCGCATCATAATCCTATCTCGAGGGCTCCAAGCCTTACTCCCACTGATAGCCTTCTGATGACTCGCAGCAAGCCTCGCTAACCTCGCCCTACCCCCCACTATTAACCTCCTAGGTGAACTCTTCGTACTAATGGCCTCCTTCTCCTGGGCAAACACTACTATTACACTCACCGGGCTCAACGTACTAATCACGGCCCTATACTCCCTTTACATATTTATCATAACACAACGAGGCACACTTACACACCACATTAAAAACATAAAACCCTCACTCACACGAGAAAACATATTAATACTTATGCACCTCTTCCCCCTCCTCCTCCTAACCCTCAACCCTAACATCATTACTGGCTTTACTCCCTGTAAACATAGTTTAATCAAAACATTAGATTGTGAATCTAACAATAGAGGCTCG-AAACCTCTTGCTTACCGAGAAAGCC-CACAAGAACTGCTAACTCACTATCCCATGTATGACAACATGGCTTTCTCAACTTTTAAAGGATAACAGCTATCCATTGGTCTTAGGACCCAAAAATTTTGGTGCAACTCCAAATAAAAGTAATAGCAATGTACACCACCATAGCCATTCTAACGCTAACCTCCCTAATTCCCCCCATTACAGCCACCCTTATTAACCCCAATAAAAAGAACTTATACCCGCACTACGTAAAAATGACCATTGCCTCTACCTTTATAATCAGCCTATTTCCCACAATAATATTCATGTGCACAGACCAAGAAACCATTATTTCAAACTGACACTGAACTGCAACCCAAACGCTAGAACTCTCCCTAAGCTT\"/>\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_Macaca_fuscata\" spec=\"Sequence\" taxon=\"Macaca_fuscata\" totalcount=\"4\" value=\"AAGCTTTTCCGGCGCAACCATCCTTATGATCGCTCACGGACTCACCTCTTCCATATATTTCTGCCTAGCCAATTCAAACTATGAACGCACTCACAACCGTACCATACTACTGTCCCGAGGACTTCAAATCCTACTTCCACTAACAGCCTTTTGATGATTAACAGCAAGCCTTACTAACCTTGCCCTACCCCCCACTATCAATCTACTAGGTGAACTCTTTGTAATCGCAACCTCATTCTCCTGATCCCATATCACCATTATGCTAACAGGACTTAACATATTAATTACGGCCCTCTACTCTCTCCACATATTCACTACAACACAACGAGGAACACTCACACATCACATAATCAACATAAAGCCCCCCTTCACACGAGAAAACACATTAATATTCATACACCTCGCTCCAATTATCCTTCTATCCCTCAACCCCAACATCATCCTGGGGTTTACCTCCTGTAGATATAGTTTAACTAAAACACTAGATTGTGAATCTAACCATAGAGACTCA-CCACCTCTTATTTACCGAGAAAACT-CGCAAGGACTGCTAACCCATGTACCCGTACCTAAAATTACGGTTTTCTCAACTTTTAAAGGATAACAGCTATCCATTGACCTTAGGAGTCAAAAACATTGGTGCAACTCCAAATAAAAGTAATAATCATGCACACCCCCATCATTATAACAACCCTTATCTCCCTAACTCTCCCAATTTTTGCCACCCTCATCAACCCTTACAAAAAACGTCCATACCCAGATTACGTAAAAACAACCGTAATATATGCTTTCATCATCAGCCTCCCCTCAACAACTTTATTCATCTTCTCAAACCAAGAAACAACCATTTGGAGCTGACATTGAATAATGACCCAAACACTAGACCTAACGCTAAGCTT\"/>\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_M._mulatta\" spec=\"Sequence\" taxon=\"M._mulatta\" totalcount=\"4\" value=\"AAGCTTTTCTGGCGCAACCATCCTCATGATTGCTCACGGACTCACCTCTTCCATATATTTCTGCCTAGCCAATTCAAACTATGAACGCACTCACAACCGTACCATACTACTGTCCCGGGGACTTCAAATCCTACTTCCACTAACAGCTTTCTGATGATTAACAGCAAGCCTTACTAACCTTGCCCTACCCCCCACTATCAACCTACTAGGTGAACTCTTTGTAATCGCGACCTCATTCTCCTGGTCCCATATCACCATTATATTAACAGGATTTAACATACTAATTACGGCCCTCTACTCCCTCCACATATTCACCACAACACAACGAGGAGCACTCACACATCACATAATCAACATAAAACCCCCCTTCACACGAGAAAACATATTAATATTCATACACCTCGCTCCAATCATCCTCCTATCTCTCAACCCCAACATCATCCTGGGGTTTACTTCCTGTAGATATAGTTTAACTAAAACATTAGATTGTGAATCTAACCATAGAGACTTA-CCACCTCTTATTTACCGAGAAAACT-CGCGAGGACTGCTAACCCATGTATCCGTACCTAAAATTACGGTTTTCTCAACTTTTAAAGGATAACAGCTATCCATTGACCTTAGGAGTCAAAAATATTGGTGCAACTCCAAATAAAAGTAATAATCATGCACACCCCTATCATAATAACAACCCTTATCTCCCTAACTCTCCCAATTTTTGCCACCCTCATCAACCCTTACAAAAAACGTCCATACCCAGATTACGTAAAAACAACCGTAATATATGCTTTCATCATCAGCCTCCCCTCAACAACTTTATTCATCTTCTCAAACCAAGAAACAACCATTTGAAGCTGACATTGAATAATAACCCAAACACTAGACCTAACACTAAGCTT\"/>\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_M._fascicularis\" spec=\"Sequence\" taxon=\"M._fascicularis\" totalcount=\"4\" value=\"AAGCTTCTCCGGCGCAACCACCCTTATAATCGCCCACGGGCTCACCTCTTCCATGTATTTCTGCTTGGCCAATTCAAACTATGAGCGCACTCATAACCGTACCATACTACTATCCCGAGGACTTCAAATTCTACTTCCATTGACAGCCTTCTGATGACTCACAGCAAGCCTTACTAACCTTGCCCTACCCCCCACTATTAATCTACTAGGCGAACTCTTTGTAATCACAACTTCATTTTCCTGATCCCATATCACCATTGTGTTAACGGGCCTTAATATACTAATCACAGCCCTCTACTCTCTCCACATGTTCATTACAGTACAACGAGGAACACTCACACACCACATAATCAATATAAAACCCCCCTTCACACGAGAAAACATATTAATATTCATACACCTCGCTCCAATTATCCTTCTATCTCTCAACCCCAACATCATCCTGGGGTTTACCTCCTGTAAATATAGTTTAACTAAAACATTAGATTGTGAATCTAACTATAGAGGCCTA-CCACTTCTTATTTACCGAGAAAACT-CGCAAGGACTGCTAATCCATGCCTCCGTACTTAAAACTACGGTTTCCTCAACTTTTAAAGGATAACAGCTATCCATTGACCTTAGGAGTCAAAAACATTGGTGCAACTCCAAATAAAAGTAATAATCATGCACACCCCCATCATAATAACAACCCTCATCTCCCTGACCCTTCCAATTTTTGCCACCCTCACCAACCCCTATAAAAAACGTTCATACCCAGACTACGTAAAAACAACCGTAATATATGCTTTTATTACCAGTCTCCCCTCAACAACCCTATTCATCCTCTCAAACCAAGAAACAACCATTTGGAGTTGACATTGAATAACAACCCAAACATTAGACCTAACACTAAGCTT\"/>\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_M._sylvanus\" spec=\"Sequence\" taxon=\"M._sylvanus\" totalcount=\"4\" value=\"AAGCTTCTCCGGTGCAACTATCCTTATAGTTGCCCATGGACTCACCTCTTCCATATACTTCTGCTTGGCCAACTCAAACTACGAACGCACCCACAGCCGCATCATACTACTATCCCGAGGACTCCAAATCCTACTCCCACTAACAGCCTTCTGATGATTCACAGCAAGCCTTACTAATCTTGCTCTACCCTCCACTATTAATCTACTGGGCGAACTCTTCGTAATCGCAACCTCATTTTCCTGATCCCACATCACCATCATACTAACAGGACTGAACATACTAATTACAGCCCTCTACTCTCTTCACATATTCACCACAACACAACGAGGAGCGCTCACACACCACATAATTAACATAAAACCACCTTTCACACGAGAAAACATATTAATACTCATACACCTCGCTCCAATTATTCTTCTATCTCTTAACCCCAACATCATTCTAGGATTTACTTCCTGTAAATATAGTTTAATTAAAACATTAGACTGTGAATCTAACTATAGAAGCTTA-CCACTTCTTATTTACCGAGAAAACT-TGCAAGGACCGCTAATCCACACCTCCGTACTTAAAACTACGGTTTTCTCAACTTTTAAAGGATAACAGCTATCCATTGGCCTTAGGAGTCAAAAATATTGGTGCAACTCCAAATAAAAGTAATAATCATGTATACCCCCATCATAATAACAACTCTCATCTCCCTAACTCTTCCAATTTTCGCTACCCTTATCAACCCCAACAAAAAACACCTATATCCAAACTACGTAAAAACAGCCGTAATATATGCTTTCATTACCAGCCTCTCTTCAACAACTTTATATATATTCTTAAACCAAGAAACAATCATCTGAAGCTGGCACTGAATAATAACCCAAACACTAAGCCTAACATTAAGCTT\"/>\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_Saimiri_sciureus\" spec=\"Sequence\" taxon=\"Saimiri_sciureus\" totalcount=\"4\" value=\"AAGCTTCACCGGCGCAATGATCCTAATAATCGCTCACGGGTTTACTTCGTCTATGCTATTCTGCCTAGCAAACTCAAATTACGAACGAATTCACAGCCGAACAATAACATTTACTCGAGGGCTCCAAACACTATTCCCGCTTATAGGCCTCTGATGACTCCTAGCAAATCTCGCTAACCTCGCCCTACCCACAGCTATTAATCTAGTAGGAGAATTACTCACAATCGTATCTTCCTTCTCTTGATCCAACTTTACTATTATATTCACAGGACTTAATATACTAATTACAGCACTCTACTCACTTCATATGTATGCCTCTACACAGCGAGGTCCACTTACATACAGCACCAGCAATATAAAACCAATATTTACACGAGAAAATACGCTAATATTTATACATATAACACCAATCCTCCTCCTTACCTTGAGCCCCAAGGTAATTATAGGACCCTCACCTTGTAATTATAGTTTAGCTAAAACATTAGATTGTGAATCTAATAATAGAAGAATA-TAACTTCTTAATTACCGAGAAAGTG-CGCAAGAACTGCTAATTCATGCTCCCAAGACTAACAACTTGGCTTCCTCAACTTTTAAAGGATAGTAGTTATCCATTGGTCTTAGGAGCCAAAAACATTGGTGCAACTCCAAATAAAAGTAATA---ATACACTTCTCCATCACTCTAATAACACTAATTAGCCTACTAGCGCCAATCCTAGCTACCCTCATTAACCCTAACAAAAGCACACTATACCCGTACTACGTAAAACTAGCCATCATCTACGCCCTCATTACCAGTACCTTATCTATAATATTCTTTATCCTTACAGGCCAAGAATCAATAATTTCAAACTGACACTGAATAACTATCCAAACCATCAAACTATCCCTAAGCTT\"/>\r\n",
      "                                \r\n",
      "        <sequence id=\"seq_Tarsius_syrichta\" spec=\"Sequence\" taxon=\"Tarsius_syrichta\" totalcount=\"4\" value=\"AAGTTTCATTGGAGCCACCACTCTTATAATTGCCCATGGCCTCACCTCCTCCCTATTATTTTGCCTAGCAAATACAAACTACGAACGAGTCCACAGTCGAACAATAGCACTAGCCCGTGGCCTTCAAACCCTATTACCTCTTGCAGCAACATGATGACTCCTCGCCAGCTTAACCAACCTGGCCCTTCCCCCAACAATTAATTTAATCGGTGAACTGTCCGTAATAATAGCAGCATTTTCATGGTCACACCTAACTATTATCTTAGTAGGCCTTAACACCCTTATCACCGCCCTATATTCCCTATATATACTAATCATAACTCAACGAGGAAAATACACATATCATATCAACAATATCATGCCCCCTTTCACCCGAGAAAATACATTAATAATCATACACCTATTTCCCTTAATCCTACTATCTACCAACCCCAAAGTAATTATAGGAACCATGTACTGTAAATATAGTTTAAACAAAACATTAGATTGTGAGTCTAATAATAGAAGCCCAAAGATTTCTTATTTACCAAGAAAGTA-TGCAAGAACTGCTAACTCATGCCTCCATATATAACAATGTGGCTTTCTT-ACTTTTAAAGGATAGAAGTAATCCATCGGTCTTAGGAACCGAAAA-ATTGGTGCAACTCCAAATAAAAGTAATAAATTTATTTTCATCCTCCATTTTACTATCACTTACACTCTTAATTACCCCATTTATTATTACAACAACTAAAAAATATGAAACACATGCATACCCTTACTACGTAAAAAACTCTATCGCCTGCGCATTTATAACAAGCCTAGTCCCAATGCTCATATTTCTATACACAAATCAAGAAATAATCATTTCCAACTGACATTGAATAACGATTCATACTATCAAATTATGCCTAAGCTT\"/>\r\n",
      "                            \r\n",
      "    </data>\r\n",
      "        \r\n",
      "\r\n",
      "\r\n",
      "    \r\n",
      "\r\n",
      "\r\n",
      "    \r\n",
      "\r\n",
      "    <map name=\"Uniform\" >beast.math.distributions.Uniform</map>\r\n",
      "    \r\n",
      "    <map name=\"Exponential\" >beast.math.distributions.Exponential</map>\r\n",
      "    \r\n",
      "    <map name=\"LogNormal\" >beast.math.distributions.LogNormalDistributionModel</map>\r\n",
      "    \r\n",
      "    <map name=\"Normal\" >beast.math.distributions.Normal</map>\r\n",
      "    \r\n",
      "    <map name=\"Beta\" >beast.math.distributions.Beta</map>\r\n",
      "    \r\n",
      "    <map name=\"Gamma\" >beast.math.distributions.Gamma</map>\r\n",
      "    \r\n",
      "    <map name=\"LaplaceDistribution\" >beast.math.distributions.LaplaceDistribution</map>\r\n",
      "    \r\n",
      "    <map name=\"prior\" >beast.math.distributions.Prior</map>\r\n",
      "    \r\n",
      "    <map name=\"InverseGamma\" >beast.math.distributions.InverseGamma</map>\r\n",
      "    \r\n",
      "    <map name=\"OneOnX\" >beast.math.distributions.OneOnX</map>\r\n",
      "    \r\n",
      "    <run id=\"mcmc\" spec=\"MCMC\" chainLength=\"1000000\">\r\n",
      "            \r\n",
      "        <state id=\"state\" spec=\"State\" storeEvery=\"5000\">\r\n",
      "                    \r\n",
      "            <tree id=\"Tree.t:tree\" spec=\"beast.evolution.tree.Tree\" name=\"stateNode\">\r\n",
      "                            \r\n",
      "                <taxonset id=\"TaxonSet.noncoding\" spec=\"TaxonSet\">\r\n",
      "                                    \r\n",
      "                    <alignment id=\"noncoding\" spec=\"FilteredAlignment\" filter=\"1,458-659,897-898\">\r\n",
      "                                            \r\n",
      "                        <data idref=\"primate-mtDNA\"/>\r\n",
      "                                        \r\n",
      "                    </alignment>\r\n",
      "                                \r\n",
      "                </taxonset>\r\n",
      "                        \r\n",
      "            </tree>\r\n",
      "                    \r\n",
      "            <parameter id=\"mutationRate.s:noncoding\" spec=\"parameter.RealParameter\" name=\"stateNode\">1.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"gammaShape.s:noncoding\" spec=\"parameter.RealParameter\" name=\"stateNode\">1.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"kappa.s:noncoding\" spec=\"parameter.RealParameter\" lower=\"0.0\" name=\"stateNode\">2.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"kappa.s:1stpos\" spec=\"parameter.RealParameter\" lower=\"0.0\" name=\"stateNode\">2.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"mutationRate.s:1stpos\" spec=\"parameter.RealParameter\" name=\"stateNode\">1.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"gammaShape.s:1stpos\" spec=\"parameter.RealParameter\" name=\"stateNode\">1.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"kappa.s:2ndpos\" spec=\"parameter.RealParameter\" lower=\"0.0\" name=\"stateNode\">2.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"mutationRate.s:2ndpos\" spec=\"parameter.RealParameter\" name=\"stateNode\">1.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"gammaShape.s:2ndpos\" spec=\"parameter.RealParameter\" name=\"stateNode\">1.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"kappa.s:3rdpos\" spec=\"parameter.RealParameter\" lower=\"0.0\" name=\"stateNode\">2.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"mutationRate.s:3rdpos\" spec=\"parameter.RealParameter\" name=\"stateNode\">1.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"gammaShape.s:3rdpos\" spec=\"parameter.RealParameter\" name=\"stateNode\">1.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"birthRateY.t:tree\" spec=\"parameter.RealParameter\" name=\"stateNode\">1.0</parameter>\r\n",
      "                    \r\n",
      "            <parameter id=\"clockRate.c:clock\" spec=\"parameter.RealParameter\" name=\"stateNode\">1.0</parameter>\r\n",
      "                \r\n",
      "        </state>\r\n",
      "            \r\n",
      "        <init id=\"RandomTree.t:1stpos\" spec=\"beast.evolution.tree.RandomTree\" estimate=\"false\" initial=\"@Tree.t:tree\">\r\n",
      "                    \r\n",
      "            <taxa id=\"1stpos\" spec=\"FilteredAlignment\" data=\"@primate-mtDNA\" filter=\"2-457\\3,660-896\\3\"/>\r\n",
      "                    \r\n",
      "            <populationModel id=\"ConstantPopulation0.t:1stpos\" spec=\"ConstantPopulation\">\r\n",
      "                            \r\n",
      "                <parameter id=\"randomPopSize.t:1stpos\" spec=\"parameter.RealParameter\" name=\"popSize\">1.0</parameter>\r\n",
      "                        \r\n",
      "            </populationModel>\r\n",
      "                \r\n",
      "        </init>\r\n",
      "            \r\n",
      "        <distribution id=\"posterior\" spec=\"util.CompoundDistribution\">\r\n",
      "                    \r\n",
      "            <distribution id=\"prior\" spec=\"util.CompoundDistribution\">\r\n",
      "                            \r\n",
      "                <distribution id=\"CalibratedYuleModel.t:tree\" spec=\"beast.evolution.speciation.CalibratedYuleModel\" birthRate=\"@birthRateY.t:tree\" tree=\"@Tree.t:tree\"/>\r\n",
      "                            \r\n",
      "                <prior id=\"CalibratedYuleBirthRatePrior.t:tree\" name=\"distribution\" x=\"@birthRateY.t:tree\">\r\n",
      "                                    \r\n",
      "                    <Gamma id=\"Gamma.0\" name=\"distr\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.6\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"alpha\">0.001</parameter>\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.7\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"beta\">1000.0</parameter>\r\n",
      "                                        \r\n",
      "                    </Gamma>\r\n",
      "                                \r\n",
      "                </prior>\r\n",
      "                            \r\n",
      "                <prior id=\"ClockPrior.c:clock\" name=\"distribution\" x=\"@clockRate.c:clock\">\r\n",
      "                                    \r\n",
      "                    <Uniform id=\"Uniform.0\" name=\"distr\" upper=\"Infinity\"/>\r\n",
      "                                \r\n",
      "                </prior>\r\n",
      "                            \r\n",
      "                <prior id=\"GammaShapePrior.s:1stpos\" name=\"distribution\" x=\"@gammaShape.s:1stpos\">\r\n",
      "                                    \r\n",
      "                    <Exponential id=\"Exponential.0.1stpos\" name=\"distr\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.0.1stpos\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"mean\">1.0</parameter>\r\n",
      "                                        \r\n",
      "                    </Exponential>\r\n",
      "                                \r\n",
      "                </prior>\r\n",
      "                            \r\n",
      "                <prior id=\"GammaShapePrior.s:2ndpos\" name=\"distribution\" x=\"@gammaShape.s:2ndpos\">\r\n",
      "                                    \r\n",
      "                    <Exponential id=\"Exponential.0.2ndpos\" name=\"distr\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.0.2ndpos\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"mean\">1.0</parameter>\r\n",
      "                                        \r\n",
      "                    </Exponential>\r\n",
      "                                \r\n",
      "                </prior>\r\n",
      "                            \r\n",
      "                <prior id=\"GammaShapePrior.s:3rdpos\" name=\"distribution\" x=\"@gammaShape.s:3rdpos\">\r\n",
      "                                    \r\n",
      "                    <Exponential id=\"Exponential.0.3rdpos\" name=\"distr\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.0.3rdpos\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"mean\">1.0</parameter>\r\n",
      "                                        \r\n",
      "                    </Exponential>\r\n",
      "                                \r\n",
      "                </prior>\r\n",
      "                            \r\n",
      "                <prior id=\"GammaShapePrior.s:noncoding\" name=\"distribution\" x=\"@gammaShape.s:noncoding\">\r\n",
      "                                    \r\n",
      "                    <Exponential id=\"Exponential.0\" name=\"distr\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.0\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"mean\">1.0</parameter>\r\n",
      "                                        \r\n",
      "                    </Exponential>\r\n",
      "                                \r\n",
      "                </prior>\r\n",
      "                            \r\n",
      "                <prior id=\"KappaPrior.s:1stpos\" name=\"distribution\" x=\"@kappa.s:1stpos\">\r\n",
      "                                    \r\n",
      "                    <LogNormal id=\"LogNormalDistributionModel.0.1stpos\" name=\"distr\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.4.1stpos\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"M\">1.0</parameter>\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.5.1stpos\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"S\">1.25</parameter>\r\n",
      "                                        \r\n",
      "                    </LogNormal>\r\n",
      "                                \r\n",
      "                </prior>\r\n",
      "                            \r\n",
      "                <prior id=\"KappaPrior.s:2ndpos\" name=\"distribution\" x=\"@kappa.s:2ndpos\">\r\n",
      "                                    \r\n",
      "                    <LogNormal id=\"LogNormalDistributionModel.0.2ndpos\" name=\"distr\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.4.2ndpos\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"M\">1.0</parameter>\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.5.2ndpos\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"S\">1.25</parameter>\r\n",
      "                                        \r\n",
      "                    </LogNormal>\r\n",
      "                                \r\n",
      "                </prior>\r\n",
      "                            \r\n",
      "                <prior id=\"KappaPrior.s:3rdpos\" name=\"distribution\" x=\"@kappa.s:3rdpos\">\r\n",
      "                                    \r\n",
      "                    <LogNormal id=\"LogNormalDistributionModel.0.3rdpos\" name=\"distr\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.4.3rdpos\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"M\">1.0</parameter>\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.5.3rdpos\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"S\">1.25</parameter>\r\n",
      "                                        \r\n",
      "                    </LogNormal>\r\n",
      "                                \r\n",
      "                </prior>\r\n",
      "                            \r\n",
      "                <prior id=\"KappaPrior.s:noncoding\" name=\"distribution\" x=\"@kappa.s:noncoding\">\r\n",
      "                                    \r\n",
      "                    <LogNormal id=\"LogNormalDistributionModel.0\" name=\"distr\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.4\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"M\">1.0</parameter>\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.5\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"S\">1.25</parameter>\r\n",
      "                                        \r\n",
      "                    </LogNormal>\r\n",
      "                                \r\n",
      "                </prior>\r\n",
      "                            \r\n",
      "                <distribution id=\"human-chimp.prior\" spec=\"beast.math.distributions.MRCAPrior\" monophyletic=\"true\" tree=\"@Tree.t:tree\">\r\n",
      "                                    \r\n",
      "                    <taxonset id=\"human-chimp\" spec=\"TaxonSet\">\r\n",
      "                                            \r\n",
      "                        <taxon id=\"Homo_sapiens\" spec=\"Taxon\"/>\r\n",
      "                                            \r\n",
      "                        <taxon id=\"Pan\" spec=\"Taxon\"/>\r\n",
      "                                        \r\n",
      "                    </taxonset>\r\n",
      "                                    \r\n",
      "                    <Normal id=\"Normal.0\" name=\"distr\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.8\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"mean\">6.0</parameter>\r\n",
      "                                            \r\n",
      "                        <parameter id=\"RealParameter.9\" spec=\"parameter.RealParameter\" estimate=\"false\" name=\"sigma\">0.5</parameter>\r\n",
      "                                        \r\n",
      "                    </Normal>\r\n",
      "                                \r\n",
      "                </distribution>\r\n",
      "                        \r\n",
      "            </distribution>\r\n",
      "                    \r\n",
      "            <distribution id=\"likelihood\" spec=\"util.CompoundDistribution\" useThreads=\"true\">\r\n",
      "                            \r\n",
      "                <distribution id=\"treeLikelihood.noncoding\" spec=\"ThreadedTreeLikelihood\" data=\"@noncoding\" tree=\"@Tree.t:tree\">\r\n",
      "                                    \r\n",
      "                    <siteModel id=\"SiteModel.s:noncoding\" spec=\"SiteModel\" gammaCategoryCount=\"4\" mutationRate=\"@mutationRate.s:noncoding\" shape=\"@gammaShape.s:noncoding\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"proportionInvariant.s:noncoding\" spec=\"parameter.RealParameter\" estimate=\"false\" lower=\"0.0\" name=\"proportionInvariant\" upper=\"1.0\">0.0</parameter>\r\n",
      "                                            \r\n",
      "                        <substModel id=\"hky.s:noncoding\" spec=\"HKY\" kappa=\"@kappa.s:noncoding\">\r\n",
      "                                                    \r\n",
      "                            <frequencies id=\"empiricalFreqs.s:noncoding\" spec=\"Frequencies\" data=\"@noncoding\"/>\r\n",
      "                                                \r\n",
      "                        </substModel>\r\n",
      "                                        \r\n",
      "                    </siteModel>\r\n",
      "                                    \r\n",
      "                    <branchRateModel id=\"StrictClock.c:clock\" spec=\"beast.evolution.branchratemodel.StrictClockModel\" clock.rate=\"@clockRate.c:clock\"/>\r\n",
      "                                \r\n",
      "                </distribution>\r\n",
      "                            \r\n",
      "                <distribution id=\"treeLikelihood.1stpos\" spec=\"ThreadedTreeLikelihood\" branchRateModel=\"@StrictClock.c:clock\" data=\"@1stpos\" tree=\"@Tree.t:tree\">\r\n",
      "                                    \r\n",
      "                    <siteModel id=\"SiteModel.s:1stpos\" spec=\"SiteModel\" gammaCategoryCount=\"4\" mutationRate=\"@mutationRate.s:1stpos\" shape=\"@gammaShape.s:1stpos\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"proportionInvariant.s:1stpos\" spec=\"parameter.RealParameter\" estimate=\"false\" lower=\"0.0\" name=\"proportionInvariant\" upper=\"1.0\">0.0</parameter>\r\n",
      "                                            \r\n",
      "                        <substModel id=\"hky.s:1stpos\" spec=\"HKY\" kappa=\"@kappa.s:1stpos\">\r\n",
      "                                                    \r\n",
      "                            <frequencies id=\"empiricalFreqs.s:1stpos\" spec=\"Frequencies\" data=\"@1stpos\"/>\r\n",
      "                                                \r\n",
      "                        </substModel>\r\n",
      "                                        \r\n",
      "                    </siteModel>\r\n",
      "                                \r\n",
      "                </distribution>\r\n",
      "                            \r\n",
      "                <distribution id=\"treeLikelihood.2ndpos\" spec=\"ThreadedTreeLikelihood\" branchRateModel=\"@StrictClock.c:clock\" tree=\"@Tree.t:tree\">\r\n",
      "                                    \r\n",
      "                    <data\r\n",
      "id=\"2ndpos\"\r\n",
      "spec=\"FilteredAlignment\"\r\n",
      "data=\"@primate-mtDNA\"\r\n",
      "filter=\"3-457\\3,661-896\\3\"/>\r\n",
      "                                    \r\n",
      "                    <siteModel id=\"SiteModel.s:2ndpos\" spec=\"SiteModel\" gammaCategoryCount=\"4\" mutationRate=\"@mutationRate.s:2ndpos\" shape=\"@gammaShape.s:2ndpos\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"proportionInvariant.s:2ndpos\" spec=\"parameter.RealParameter\" estimate=\"false\" lower=\"0.0\" name=\"proportionInvariant\" upper=\"1.0\">0.0</parameter>\r\n",
      "                                            \r\n",
      "                        <substModel id=\"hky.s:2ndpos\" spec=\"HKY\" kappa=\"@kappa.s:2ndpos\">\r\n",
      "                                                    \r\n",
      "                            <frequencies id=\"empiricalFreqs.s:2ndpos\" spec=\"Frequencies\" data=\"@2ndpos\"/>\r\n",
      "                                                \r\n",
      "                        </substModel>\r\n",
      "                                        \r\n",
      "                    </siteModel>\r\n",
      "                                \r\n",
      "                </distribution>\r\n",
      "                            \r\n",
      "                <distribution id=\"treeLikelihood.3rdpos\" spec=\"ThreadedTreeLikelihood\" branchRateModel=\"@StrictClock.c:clock\" tree=\"@Tree.t:tree\">\r\n",
      "                                    \r\n",
      "                    <data\r\n",
      "id=\"3rdpos\"\r\n",
      "spec=\"FilteredAlignment\"\r\n",
      "data=\"@primate-mtDNA\"\r\n",
      "filter=\"4-457\\3,662-.\\3\"/>\r\n",
      "                                    \r\n",
      "                    <siteModel id=\"SiteModel.s:3rdpos\" spec=\"SiteModel\" gammaCategoryCount=\"4\" mutationRate=\"@mutationRate.s:3rdpos\" shape=\"@gammaShape.s:3rdpos\">\r\n",
      "                                            \r\n",
      "                        <parameter id=\"proportionInvariant.s:3rdpos\" spec=\"parameter.RealParameter\" estimate=\"false\" lower=\"0.0\" name=\"proportionInvariant\" upper=\"1.0\">0.0</parameter>\r\n",
      "                                            \r\n",
      "                        <substModel id=\"hky.s:3rdpos\" spec=\"HKY\" kappa=\"@kappa.s:3rdpos\">\r\n",
      "                                                    \r\n",
      "                            <frequencies id=\"empiricalFreqs.s:3rdpos\" spec=\"Frequencies\" data=\"@3rdpos\"/>\r\n",
      "                                                \r\n",
      "                        </substModel>\r\n",
      "                                        \r\n",
      "                    </siteModel>\r\n",
      "                                \r\n",
      "                </distribution>\r\n",
      "                        \r\n",
      "            </distribution>\r\n",
      "                \r\n",
      "        </distribution>\r\n",
      "            \r\n",
      "        <operator id=\"FixMeanMutationRatesOperator\" spec=\"DeltaExchangeOperator\" delta=\"0.75\" weight=\"2.0\">\r\n",
      "                    \r\n",
      "            <parameter idref=\"mutationRate.s:noncoding\"/>\r\n",
      "                    \r\n",
      "            <parameter idref=\"mutationRate.s:1stpos\"/>\r\n",
      "                    \r\n",
      "            <parameter idref=\"mutationRate.s:2ndpos\"/>\r\n",
      "                    \r\n",
      "            <parameter idref=\"mutationRate.s:3rdpos\"/>\r\n",
      "                    \r\n",
      "            <weightvector id=\"weightparameter\" spec=\"parameter.IntegerParameter\" dimension=\"4\" estimate=\"false\" lower=\"0\" upper=\"0\">205 231 231 231</weightvector>\r\n",
      "                \r\n",
      "        </operator>\r\n",
      "            \r\n",
      "        <operator id=\"gammaShapeScaler.s:noncoding\" spec=\"ScaleOperator\" parameter=\"@gammaShape.s:noncoding\" scaleFactor=\"0.5\" weight=\"0.1\"/>\r\n",
      "            \r\n",
      "        <operator id=\"KappaScaler.s:noncoding\" spec=\"ScaleOperator\" parameter=\"@kappa.s:noncoding\" scaleFactor=\"0.5\" weight=\"0.1\"/>\r\n",
      "            \r\n",
      "        <operator id=\"gammaShapeScaler.s:1stpos\" spec=\"ScaleOperator\" parameter=\"@gammaShape.s:1stpos\" scaleFactor=\"0.5\" weight=\"0.1\"/>\r\n",
      "            \r\n",
      "        <operator id=\"KappaScaler.s:1stpos\" spec=\"ScaleOperator\" parameter=\"@kappa.s:1stpos\" scaleFactor=\"0.5\" weight=\"0.1\"/>\r\n",
      "            \r\n",
      "        <operator id=\"gammaShapeScaler.s:2ndpos\" spec=\"ScaleOperator\" parameter=\"@gammaShape.s:2ndpos\" scaleFactor=\"0.5\" weight=\"0.1\"/>\r\n",
      "            \r\n",
      "        <operator id=\"KappaScaler.s:2ndpos\" spec=\"ScaleOperator\" parameter=\"@kappa.s:2ndpos\" scaleFactor=\"0.5\" weight=\"0.1\"/>\r\n",
      "            \r\n",
      "        <operator id=\"gammaShapeScaler.s:3rdpos\" spec=\"ScaleOperator\" parameter=\"@gammaShape.s:3rdpos\" scaleFactor=\"0.5\" weight=\"0.1\"/>\r\n",
      "            \r\n",
      "        <operator id=\"KappaScaler.s:3rdpos\" spec=\"ScaleOperator\" parameter=\"@kappa.s:3rdpos\" scaleFactor=\"0.5\" weight=\"0.1\"/>\r\n",
      "            \r\n",
      "        <operator id=\"CalibratedYuleModelTreeScaler.t:tree\" spec=\"ScaleOperator\" scaleFactor=\"0.5\" tree=\"@Tree.t:tree\" weight=\"3.0\"/>\r\n",
      "            \r\n",
      "        <operator id=\"CalibratedYuleModelTreeRootScaler.t:tree\" spec=\"ScaleOperator\" rootOnly=\"true\" scaleFactor=\"0.5\" tree=\"@Tree.t:tree\" weight=\"3.0\"/>\r\n",
      "            \r\n",
      "        <operator id=\"CalibratedYuleModelUniformOperator.t:tree\" spec=\"Uniform\" tree=\"@Tree.t:tree\" weight=\"30.0\"/>\r\n",
      "            \r\n",
      "        <operator id=\"CalibratedYuleModelSubtreeSlide.t:tree\" spec=\"SubtreeSlide\" tree=\"@Tree.t:tree\" weight=\"15.0\"/>\r\n",
      "            \r\n",
      "        <operator id=\"CalibratedYuleModelNarrow.t:tree\" spec=\"Exchange\" tree=\"@Tree.t:tree\" weight=\"15.0\"/>\r\n",
      "            \r\n",
      "        <operator id=\"CalibratedYuleModelWide.t:tree\" spec=\"Exchange\" isNarrow=\"false\" tree=\"@Tree.t:tree\" weight=\"3.0\"/>\r\n",
      "            \r\n",
      "        <operator id=\"CalibratedYuleModelWilsonBalding.t:tree\" spec=\"WilsonBalding\" tree=\"@Tree.t:tree\" weight=\"3.0\"/>\r\n",
      "            \r\n",
      "        <operator id=\"CalibratedYuleBirthRateScaler.t:tree\" spec=\"ScaleOperator\" parameter=\"@birthRateY.t:tree\" weight=\"3.0\"/>\r\n",
      "            \r\n",
      "        <operator id=\"StrictClockRateScaler.c:clock\" spec=\"ScaleOperator\" parameter=\"@clockRate.c:clock\" weight=\"3.0\"/>\r\n",
      "            \r\n",
      "        <operator id=\"strictClockUpDownOperator.c:clock\" spec=\"UpDownOperator\" scaleFactor=\"0.75\" weight=\"3.0\">\r\n",
      "                    \r\n",
      "            <up idref=\"clockRate.c:clock\"/>\r\n",
      "                    \r\n",
      "            <down idref=\"Tree.t:tree\"/>\r\n",
      "                \r\n",
      "        </operator>\r\n",
      "            \r\n",
      "        <logger id=\"tracelog\" spec=\"Logger\" fileName=\"primate-mtDNA.log\" logEvery=\"200\" model=\"@posterior\" sanitiseHeaders=\"true\" sort=\"smart\">\r\n",
      "                    \r\n",
      "            <log idref=\"posterior\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"likelihood\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"prior\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"treeLikelihood.1stpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"treeLikelihood.3rdpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"treeLikelihood.noncoding\"/>\r\n",
      "                    \r\n",
      "            <log id=\"TreeHeight.t:tree\" spec=\"beast.evolution.tree.TreeHeightLogger\" tree=\"@Tree.t:tree\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"treeLikelihood.2ndpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"mutationRate.s:noncoding\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"gammaShape.s:noncoding\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"kappa.s:noncoding\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"kappa.s:1stpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"mutationRate.s:1stpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"gammaShape.s:1stpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"kappa.s:2ndpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"mutationRate.s:2ndpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"gammaShape.s:2ndpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"kappa.s:3rdpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"mutationRate.s:3rdpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"gammaShape.s:3rdpos\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"CalibratedYuleModel.t:tree\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"birthRateY.t:tree\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"human-chimp.prior\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"clockRate.c:clock\"/>\r\n",
      "                \r\n",
      "        </logger>\r\n",
      "            \r\n",
      "        <logger id=\"screenlog\" spec=\"Logger\" logEvery=\"1000\">\r\n",
      "                    \r\n",
      "            <log idref=\"posterior\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"likelihood\"/>\r\n",
      "                    \r\n",
      "            <log idref=\"prior\"/>\r\n",
      "                \r\n",
      "        </logger>\r\n",
      "            \r\n",
      "        <logger id=\"treelog.t:tree\" spec=\"Logger\" fileName=\"primate-mtDNA.trees\" logEvery=\"1000\" mode=\"tree\">\r\n",
      "                    \r\n",
      "            <log id=\"TreeWithMetaDataLogger.t:tree\" spec=\"beast.evolution.tree.TreeWithMetaDataLogger\" tree=\"@Tree.t:tree\"/>\r\n",
      "                \r\n",
      "        </logger>\r\n",
      "            \r\n",
      "        <operatorschedule id=\"OperatorSchedule\" spec=\"OperatorSchedule\"/>\r\n",
      "        \r\n",
      "    </run>\r\n",
      "    \r\n",
      "</beast>\r\n"
     ]
    }
   ],
   "source": [
    "! cat ../data/BEAST2_tutorial/Primates.xml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf94c867",
   "metadata": {},
   "source": [
    "# Tonight: begin at \"Analysing the posterior estimate quality\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d286f3d6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
